// @devvit/protos/types/devvit/dev_portal/reddit/redditor.js
/**
 * #redditor.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseRedditor() {
    return { id: "", displayName: "", snoovatarIconUrl: "" };
}
export const Redditor = {
    $type: "devvit.dev_portal.reddit.Redditor",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.displayName !== "") {
            writer.uint32(18).string(message.displayName);
        }
        if (message.snoovatarIconUrl !== "") {
            writer.uint32(26).string(message.snoovatarIconUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRedditor();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.displayName = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.snoovatarIconUrl = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            displayName: isSet(object.displayName) ? globalThis.String(object.displayName) : "",
            snoovatarIconUrl: isSet(object.snoovatarIconUrl) ? globalThis.String(object.snoovatarIconUrl) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.displayName !== "") {
            obj.displayName = message.displayName;
        }
        if (message.snoovatarIconUrl !== "") {
            obj.snoovatarIconUrl = message.snoovatarIconUrl;
        }
        return obj;
    },
    create(base) {
        return Redditor.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRedditor();
        message.id = object.id ?? "";
        message.displayName = object.displayName ?? "";
        message.snoovatarIconUrl = object.snoovatarIconUrl ?? "";
        return message;
    },
};
messageTypeRegistry.set(Redditor.$type, Redditor);
function isSet(value) {
    return value !== null && value !== undefined;
}
