// @devvit/protos/types/devvit/plugin/v2events/v2events.js
/**
 * #v2events.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseV2EventsActionInfo() {
    return { pageType: "", paneName: "" };
}
export const V2EventsActionInfo = {
    $type: "devvit.plugin.v2events.V2EventsActionInfo",
    encode(message, writer = _m0.Writer.create()) {
        if (message.pageType !== "") {
            writer.uint32(10).string(message.pageType);
        }
        if (message.paneName !== "") {
            writer.uint32(18).string(message.paneName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseV2EventsActionInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pageType = reader.string();
                    break;
                case 2:
                    message.paneName = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            pageType: isSet(object.pageType) ? String(object.pageType) : "",
            paneName: isSet(object.paneName) ? String(object.paneName) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.pageType !== undefined && (obj.pageType = message.pageType);
        message.paneName !== undefined && (obj.paneName = message.paneName);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseV2EventsActionInfo();
        message.pageType = object.pageType ?? "";
        message.paneName = object.paneName ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.v2events.V2EventsActionInfo", V2EventsActionInfo);
function createBaseV2EventsPostGuidanceCreate() {
    return { actionInfo: undefined };
}
export const V2EventsPostGuidanceCreate = {
    $type: "devvit.plugin.v2events.V2EventsPostGuidanceCreate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.actionInfo !== undefined) {
            V2EventsActionInfo.encode(message.actionInfo, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseV2EventsPostGuidanceCreate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.actionInfo = V2EventsActionInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { actionInfo: isSet(object.actionInfo) ? V2EventsActionInfo.fromJSON(object.actionInfo) : undefined };
    },
    toJSON(message) {
        const obj = {};
        message.actionInfo !== undefined &&
            (obj.actionInfo = message.actionInfo ? V2EventsActionInfo.toJSON(message.actionInfo) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseV2EventsPostGuidanceCreate();
        message.actionInfo = (object.actionInfo !== undefined && object.actionInfo !== null)
            ? V2EventsActionInfo.fromPartial(object.actionInfo)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.v2events.V2EventsPostGuidanceCreate", V2EventsPostGuidanceCreate);
function createBaseV2EventsPostGuidanceUpdate() {
    return { actionInfo: undefined };
}
export const V2EventsPostGuidanceUpdate = {
    $type: "devvit.plugin.v2events.V2EventsPostGuidanceUpdate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.actionInfo !== undefined) {
            V2EventsActionInfo.encode(message.actionInfo, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseV2EventsPostGuidanceUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.actionInfo = V2EventsActionInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { actionInfo: isSet(object.actionInfo) ? V2EventsActionInfo.fromJSON(object.actionInfo) : undefined };
    },
    toJSON(message) {
        const obj = {};
        message.actionInfo !== undefined &&
            (obj.actionInfo = message.actionInfo ? V2EventsActionInfo.toJSON(message.actionInfo) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseV2EventsPostGuidanceUpdate();
        message.actionInfo = (object.actionInfo !== undefined && object.actionInfo !== null)
            ? V2EventsActionInfo.fromPartial(object.actionInfo)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.v2events.V2EventsPostGuidanceUpdate", V2EventsPostGuidanceUpdate);
function createBaseV2EventsPublishRequest() {
    return { postGuidanceCreate: undefined, postGuidanceUpdate: undefined };
}
export const V2EventsPublishRequest = {
    $type: "devvit.plugin.v2events.V2EventsPublishRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.postGuidanceCreate !== undefined) {
            V2EventsPostGuidanceCreate.encode(message.postGuidanceCreate, writer.uint32(10).fork()).ldelim();
        }
        if (message.postGuidanceUpdate !== undefined) {
            V2EventsPostGuidanceUpdate.encode(message.postGuidanceUpdate, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseV2EventsPublishRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.postGuidanceCreate = V2EventsPostGuidanceCreate.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.postGuidanceUpdate = V2EventsPostGuidanceUpdate.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            postGuidanceCreate: isSet(object.postGuidanceCreate)
                ? V2EventsPostGuidanceCreate.fromJSON(object.postGuidanceCreate)
                : undefined,
            postGuidanceUpdate: isSet(object.postGuidanceUpdate)
                ? V2EventsPostGuidanceUpdate.fromJSON(object.postGuidanceUpdate)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.postGuidanceCreate !== undefined && (obj.postGuidanceCreate = message.postGuidanceCreate
            ? V2EventsPostGuidanceCreate.toJSON(message.postGuidanceCreate)
            : undefined);
        message.postGuidanceUpdate !== undefined && (obj.postGuidanceUpdate = message.postGuidanceUpdate
            ? V2EventsPostGuidanceUpdate.toJSON(message.postGuidanceUpdate)
            : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseV2EventsPublishRequest();
        message.postGuidanceCreate = (object.postGuidanceCreate !== undefined && object.postGuidanceCreate !== null)
            ? V2EventsPostGuidanceCreate.fromPartial(object.postGuidanceCreate)
            : undefined;
        message.postGuidanceUpdate = (object.postGuidanceUpdate !== undefined && object.postGuidanceUpdate !== null)
            ? V2EventsPostGuidanceUpdate.fromPartial(object.postGuidanceUpdate)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.v2events.V2EventsPublishRequest", V2EventsPublishRequest);
function createBaseV2EventsPublishResponse() {
    return {};
}
export const V2EventsPublishResponse = {
    $type: "devvit.plugin.v2events.V2EventsPublishResponse",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseV2EventsPublishResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseV2EventsPublishResponse();
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.v2events.V2EventsPublishResponse", V2EventsPublishResponse);
export class V2EventsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.v2events.V2Events";
        this.rpc = rpc;
        this.Publish = this.Publish.bind(this);
    }
    Publish(request) {
        const data = V2EventsPublishRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Publish", data);
        return promise.then((data) => V2EventsPublishResponse.decode(new _m0.Reader(data)));
    }
}
export const V2EventsDefinition = {
    name: "V2Events",
    fullName: "devvit.plugin.v2events.V2Events",
    methods: {
        publish: {
            name: "Publish",
            requestType: V2EventsPublishRequest,
            requestStream: false,
            responseType: V2EventsPublishResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
