// @devvit/shared-types/tid.js
import { assert } from './assert.js';
// CALMS!
export var T_PREFIX;
(function (T_PREFIX) {
    T_PREFIX["COMMENT"] = "t1_";
    T_PREFIX["ACCOUNT"] = "t2_";
    T_PREFIX["LINK"] = "t3_";
    T_PREFIX["MESSAGE"] = "t4_";
    T_PREFIX["SUBREDDIT"] = "t5_";
    T_PREFIX["AWARD"] = "t6_";
})(T_PREFIX || (T_PREFIX = {}));
// type guards
export function isT1ID(id) {
    return id.startsWith(T_PREFIX.COMMENT);
}
export function isT2ID(id) {
    return id.startsWith(T_PREFIX.ACCOUNT);
}
export function isT3ID(id) {
    return id.startsWith(T_PREFIX.LINK);
}
export function isT4ID(id) {
    return id.startsWith(T_PREFIX.MESSAGE);
}
export function isT5ID(id) {
    return id.startsWith(T_PREFIX.SUBREDDIT);
}
export function isT6ID(id) {
    return id.startsWith(T_PREFIX.AWARD);
}
// assertion functions
export function assertT1ID(id) {
    assert(isT1ID(id), `Expected comment id to start with ${T_PREFIX.COMMENT}, got ${id}}`);
}
export function assertT2ID(id) {
    assert(isT2ID(id), `Expected account id to start with ${T_PREFIX.ACCOUNT}, got ${id}}`);
}
export function assertT3ID(id) {
    assert(isT3ID(id), `Expected link id to start with ${T_PREFIX.LINK}, got ${id}}`);
}
export function assertT4ID(id) {
    assert(isT4ID(id), `Expected message id to start with ${T_PREFIX.MESSAGE}, got ${id}}`);
}
export function assertT5ID(id) {
    assert(isT5ID(id), `Expected subreddit id to start with ${T_PREFIX.SUBREDDIT}, got ${id}}`);
}
export function assertT6ID(id) {
    assert(isT6ID(id), `Expected award id to start with ${T_PREFIX.AWARD}, got ${id}}`);
}
// factory functions
export function asT1ID(id) {
    assertT1ID(id);
    return id;
}
export function asT2ID(id) {
    assertT2ID(id);
    return id;
}
export function asT3ID(id) {
    assertT3ID(id);
    return id;
}
export function asT4ID(id) {
    assertT4ID(id);
    return id;
}
export function asT5ID(id) {
    assertT5ID(id);
    return id;
}
export function asT6ID(id) {
    assertT6ID(id);
    return id;
}
export function asTID(id) {
    if (isT1ID(id)) {
        return asT1ID(id);
    }
    if (isT2ID(id)) {
        return asT2ID(id);
    }
    if (isT3ID(id)) {
        return asT3ID(id);
    }
    if (isT4ID(id)) {
        return asT4ID(id);
    }
    if (isT5ID(id)) {
        return asT5ID(id);
    }
    if (isT6ID(id)) {
        return asT6ID(id);
    }
    throw new Error(`Expected thing id to start with ${Object.values(T_PREFIX).join(', ')} got ${id}}`);
}
// convenience functions
export function isCommentId(id) {
    return isT1ID(id);
}
export function isAccountId(id) {
    return isT2ID(id);
}
export function isLinkId(id) {
    return isT3ID(id);
}
export function isMessageId(id) {
    return isT4ID(id);
}
export function isSubredditId(id) {
    return isT5ID(id);
}
export function isAwardId(id) {
    return isT6ID(id);
}
