// @devvit/protos/types/devvit/plugin/settings/v1alpha/settings.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { SettingsRequest, SettingsResponse } from './settings.js';
export var SettingsMethod;
(function (SettingsMethod) {
    SettingsMethod["GetSettings"] = "GetSettings";
})(SettingsMethod || (SettingsMethod = {}));
export const SettingsMethodList = [SettingsMethod.GetSettings];
export function createSettingsServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.settings.v1alpha',
        serviceName: 'Settings',
        methodList: SettingsMethodList,
        matchRoute: matchSettingsRoute,
    });
}
function matchSettingsRoute(method, events) {
    switch (method) {
        case 'GetSettings':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetSettings' };
                await events.onMatch(ctx);
                return handleSettingsGetSettingsRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleSettingsGetSettingsRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleSettingsGetSettingsJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleSettingsGetSettingsProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleSettingsGetSettingsJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = SettingsRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSettings(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSettings(ctx, request);
    }
    return JSON.stringify(SettingsResponse.toJSON(response));
}
async function handleSettingsGetSettingsProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = SettingsRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSettings(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSettings(ctx, request);
    }
    return Buffer.from(SettingsResponse.encode(response).finish());
}
