// @devvit/protos/types/devvit/runtime/actor/runtime.js
/**
 * #runtime.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { LinkedBundle } from '../bundle.js';
function createBaseWorkerQuery() {
    return { hostname: "" };
}
export const WorkerQuery = {
    $type: "devvit.runtime.actor.WorkerQuery",
    encode(message, writer = _m0.Writer.create()) {
        if (message.hostname !== "") {
            writer.uint32(10).string(message.hostname);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWorkerQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.hostname = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { hostname: isSet(object.hostname) ? String(object.hostname) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.hostname !== undefined && (obj.hostname = message.hostname);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseWorkerQuery();
        message.hostname = object.hostname ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.runtime.actor.WorkerQuery", WorkerQuery);
export class RuntimeActorClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.runtime.actor.RuntimeActor";
        this.rpc = rpc;
        this.NewWorker = this.NewWorker.bind(this);
        this.TerminateWorker = this.TerminateWorker.bind(this);
    }
    NewWorker(request) {
        const data = LinkedBundle.encode(request).finish();
        const promise = this.rpc.request(this.service, "NewWorker", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    TerminateWorker(request) {
        const data = WorkerQuery.encode(request).finish();
        const promise = this.rpc.request(this.service, "TerminateWorker", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const RuntimeActorDefinition = {
    name: "RuntimeActor",
    fullName: "devvit.runtime.actor.RuntimeActor",
    methods: {
        newWorker: {
            name: "NewWorker",
            requestType: LinkedBundle,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        terminateWorker: {
            name: "TerminateWorker",
            requestType: WorkerQuery,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
