// @devvit/protos/types/devvit/plugin/settings/v1alpha/settings.js
/**
 * #settings.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Timestamp } from '../../../../google/protobuf/timestamp.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
import { FormFieldValue } from '../../../ui/form_builder/v1alpha/value.js';
function createBaseSettingsValues() {
    return { version: "", settings: {}, timestamp: undefined };
}
export const SettingsValues = {
    $type: "devvit.plugin.settings.v1alpha.SettingsValues",
    encode(message, writer = _m0.Writer.create()) {
        if (message.version !== "") {
            writer.uint32(10).string(message.version);
        }
        Object.entries(message.settings).forEach(([key, value]) => {
            SettingsValues_SettingsEntry.encode({ key: key, value }, writer.uint32(18).fork()).ldelim();
        });
        if (message.timestamp !== undefined) {
            Timestamp.encode(toTimestamp(message.timestamp), writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSettingsValues();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.version = reader.string();
                    break;
                case 2:
                    const entry2 = SettingsValues_SettingsEntry.decode(reader, reader.uint32());
                    if (entry2.value !== undefined) {
                        message.settings[entry2.key] = entry2.value;
                    }
                    break;
                case 3:
                    message.timestamp = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            version: isSet(object.version) ? String(object.version) : "",
            settings: isObject(object.settings)
                ? Object.entries(object.settings).reduce((acc, [key, value]) => {
                    acc[key] = FormFieldValue.fromJSON(value);
                    return acc;
                }, {})
                : {},
            timestamp: isSet(object.timestamp) ? fromJsonTimestamp(object.timestamp) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.version !== undefined && (obj.version = message.version);
        obj.settings = {};
        if (message.settings) {
            Object.entries(message.settings).forEach(([k, v]) => {
                obj.settings[k] = FormFieldValue.toJSON(v);
            });
        }
        message.timestamp !== undefined && (obj.timestamp = message.timestamp.toISOString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseSettingsValues();
        message.version = object.version ?? "";
        message.settings = Object.entries(object.settings ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = FormFieldValue.fromPartial(value);
            }
            return acc;
        }, {});
        message.timestamp = object.timestamp ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.settings.v1alpha.SettingsValues", SettingsValues);
function createBaseSettingsValues_SettingsEntry() {
    return { key: "", value: undefined };
}
export const SettingsValues_SettingsEntry = {
    $type: "devvit.plugin.settings.v1alpha.SettingsValues.SettingsEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== undefined) {
            FormFieldValue.encode(message.value, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSettingsValues_SettingsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.string();
                    break;
                case 2:
                    message.value = FormFieldValue.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? String(object.key) : "",
            value: isSet(object.value) ? FormFieldValue.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined && (obj.value = message.value ? FormFieldValue.toJSON(message.value) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseSettingsValues_SettingsEntry();
        message.key = object.key ?? "";
        message.value = (object.value !== undefined && object.value !== null)
            ? FormFieldValue.fromPartial(object.value)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.settings.v1alpha.SettingsValues.SettingsEntry", SettingsValues_SettingsEntry);
function createBaseSettingsRequest() {
    return {};
}
export const SettingsRequest = {
    $type: "devvit.plugin.settings.v1alpha.SettingsRequest",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSettingsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseSettingsRequest();
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.settings.v1alpha.SettingsRequest", SettingsRequest);
function createBaseSettingsResponse() {
    return { installationSettings: undefined, instanceSettings: undefined, appSettings: undefined };
}
export const SettingsResponse = {
    $type: "devvit.plugin.settings.v1alpha.SettingsResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installationSettings !== undefined) {
            SettingsValues.encode(message.installationSettings, writer.uint32(10).fork()).ldelim();
        }
        if (message.instanceSettings !== undefined) {
            SettingsValues.encode(message.instanceSettings, writer.uint32(18).fork()).ldelim();
        }
        if (message.appSettings !== undefined) {
            SettingsValues.encode(message.appSettings, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSettingsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.installationSettings = SettingsValues.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.instanceSettings = SettingsValues.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.appSettings = SettingsValues.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            installationSettings: isSet(object.installationSettings)
                ? SettingsValues.fromJSON(object.installationSettings)
                : undefined,
            instanceSettings: isSet(object.instanceSettings) ? SettingsValues.fromJSON(object.instanceSettings) : undefined,
            appSettings: isSet(object.appSettings) ? SettingsValues.fromJSON(object.appSettings) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.installationSettings !== undefined && (obj.installationSettings = message.installationSettings
            ? SettingsValues.toJSON(message.installationSettings)
            : undefined);
        message.instanceSettings !== undefined &&
            (obj.instanceSettings = message.instanceSettings ? SettingsValues.toJSON(message.instanceSettings) : undefined);
        message.appSettings !== undefined &&
            (obj.appSettings = message.appSettings ? SettingsValues.toJSON(message.appSettings) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseSettingsResponse();
        message.installationSettings = (object.installationSettings !== undefined && object.installationSettings !== null)
            ? SettingsValues.fromPartial(object.installationSettings)
            : undefined;
        message.instanceSettings = (object.instanceSettings !== undefined && object.instanceSettings !== null)
            ? SettingsValues.fromPartial(object.instanceSettings)
            : undefined;
        message.appSettings = (object.appSettings !== undefined && object.appSettings !== null)
            ? SettingsValues.fromPartial(object.appSettings)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.settings.v1alpha.SettingsResponse", SettingsResponse);
export class SettingsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.settings.v1alpha.Settings";
        this.rpc = rpc;
        this.GetSettings = this.GetSettings.bind(this);
    }
    GetSettings(request) {
        const data = SettingsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetSettings", data);
        return promise.then((data) => SettingsResponse.decode(new _m0.Reader(data)));
    }
}
export const SettingsDefinition = {
    name: "Settings",
    fullName: "devvit.plugin.settings.v1alpha.Settings",
    methods: {
        /** Returns the installation and instance settings */
        getSettings: {
            name: "GetSettings",
            requestType: SettingsRequest,
            requestStream: false,
            responseType: SettingsResponse,
            responseStream: false,
            options: {},
        },
    },
};
function toTimestamp(date) {
    const seconds = date.getTime() / 1000;
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = t.seconds * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
