// @devvit/protos/types/devvit/plugin/redditapi/privatemessages/privatemessages_svc.js
/**
 * #privatemessages_svc.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../../google/protobuf/empty.js';
import { Listing } from '../common/common_msg.js';
import { ComposeRequest, GenericPrivateMessagesRequest, MessageWhereRequest, ReadAllMessagesRequest, } from './privatemessages_msg.js';
export class PrivateMessagesClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.redditapi.privatemessages.PrivateMessages";
        this.rpc = rpc;
        this.Block = this.Block.bind(this);
        this.CollapseMessage = this.CollapseMessage.bind(this);
        this.Compose = this.Compose.bind(this);
        this.DelMsg = this.DelMsg.bind(this);
        this.ReadAllMessages = this.ReadAllMessages.bind(this);
        this.ReadMessage = this.ReadMessage.bind(this);
        this.UnblockSubreddit = this.UnblockSubreddit.bind(this);
        this.UncollapseMessage = this.UncollapseMessage.bind(this);
        this.UnreadMessage = this.UnreadMessage.bind(this);
        this.MessageWhere = this.MessageWhere.bind(this);
    }
    Block(request) {
        const data = GenericPrivateMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Block", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    CollapseMessage(request) {
        const data = GenericPrivateMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CollapseMessage", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    Compose(request) {
        const data = ComposeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Compose", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    DelMsg(request) {
        const data = GenericPrivateMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DelMsg", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    ReadAllMessages(request) {
        const data = ReadAllMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ReadAllMessages", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    ReadMessage(request) {
        const data = GenericPrivateMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ReadMessage", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    UnblockSubreddit(request) {
        const data = GenericPrivateMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UnblockSubreddit", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    UncollapseMessage(request) {
        const data = GenericPrivateMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UncollapseMessage", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    UnreadMessage(request) {
        const data = GenericPrivateMessagesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UnreadMessage", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    MessageWhere(request) {
        const data = MessageWhereRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "MessageWhere", data);
        return promise.then((data) => Listing.decode(new _m0.Reader(data)));
    }
}
export const PrivateMessagesDefinition = {
    name: "PrivateMessages",
    fullName: "devvit.plugin.redditapi.privatemessages.PrivateMessages",
    methods: {
        /** https://www.reddit.com/dev/api#POST_api_block */
        block: {
            name: "Block",
            requestType: GenericPrivateMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_collapse_message */
        collapseMessage: {
            name: "CollapseMessage",
            requestType: GenericPrivateMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_compose */
        compose: {
            name: "Compose",
            requestType: ComposeRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_del_msg */
        delMsg: {
            name: "DelMsg",
            requestType: GenericPrivateMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_read_all_messages */
        readAllMessages: {
            name: "ReadAllMessages",
            requestType: ReadAllMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_read_message */
        readMessage: {
            name: "ReadMessage",
            requestType: GenericPrivateMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_unblock_subreddit */
        unblockSubreddit: {
            name: "UnblockSubreddit",
            requestType: GenericPrivateMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_uncollapse_message */
        uncollapseMessage: {
            name: "UncollapseMessage",
            requestType: GenericPrivateMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_unread_message */
        unreadMessage: {
            name: "UnreadMessage",
            requestType: GenericPrivateMessagesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_message_where */
        messageWhere: {
            name: "MessageWhere",
            requestType: MessageWhereRequest,
            requestStream: false,
            responseType: Listing,
            responseStream: false,
            options: {},
        },
    },
};
