// @devvit/protos/types/devvit/plugin/redditapi/privatemessages/privatemessages_msg.js
/**
 * #privatemessages_msg.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { BoolValue, Int64Value, StringValue } from '../../../../google/protobuf/wrappers.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
function createBaseGenericPrivateMessagesRequest() {
    return { id: "" };
}
export const GenericPrivateMessagesRequest = {
    $type: "devvit.plugin.redditapi.privatemessages.GenericPrivateMessagesRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenericPrivateMessagesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { id: isSet(object.id) ? String(object.id) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGenericPrivateMessagesRequest();
        message.id = object.id ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.redditapi.privatemessages.GenericPrivateMessagesRequest", GenericPrivateMessagesRequest);
function createBaseComposeRequest() {
    return { to: "", subject: "", text: "", fromSr: "" };
}
export const ComposeRequest = {
    $type: "devvit.plugin.redditapi.privatemessages.ComposeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.to !== "") {
            writer.uint32(10).string(message.to);
        }
        if (message.subject !== "") {
            writer.uint32(18).string(message.subject);
        }
        if (message.text !== "") {
            writer.uint32(26).string(message.text);
        }
        if (message.fromSr !== "") {
            writer.uint32(34).string(message.fromSr);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseComposeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.to = reader.string();
                    break;
                case 2:
                    message.subject = reader.string();
                    break;
                case 3:
                    message.text = reader.string();
                    break;
                case 4:
                    message.fromSr = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            to: isSet(object.to) ? String(object.to) : "",
            subject: isSet(object.subject) ? String(object.subject) : "",
            text: isSet(object.text) ? String(object.text) : "",
            fromSr: isSet(object.fromSr) ? String(object.fromSr) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.to !== undefined && (obj.to = message.to);
        message.subject !== undefined && (obj.subject = message.subject);
        message.text !== undefined && (obj.text = message.text);
        message.fromSr !== undefined && (obj.fromSr = message.fromSr);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseComposeRequest();
        message.to = object.to ?? "";
        message.subject = object.subject ?? "";
        message.text = object.text ?? "";
        message.fromSr = object.fromSr ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.redditapi.privatemessages.ComposeRequest", ComposeRequest);
function createBaseReadAllMessagesRequest() {
    return { filterTypes: "" };
}
export const ReadAllMessagesRequest = {
    $type: "devvit.plugin.redditapi.privatemessages.ReadAllMessagesRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.filterTypes !== "") {
            writer.uint32(10).string(message.filterTypes);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReadAllMessagesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.filterTypes = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { filterTypes: isSet(object.filterTypes) ? String(object.filterTypes) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.filterTypes !== undefined && (obj.filterTypes = message.filterTypes);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseReadAllMessagesRequest();
        message.filterTypes = object.filterTypes ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.redditapi.privatemessages.ReadAllMessagesRequest", ReadAllMessagesRequest);
function createBaseMessageWhereRequest() {
    return {
        where: "",
        mark: undefined,
        mid: undefined,
        after: undefined,
        before: undefined,
        count: undefined,
        limit: undefined,
        show: undefined,
    };
}
export const MessageWhereRequest = {
    $type: "devvit.plugin.redditapi.privatemessages.MessageWhereRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.where !== "") {
            writer.uint32(10).string(message.where);
        }
        if (message.mark !== undefined) {
            BoolValue.encode({ value: message.mark }, writer.uint32(18).fork()).ldelim();
        }
        if (message.mid !== undefined) {
            StringValue.encode({ value: message.mid }, writer.uint32(26).fork()).ldelim();
        }
        if (message.after !== undefined) {
            StringValue.encode({ value: message.after }, writer.uint32(34).fork()).ldelim();
        }
        if (message.before !== undefined) {
            StringValue.encode({ value: message.before }, writer.uint32(42).fork()).ldelim();
        }
        if (message.count !== undefined) {
            Int64Value.encode({ value: message.count }, writer.uint32(50).fork()).ldelim();
        }
        if (message.limit !== undefined) {
            Int64Value.encode({ value: message.limit }, writer.uint32(58).fork()).ldelim();
        }
        if (message.show !== undefined) {
            StringValue.encode({ value: message.show }, writer.uint32(66).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMessageWhereRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.where = reader.string();
                    break;
                case 2:
                    message.mark = BoolValue.decode(reader, reader.uint32()).value;
                    break;
                case 3:
                    message.mid = StringValue.decode(reader, reader.uint32()).value;
                    break;
                case 4:
                    message.after = StringValue.decode(reader, reader.uint32()).value;
                    break;
                case 5:
                    message.before = StringValue.decode(reader, reader.uint32()).value;
                    break;
                case 6:
                    message.count = Int64Value.decode(reader, reader.uint32()).value;
                    break;
                case 7:
                    message.limit = Int64Value.decode(reader, reader.uint32()).value;
                    break;
                case 8:
                    message.show = StringValue.decode(reader, reader.uint32()).value;
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            where: isSet(object.where) ? String(object.where) : "",
            mark: isSet(object.mark) ? Boolean(object.mark) : undefined,
            mid: isSet(object.mid) ? String(object.mid) : undefined,
            after: isSet(object.after) ? String(object.after) : undefined,
            before: isSet(object.before) ? String(object.before) : undefined,
            count: isSet(object.count) ? Number(object.count) : undefined,
            limit: isSet(object.limit) ? Number(object.limit) : undefined,
            show: isSet(object.show) ? String(object.show) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.where !== undefined && (obj.where = message.where);
        message.mark !== undefined && (obj.mark = message.mark);
        message.mid !== undefined && (obj.mid = message.mid);
        message.after !== undefined && (obj.after = message.after);
        message.before !== undefined && (obj.before = message.before);
        message.count !== undefined && (obj.count = message.count);
        message.limit !== undefined && (obj.limit = message.limit);
        message.show !== undefined && (obj.show = message.show);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMessageWhereRequest();
        message.where = object.where ?? "";
        message.mark = object.mark ?? undefined;
        message.mid = object.mid ?? undefined;
        message.after = object.after ?? undefined;
        message.before = object.before ?? undefined;
        message.count = object.count ?? undefined;
        message.limit = object.limit ?? undefined;
        message.show = object.show ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.redditapi.privatemessages.MessageWhereRequest", MessageWhereRequest);
function isSet(value) {
    return value !== null && value !== undefined;
}
