// @devvit/protos/types/devvit/actor/bot/bot.js
/**
 * #bot.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseCommandDescription() {
    return { name: "", command: "", description: "", usage: "" };
}
export const CommandDescription = {
    $type: "devvit.actor.bot.CommandDescription",
    encode(message, writer = _m0.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.command !== "") {
            writer.uint32(18).string(message.command);
        }
        if (message.description !== "") {
            writer.uint32(26).string(message.description);
        }
        if (message.usage !== "") {
            writer.uint32(34).string(message.usage);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommandDescription();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.command = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.description = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.usage = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            command: isSet(object.command) ? globalThis.String(object.command) : "",
            description: isSet(object.description) ? globalThis.String(object.description) : "",
            usage: isSet(object.usage) ? globalThis.String(object.usage) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.command !== "") {
            obj.command = message.command;
        }
        if (message.description !== "") {
            obj.description = message.description;
        }
        if (message.usage !== "") {
            obj.usage = message.usage;
        }
        return obj;
    },
    create(base) {
        return CommandDescription.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCommandDescription();
        message.name = object.name ?? "";
        message.command = object.command ?? "";
        message.description = object.description ?? "";
        message.usage = object.usage ?? "";
        return message;
    },
};
messageTypeRegistry.set(CommandDescription.$type, CommandDescription);
function createBaseCommandsList() {
    return { commands: [] };
}
export const CommandsList = {
    $type: "devvit.actor.bot.CommandsList",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.commands) {
            CommandDescription.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommandsList();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.commands.push(CommandDescription.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            commands: globalThis.Array.isArray(object?.commands)
                ? object.commands.map((e) => CommandDescription.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.commands?.length) {
            obj.commands = message.commands.map((e) => CommandDescription.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return CommandsList.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCommandsList();
        message.commands = object.commands?.map((e) => CommandDescription.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set(CommandsList.$type, CommandsList);
function createBaseCommandContext() {
    return { user: "", subreddit: "", post: "", comment: "" };
}
export const CommandContext = {
    $type: "devvit.actor.bot.CommandContext",
    encode(message, writer = _m0.Writer.create()) {
        if (message.user !== "") {
            writer.uint32(10).string(message.user);
        }
        if (message.subreddit !== "") {
            writer.uint32(18).string(message.subreddit);
        }
        if (message.post !== "") {
            writer.uint32(26).string(message.post);
        }
        if (message.comment !== "") {
            writer.uint32(34).string(message.comment);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommandContext();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.user = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.subreddit = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.post = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.comment = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            user: isSet(object.user) ? globalThis.String(object.user) : "",
            subreddit: isSet(object.subreddit) ? globalThis.String(object.subreddit) : "",
            post: isSet(object.post) ? globalThis.String(object.post) : "",
            comment: isSet(object.comment) ? globalThis.String(object.comment) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.user !== "") {
            obj.user = message.user;
        }
        if (message.subreddit !== "") {
            obj.subreddit = message.subreddit;
        }
        if (message.post !== "") {
            obj.post = message.post;
        }
        if (message.comment !== "") {
            obj.comment = message.comment;
        }
        return obj;
    },
    create(base) {
        return CommandContext.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCommandContext();
        message.user = object.user ?? "";
        message.subreddit = object.subreddit ?? "";
        message.post = object.post ?? "";
        message.comment = object.comment ?? "";
        return message;
    },
};
messageTypeRegistry.set(CommandContext.$type, CommandContext);
function createBaseCommandRequest() {
    return { context: undefined, command: "", args: "" };
}
export const CommandRequest = {
    $type: "devvit.actor.bot.CommandRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.context !== undefined) {
            CommandContext.encode(message.context, writer.uint32(10).fork()).ldelim();
        }
        if (message.command !== "") {
            writer.uint32(18).string(message.command);
        }
        if (message.args !== "") {
            writer.uint32(26).string(message.args);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommandRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.context = CommandContext.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.command = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.args = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            context: isSet(object.context) ? CommandContext.fromJSON(object.context) : undefined,
            command: isSet(object.command) ? globalThis.String(object.command) : "",
            args: isSet(object.args) ? globalThis.String(object.args) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.context !== undefined) {
            obj.context = CommandContext.toJSON(message.context);
        }
        if (message.command !== "") {
            obj.command = message.command;
        }
        if (message.args !== "") {
            obj.args = message.args;
        }
        return obj;
    },
    create(base) {
        return CommandRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCommandRequest();
        message.context = (object.context !== undefined && object.context !== null)
            ? CommandContext.fromPartial(object.context)
            : undefined;
        message.command = object.command ?? "";
        message.args = object.args ?? "";
        return message;
    },
};
messageTypeRegistry.set(CommandRequest.$type, CommandRequest);
export const BotServiceName = "devvit.actor.bot.Bot";
export class BotClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || BotServiceName;
        this.rpc = rpc;
        this.GetCommands = this.GetCommands.bind(this);
        this.OnCommand = this.OnCommand.bind(this);
    }
    GetCommands(request, metadata) {
        const data = Empty.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetCommands", data, metadata);
        return promise.then((data) => CommandsList.decode(_m0.Reader.create(data)));
    }
    OnCommand(request, metadata) {
        const data = CommandRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "OnCommand", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
}
export const BotDefinition = {
    name: "Bot",
    fullName: "devvit.actor.bot.Bot",
    methods: {
        getCommands: {
            name: "GetCommands",
            requestType: Empty,
            requestStream: false,
            responseType: CommandsList,
            responseStream: false,
            options: {},
        },
        onCommand: {
            name: "OnCommand",
            requestType: CommandRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
