// @devvit/protos/types/devvit/plugin/gl/methodtype.js
/**
 * #methodtype.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
export var GLMethodType;
(function (GLMethodType) {
    GLMethodType[GLMethodType["ASYNC"] = 0] = "ASYNC";
    GLMethodType[GLMethodType["CONSTRUCTOR"] = 1] = "CONSTRUCTOR";
    GLMethodType[GLMethodType["DESTRUCTOR"] = 2] = "DESTRUCTOR";
    GLMethodType[GLMethodType["SYNC"] = 3] = "SYNC";
    GLMethodType[GLMethodType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(GLMethodType || (GLMethodType = {}));
export function gLMethodTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "ASYNC":
            return GLMethodType.ASYNC;
        case 1:
        case "CONSTRUCTOR":
            return GLMethodType.CONSTRUCTOR;
        case 2:
        case "DESTRUCTOR":
            return GLMethodType.DESTRUCTOR;
        case 3:
        case "SYNC":
            return GLMethodType.SYNC;
        case -1:
        case "UNRECOGNIZED":
        default:
            return GLMethodType.UNRECOGNIZED;
    }
}
export function gLMethodTypeToJSON(object) {
    switch (object) {
        case GLMethodType.ASYNC:
            return 0;
        case GLMethodType.CONSTRUCTOR:
            return 1;
        case GLMethodType.DESTRUCTOR:
            return 2;
        case GLMethodType.SYNC:
            return 3;
        case GLMethodType.UNRECOGNIZED:
        default:
            return -1;
    }
}
