// @devvit/protos/types/devvit/plugin/assetresolver/assetresolver.js
/**
 * #assetresolver.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseGetAssetURLRequest() {
    return { path: "" };
}
export const GetAssetURLRequest = {
    $type: "devvit.plugin.assetresolver.GetAssetURLRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.path !== "") {
            writer.uint32(10).string(message.path);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAssetURLRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.path = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { path: isSet(object.path) ? String(object.path) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.path !== undefined && (obj.path = message.path);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetAssetURLRequest();
        message.path = object.path ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.assetresolver.GetAssetURLRequest", GetAssetURLRequest);
function createBaseGetAssetURLResponse() {
    return { url: "", found: false };
}
export const GetAssetURLResponse = {
    $type: "devvit.plugin.assetresolver.GetAssetURLResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        if (message.found === true) {
            writer.uint32(16).bool(message.found);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAssetURLResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.url = reader.string();
                    break;
                case 2:
                    message.found = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            url: isSet(object.url) ? String(object.url) : "",
            found: isSet(object.found) ? Boolean(object.found) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.url !== undefined && (obj.url = message.url);
        message.found !== undefined && (obj.found = message.found);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetAssetURLResponse();
        message.url = object.url ?? "";
        message.found = object.found ?? false;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.assetresolver.GetAssetURLResponse", GetAssetURLResponse);
function createBaseGetMultipleAssetURLsRequest() {
    return { paths: [] };
}
export const GetMultipleAssetURLsRequest = {
    $type: "devvit.plugin.assetresolver.GetMultipleAssetURLsRequest",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.paths) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetMultipleAssetURLsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.paths.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { paths: Array.isArray(object?.paths) ? object.paths.map((e) => String(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.paths) {
            obj.paths = message.paths.map((e) => e);
        }
        else {
            obj.paths = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetMultipleAssetURLsRequest();
        message.paths = object.paths?.map((e) => e) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.assetresolver.GetMultipleAssetURLsRequest", GetMultipleAssetURLsRequest);
function createBaseGetMultipleAssetURLsResponse() {
    return { urls: {} };
}
export const GetMultipleAssetURLsResponse = {
    $type: "devvit.plugin.assetresolver.GetMultipleAssetURLsResponse",
    encode(message, writer = _m0.Writer.create()) {
        Object.entries(message.urls).forEach(([key, value]) => {
            GetMultipleAssetURLsResponse_UrlsEntry.encode({ key: key, value }, writer.uint32(10).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetMultipleAssetURLsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    const entry1 = GetMultipleAssetURLsResponse_UrlsEntry.decode(reader, reader.uint32());
                    if (entry1.value !== undefined) {
                        message.urls[entry1.key] = entry1.value;
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            urls: isObject(object.urls)
                ? Object.entries(object.urls).reduce((acc, [key, value]) => {
                    acc[key] = GetMultipleAssetURLsResponse_GlobResult.fromJSON(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        obj.urls = {};
        if (message.urls) {
            Object.entries(message.urls).forEach(([k, v]) => {
                obj.urls[k] = GetMultipleAssetURLsResponse_GlobResult.toJSON(v);
            });
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetMultipleAssetURLsResponse();
        message.urls = Object.entries(object.urls ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = GetMultipleAssetURLsResponse_GlobResult.fromPartial(value);
            }
            return acc;
        }, {});
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.assetresolver.GetMultipleAssetURLsResponse", GetMultipleAssetURLsResponse);
function createBaseGetMultipleAssetURLsResponse_GlobResult() {
    return { found: false, paths: [] };
}
export const GetMultipleAssetURLsResponse_GlobResult = {
    $type: "devvit.plugin.assetresolver.GetMultipleAssetURLsResponse.GlobResult",
    encode(message, writer = _m0.Writer.create()) {
        if (message.found === true) {
            writer.uint32(8).bool(message.found);
        }
        for (const v of message.paths) {
            writer.uint32(18).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetMultipleAssetURLsResponse_GlobResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.found = reader.bool();
                    break;
                case 2:
                    message.paths.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            found: isSet(object.found) ? Boolean(object.found) : false,
            paths: Array.isArray(object?.paths) ? object.paths.map((e) => String(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.found !== undefined && (obj.found = message.found);
        if (message.paths) {
            obj.paths = message.paths.map((e) => e);
        }
        else {
            obj.paths = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetMultipleAssetURLsResponse_GlobResult();
        message.found = object.found ?? false;
        message.paths = object.paths?.map((e) => e) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.assetresolver.GetMultipleAssetURLsResponse.GlobResult", GetMultipleAssetURLsResponse_GlobResult);
function createBaseGetMultipleAssetURLsResponse_UrlsEntry() {
    return { key: "", value: undefined };
}
export const GetMultipleAssetURLsResponse_UrlsEntry = {
    $type: "devvit.plugin.assetresolver.GetMultipleAssetURLsResponse.UrlsEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== undefined) {
            GetMultipleAssetURLsResponse_GlobResult.encode(message.value, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetMultipleAssetURLsResponse_UrlsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.string();
                    break;
                case 2:
                    message.value = GetMultipleAssetURLsResponse_GlobResult.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? String(object.key) : "",
            value: isSet(object.value) ? GetMultipleAssetURLsResponse_GlobResult.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined &&
            (obj.value = message.value ? GetMultipleAssetURLsResponse_GlobResult.toJSON(message.value) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetMultipleAssetURLsResponse_UrlsEntry();
        message.key = object.key ?? "";
        message.value = (object.value !== undefined && object.value !== null)
            ? GetMultipleAssetURLsResponse_GlobResult.fromPartial(object.value)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.assetresolver.GetMultipleAssetURLsResponse.UrlsEntry", GetMultipleAssetURLsResponse_UrlsEntry);
export class AssetResolverClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.assetresolver.AssetResolver";
        this.rpc = rpc;
        this.GetAssetURL = this.GetAssetURL.bind(this);
        this.GetAssetURLs = this.GetAssetURLs.bind(this);
    }
    GetAssetURL(request) {
        const data = GetAssetURLRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAssetURL", data);
        return promise.then((data) => GetAssetURLResponse.decode(new _m0.Reader(data)));
    }
    GetAssetURLs(request) {
        const data = GetMultipleAssetURLsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAssetURLs", data);
        return promise.then((data) => GetMultipleAssetURLsResponse.decode(new _m0.Reader(data)));
    }
}
export const AssetResolverDefinition = {
    name: "AssetResolver",
    fullName: "devvit.plugin.assetresolver.AssetResolver",
    methods: {
        /** Gets a public URL for a given asset */
        getAssetURL: {
            name: "GetAssetURL",
            requestType: GetAssetURLRequest,
            requestStream: false,
            responseType: GetAssetURLResponse,
            responseStream: false,
            options: {},
        },
        /** Gets the public URLs for multiple assets at the same time */
        getAssetURLs: {
            name: "GetAssetURLs",
            requestType: GetMultipleAssetURLsRequest,
            requestStream: false,
            responseType: GetMultipleAssetURLsResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
