// @devvit/protos/types/devvit/dev_portal/data_api/admin/organization/organization.js
/**
 * #organization.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../../typeRegistry.js';
import { Organization } from '../../../../data/api/v1alpha/admin.js';
function createBaseCreateOrganizationRequest() {
    return { name: "" };
}
export const CreateOrganizationRequest = {
    $type: "devvit.dev_portal.data_api.admin.organization.CreateOrganizationRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateOrganizationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { name: isSet(object.name) ? globalThis.String(object.name) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        return obj;
    },
    create(base) {
        return CreateOrganizationRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCreateOrganizationRequest();
        message.name = object.name ?? "";
        return message;
    },
};
messageTypeRegistry.set(CreateOrganizationRequest.$type, CreateOrganizationRequest);
function createBaseCreateOrganizationResponse() {
    return { organizationId: "", name: "" };
}
export const CreateOrganizationResponse = {
    $type: "devvit.dev_portal.data_api.admin.organization.CreateOrganizationResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.organizationId !== "") {
            writer.uint32(10).string(message.organizationId);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateOrganizationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.organizationId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            organizationId: isSet(object.organizationId) ? globalThis.String(object.organizationId) : "",
            name: isSet(object.name) ? globalThis.String(object.name) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.organizationId !== "") {
            obj.organizationId = message.organizationId;
        }
        if (message.name !== "") {
            obj.name = message.name;
        }
        return obj;
    },
    create(base) {
        return CreateOrganizationResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCreateOrganizationResponse();
        message.organizationId = object.organizationId ?? "";
        message.name = object.name ?? "";
        return message;
    },
};
messageTypeRegistry.set(CreateOrganizationResponse.$type, CreateOrganizationResponse);
function createBaseGetOrganizationByNameRequest() {
    return { name: "" };
}
export const GetOrganizationByNameRequest = {
    $type: "devvit.dev_portal.data_api.admin.organization.GetOrganizationByNameRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetOrganizationByNameRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { name: isSet(object.name) ? globalThis.String(object.name) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        return obj;
    },
    create(base) {
        return GetOrganizationByNameRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetOrganizationByNameRequest();
        message.name = object.name ?? "";
        return message;
    },
};
messageTypeRegistry.set(GetOrganizationByNameRequest.$type, GetOrganizationByNameRequest);
function createBaseGetOrganizationByNameResponse() {
    return { organizationId: "", isInternal: false };
}
export const GetOrganizationByNameResponse = {
    $type: "devvit.dev_portal.data_api.admin.organization.GetOrganizationByNameResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.organizationId !== "") {
            writer.uint32(10).string(message.organizationId);
        }
        if (message.isInternal !== false) {
            writer.uint32(16).bool(message.isInternal);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetOrganizationByNameResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.organizationId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.isInternal = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            organizationId: isSet(object.organizationId) ? globalThis.String(object.organizationId) : "",
            isInternal: isSet(object.isInternal) ? globalThis.Boolean(object.isInternal) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.organizationId !== "") {
            obj.organizationId = message.organizationId;
        }
        if (message.isInternal !== false) {
            obj.isInternal = message.isInternal;
        }
        return obj;
    },
    create(base) {
        return GetOrganizationByNameResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetOrganizationByNameResponse();
        message.organizationId = object.organizationId ?? "";
        message.isInternal = object.isInternal ?? false;
        return message;
    },
};
messageTypeRegistry.set(GetOrganizationByNameResponse.$type, GetOrganizationByNameResponse);
function createBaseGetOrganizationsRequest() {
    return {};
}
export const GetOrganizationsRequest = {
    $type: "devvit.dev_portal.data_api.admin.organization.GetOrganizationsRequest",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetOrganizationsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return GetOrganizationsRequest.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseGetOrganizationsRequest();
        return message;
    },
};
messageTypeRegistry.set(GetOrganizationsRequest.$type, GetOrganizationsRequest);
function createBaseGetOrganizationsResponse() {
    return { organizations: [] };
}
export const GetOrganizationsResponse = {
    $type: "devvit.dev_portal.data_api.admin.organization.GetOrganizationsResponse",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.organizations) {
            Organization.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetOrganizationsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.organizations.push(Organization.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            organizations: globalThis.Array.isArray(object?.organizations)
                ? object.organizations.map((e) => Organization.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.organizations?.length) {
            obj.organizations = message.organizations.map((e) => Organization.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return GetOrganizationsResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetOrganizationsResponse();
        message.organizations = object.organizations?.map((e) => Organization.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set(GetOrganizationsResponse.$type, GetOrganizationsResponse);
function isSet(value) {
    return value !== null && value !== undefined;
}
