// @devvit/protos/types/devvit/dev_portal/internal/data_api/v1alpha/admin/admin.twirp-client.js
import { GetDataTokenByIdRequest, GetDataTokenByIdResponse, GetDataTokensByOrganizationNameRequest, GetDataTokensByOrganizationNameResponse, } from '../../../../data_api/admin/auth/auth.js';
import { GetDataSubscriptionByIdRequest, GetDataSubscriptionByIdResponse, } from '../../../../data_api/admin/subscription/subscription.js';
export class InternalDataApiAdminClientJSON {
    constructor(rpc) {
        this.rpc = rpc;
        this.GetDataTokenById.bind(this);
        this.GetDataTokensByOrganizationName.bind(this);
        this.GetDataSubscriptionById.bind(this);
    }
    GetDataTokenById(request) {
        const data = GetDataTokenByIdRequest.toJSON(request);
        const promise = this.rpc.request('devvit.dev_portal.internal.data_api.v1alpha.admin.InternalDataApiAdmin', 'GetDataTokenById', 'application/json', data);
        return promise.then((data) => GetDataTokenByIdResponse.fromJSON(data));
    }
    GetDataTokensByOrganizationName(request) {
        const data = GetDataTokensByOrganizationNameRequest.toJSON(request);
        const promise = this.rpc.request('devvit.dev_portal.internal.data_api.v1alpha.admin.InternalDataApiAdmin', 'GetDataTokensByOrganizationName', 'application/json', data);
        return promise.then((data) => GetDataTokensByOrganizationNameResponse.fromJSON(data));
    }
    GetDataSubscriptionById(request) {
        const data = GetDataSubscriptionByIdRequest.toJSON(request);
        const promise = this.rpc.request('devvit.dev_portal.internal.data_api.v1alpha.admin.InternalDataApiAdmin', 'GetDataSubscriptionById', 'application/json', data);
        return promise.then((data) => GetDataSubscriptionByIdResponse.fromJSON(data));
    }
}
export class InternalDataApiAdminClientProtobuf {
    constructor(rpc) {
        this.rpc = rpc;
        this.GetDataTokenById.bind(this);
        this.GetDataTokensByOrganizationName.bind(this);
        this.GetDataSubscriptionById.bind(this);
    }
    GetDataTokenById(request) {
        const data = GetDataTokenByIdRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.dev_portal.internal.data_api.v1alpha.admin.InternalDataApiAdmin', 'GetDataTokenById', 'application/protobuf', data);
        return promise.then((data) => GetDataTokenByIdResponse.decode(data));
    }
    GetDataTokensByOrganizationName(request) {
        const data = GetDataTokensByOrganizationNameRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.dev_portal.internal.data_api.v1alpha.admin.InternalDataApiAdmin', 'GetDataTokensByOrganizationName', 'application/protobuf', data);
        return promise.then((data) => GetDataTokensByOrganizationNameResponse.decode(data));
    }
    GetDataSubscriptionById(request) {
        const data = GetDataSubscriptionByIdRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.dev_portal.internal.data_api.v1alpha.admin.InternalDataApiAdmin', 'GetDataSubscriptionById', 'application/protobuf', data);
        return promise.then((data) => GetDataSubscriptionByIdResponse.decode(data));
    }
}
