// @devvit/protos/types/devvit/plugin/media/media.js
/**
 * #media.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseMediaUploadRequest() {
    return { url: "", type: "" };
}
export const MediaUploadRequest = {
    $type: "devvit.plugin.media.MediaUploadRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        if (message.type !== "") {
            writer.uint32(18).string(message.type);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMediaUploadRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.url = reader.string();
                    break;
                case 2:
                    message.type = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { url: isSet(object.url) ? String(object.url) : "", type: isSet(object.type) ? String(object.type) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.url !== undefined && (obj.url = message.url);
        message.type !== undefined && (obj.type = message.type);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMediaUploadRequest();
        message.url = object.url ?? "";
        message.type = object.type ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.media.MediaUploadRequest", MediaUploadRequest);
function createBaseMediaUploadResponse() {
    return { mediaId: "", mediaUrl: "" };
}
export const MediaUploadResponse = {
    $type: "devvit.plugin.media.MediaUploadResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.mediaId !== "") {
            writer.uint32(10).string(message.mediaId);
        }
        if (message.mediaUrl !== "") {
            writer.uint32(18).string(message.mediaUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMediaUploadResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.mediaId = reader.string();
                    break;
                case 2:
                    message.mediaUrl = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            mediaId: isSet(object.mediaId) ? String(object.mediaId) : "",
            mediaUrl: isSet(object.mediaUrl) ? String(object.mediaUrl) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.mediaId !== undefined && (obj.mediaId = message.mediaId);
        message.mediaUrl !== undefined && (obj.mediaUrl = message.mediaUrl);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMediaUploadResponse();
        message.mediaId = object.mediaId ?? "";
        message.mediaUrl = object.mediaUrl ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.media.MediaUploadResponse", MediaUploadResponse);
export class MediaServiceClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.media.MediaService";
        this.rpc = rpc;
        this.Upload = this.Upload.bind(this);
    }
    Upload(request) {
        const data = MediaUploadRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Upload", data);
        return promise.then((data) => MediaUploadResponse.decode(new _m0.Reader(data)));
    }
}
export const MediaServiceDefinition = {
    name: "MediaService",
    fullName: "devvit.plugin.media.MediaService",
    methods: {
        upload: {
            name: "Upload",
            requestType: MediaUploadRequest,
            requestStream: false,
            responseType: MediaUploadResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
