// @devvit/protos/types/devvit/dev_portal/data_api/admin/audit/audit.js
/**
 * #audit.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Timestamp } from '../../../../../google/protobuf/timestamp.js';
import { messageTypeRegistry } from '../../../../../typeRegistry.js';
export var AuditLogEventType;
(function (AuditLogEventType) {
    AuditLogEventType[AuditLogEventType["DATA_SUBSCRIPTION_CREATE"] = 0] = "DATA_SUBSCRIPTION_CREATE";
    AuditLogEventType[AuditLogEventType["DATA_SUBSCRIPTION_UPDATE"] = 1] = "DATA_SUBSCRIPTION_UPDATE";
    AuditLogEventType[AuditLogEventType["DATA_SUBSCRIPTION_DELETE"] = 2] = "DATA_SUBSCRIPTION_DELETE";
    AuditLogEventType[AuditLogEventType["DATA_TOKEN_CREATE"] = 3] = "DATA_TOKEN_CREATE";
    AuditLogEventType[AuditLogEventType["DATA_TOKEN_REVOKE"] = 4] = "DATA_TOKEN_REVOKE";
    AuditLogEventType[AuditLogEventType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AuditLogEventType || (AuditLogEventType = {}));
export function auditLogEventTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "DATA_SUBSCRIPTION_CREATE":
            return AuditLogEventType.DATA_SUBSCRIPTION_CREATE;
        case 1:
        case "DATA_SUBSCRIPTION_UPDATE":
            return AuditLogEventType.DATA_SUBSCRIPTION_UPDATE;
        case 2:
        case "DATA_SUBSCRIPTION_DELETE":
            return AuditLogEventType.DATA_SUBSCRIPTION_DELETE;
        case 3:
        case "DATA_TOKEN_CREATE":
            return AuditLogEventType.DATA_TOKEN_CREATE;
        case 4:
        case "DATA_TOKEN_REVOKE":
            return AuditLogEventType.DATA_TOKEN_REVOKE;
        case -1:
        case "UNRECOGNIZED":
        default:
            return AuditLogEventType.UNRECOGNIZED;
    }
}
export function auditLogEventTypeToJSON(object) {
    switch (object) {
        case AuditLogEventType.DATA_SUBSCRIPTION_CREATE:
            return 0;
        case AuditLogEventType.DATA_SUBSCRIPTION_UPDATE:
            return 1;
        case AuditLogEventType.DATA_SUBSCRIPTION_DELETE:
            return 2;
        case AuditLogEventType.DATA_TOKEN_CREATE:
            return 3;
        case AuditLogEventType.DATA_TOKEN_REVOKE:
            return 4;
        case AuditLogEventType.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseAuditLogEvent() {
    return {
        id: "",
        createdAt: undefined,
        userId: "",
        type: 0,
        organizationId: "",
        dataTokenCreatePayload: undefined,
        dataTokenRevokePayload: undefined,
        dataSubscriptionCreatePayload: undefined,
        dataSubscriptionUpdatePayload: undefined,
        dataSubscriptionDeletePayload: undefined,
    };
}
export const AuditLogEvent = {
    $type: "devvit.dev_portal.data_api.admin.audit.AuditLogEvent",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.createdAt !== undefined) {
            Timestamp.encode(toTimestamp(message.createdAt), writer.uint32(18).fork()).ldelim();
        }
        if (message.userId !== "") {
            writer.uint32(26).string(message.userId);
        }
        if (message.type !== 0) {
            writer.uint32(32).int32(message.type);
        }
        if (message.organizationId !== "") {
            writer.uint32(42).string(message.organizationId);
        }
        if (message.dataTokenCreatePayload !== undefined) {
            DataTokenCreateAuditLogEventPayload.encode(message.dataTokenCreatePayload, writer.uint32(50).fork()).ldelim();
        }
        if (message.dataTokenRevokePayload !== undefined) {
            DataTokenRevokeAuditLogEventPayload.encode(message.dataTokenRevokePayload, writer.uint32(58).fork()).ldelim();
        }
        if (message.dataSubscriptionCreatePayload !== undefined) {
            DataSubscriptionCreateAuditLogEventPayload.encode(message.dataSubscriptionCreatePayload, writer.uint32(66).fork())
                .ldelim();
        }
        if (message.dataSubscriptionUpdatePayload !== undefined) {
            DataSubscriptionUpdateAuditLogEventPayload.encode(message.dataSubscriptionUpdatePayload, writer.uint32(74).fork())
                .ldelim();
        }
        if (message.dataSubscriptionDeletePayload !== undefined) {
            DataSubscriptionDeleteAuditLogEventPayload.encode(message.dataSubscriptionDeletePayload, writer.uint32(82).fork())
                .ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAuditLogEvent();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.string();
                    break;
                case 2:
                    message.createdAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.userId = reader.string();
                    break;
                case 4:
                    message.type = reader.int32();
                    break;
                case 5:
                    message.organizationId = reader.string();
                    break;
                case 6:
                    message.dataTokenCreatePayload = DataTokenCreateAuditLogEventPayload.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.dataTokenRevokePayload = DataTokenRevokeAuditLogEventPayload.decode(reader, reader.uint32());
                    break;
                case 8:
                    message.dataSubscriptionCreatePayload = DataSubscriptionCreateAuditLogEventPayload.decode(reader, reader.uint32());
                    break;
                case 9:
                    message.dataSubscriptionUpdatePayload = DataSubscriptionUpdateAuditLogEventPayload.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.dataSubscriptionDeletePayload = DataSubscriptionDeleteAuditLogEventPayload.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? String(object.id) : "",
            createdAt: isSet(object.createdAt) ? fromJsonTimestamp(object.createdAt) : undefined,
            userId: isSet(object.userId) ? String(object.userId) : "",
            type: isSet(object.type) ? auditLogEventTypeFromJSON(object.type) : 0,
            organizationId: isSet(object.organizationId) ? String(object.organizationId) : "",
            dataTokenCreatePayload: isSet(object.dataTokenCreatePayload)
                ? DataTokenCreateAuditLogEventPayload.fromJSON(object.dataTokenCreatePayload)
                : undefined,
            dataTokenRevokePayload: isSet(object.dataTokenRevokePayload)
                ? DataTokenRevokeAuditLogEventPayload.fromJSON(object.dataTokenRevokePayload)
                : undefined,
            dataSubscriptionCreatePayload: isSet(object.dataSubscriptionCreatePayload)
                ? DataSubscriptionCreateAuditLogEventPayload.fromJSON(object.dataSubscriptionCreatePayload)
                : undefined,
            dataSubscriptionUpdatePayload: isSet(object.dataSubscriptionUpdatePayload)
                ? DataSubscriptionUpdateAuditLogEventPayload.fromJSON(object.dataSubscriptionUpdatePayload)
                : undefined,
            dataSubscriptionDeletePayload: isSet(object.dataSubscriptionDeletePayload)
                ? DataSubscriptionDeleteAuditLogEventPayload.fromJSON(object.dataSubscriptionDeletePayload)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        message.createdAt !== undefined && (obj.createdAt = message.createdAt.toISOString());
        message.userId !== undefined && (obj.userId = message.userId);
        message.type !== undefined && (obj.type = auditLogEventTypeToJSON(message.type));
        message.organizationId !== undefined && (obj.organizationId = message.organizationId);
        message.dataTokenCreatePayload !== undefined && (obj.dataTokenCreatePayload = message.dataTokenCreatePayload
            ? DataTokenCreateAuditLogEventPayload.toJSON(message.dataTokenCreatePayload)
            : undefined);
        message.dataTokenRevokePayload !== undefined && (obj.dataTokenRevokePayload = message.dataTokenRevokePayload
            ? DataTokenRevokeAuditLogEventPayload.toJSON(message.dataTokenRevokePayload)
            : undefined);
        message.dataSubscriptionCreatePayload !== undefined &&
            (obj.dataSubscriptionCreatePayload = message.dataSubscriptionCreatePayload
                ? DataSubscriptionCreateAuditLogEventPayload.toJSON(message.dataSubscriptionCreatePayload)
                : undefined);
        message.dataSubscriptionUpdatePayload !== undefined &&
            (obj.dataSubscriptionUpdatePayload = message.dataSubscriptionUpdatePayload
                ? DataSubscriptionUpdateAuditLogEventPayload.toJSON(message.dataSubscriptionUpdatePayload)
                : undefined);
        message.dataSubscriptionDeletePayload !== undefined &&
            (obj.dataSubscriptionDeletePayload = message.dataSubscriptionDeletePayload
                ? DataSubscriptionDeleteAuditLogEventPayload.toJSON(message.dataSubscriptionDeletePayload)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAuditLogEvent();
        message.id = object.id ?? "";
        message.createdAt = object.createdAt ?? undefined;
        message.userId = object.userId ?? "";
        message.type = object.type ?? 0;
        message.organizationId = object.organizationId ?? "";
        message.dataTokenCreatePayload =
            (object.dataTokenCreatePayload !== undefined && object.dataTokenCreatePayload !== null)
                ? DataTokenCreateAuditLogEventPayload.fromPartial(object.dataTokenCreatePayload)
                : undefined;
        message.dataTokenRevokePayload =
            (object.dataTokenRevokePayload !== undefined && object.dataTokenRevokePayload !== null)
                ? DataTokenRevokeAuditLogEventPayload.fromPartial(object.dataTokenRevokePayload)
                : undefined;
        message.dataSubscriptionCreatePayload =
            (object.dataSubscriptionCreatePayload !== undefined && object.dataSubscriptionCreatePayload !== null)
                ? DataSubscriptionCreateAuditLogEventPayload.fromPartial(object.dataSubscriptionCreatePayload)
                : undefined;
        message.dataSubscriptionUpdatePayload =
            (object.dataSubscriptionUpdatePayload !== undefined && object.dataSubscriptionUpdatePayload !== null)
                ? DataSubscriptionUpdateAuditLogEventPayload.fromPartial(object.dataSubscriptionUpdatePayload)
                : undefined;
        message.dataSubscriptionDeletePayload =
            (object.dataSubscriptionDeletePayload !== undefined && object.dataSubscriptionDeletePayload !== null)
                ? DataSubscriptionDeleteAuditLogEventPayload.fromPartial(object.dataSubscriptionDeletePayload)
                : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.AuditLogEvent", AuditLogEvent);
function createBaseDataTokenCreateAuditLogEventPayload() {
    return { tokenId: "", expiration: undefined };
}
export const DataTokenCreateAuditLogEventPayload = {
    $type: "devvit.dev_portal.data_api.admin.audit.DataTokenCreateAuditLogEventPayload",
    encode(message, writer = _m0.Writer.create()) {
        if (message.tokenId !== "") {
            writer.uint32(10).string(message.tokenId);
        }
        if (message.expiration !== undefined) {
            Timestamp.encode(toTimestamp(message.expiration), writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDataTokenCreateAuditLogEventPayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.tokenId = reader.string();
                    break;
                case 2:
                    message.expiration = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            tokenId: isSet(object.tokenId) ? String(object.tokenId) : "",
            expiration: isSet(object.expiration) ? fromJsonTimestamp(object.expiration) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.tokenId !== undefined && (obj.tokenId = message.tokenId);
        message.expiration !== undefined && (obj.expiration = message.expiration.toISOString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseDataTokenCreateAuditLogEventPayload();
        message.tokenId = object.tokenId ?? "";
        message.expiration = object.expiration ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.DataTokenCreateAuditLogEventPayload", DataTokenCreateAuditLogEventPayload);
function createBaseDataTokenRevokeAuditLogEventPayload() {
    return { tokenId: "" };
}
export const DataTokenRevokeAuditLogEventPayload = {
    $type: "devvit.dev_portal.data_api.admin.audit.DataTokenRevokeAuditLogEventPayload",
    encode(message, writer = _m0.Writer.create()) {
        if (message.tokenId !== "") {
            writer.uint32(10).string(message.tokenId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDataTokenRevokeAuditLogEventPayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.tokenId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { tokenId: isSet(object.tokenId) ? String(object.tokenId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.tokenId !== undefined && (obj.tokenId = message.tokenId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseDataTokenRevokeAuditLogEventPayload();
        message.tokenId = object.tokenId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.DataTokenRevokeAuditLogEventPayload", DataTokenRevokeAuditLogEventPayload);
function createBaseDataSubscriptionCreateAuditLogEventPayload() {
    return { subscriptionId: "" };
}
export const DataSubscriptionCreateAuditLogEventPayload = {
    $type: "devvit.dev_portal.data_api.admin.audit.DataSubscriptionCreateAuditLogEventPayload",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subscriptionId !== "") {
            writer.uint32(10).string(message.subscriptionId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDataSubscriptionCreateAuditLogEventPayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subscriptionId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { subscriptionId: isSet(object.subscriptionId) ? String(object.subscriptionId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.subscriptionId !== undefined && (obj.subscriptionId = message.subscriptionId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseDataSubscriptionCreateAuditLogEventPayload();
        message.subscriptionId = object.subscriptionId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.DataSubscriptionCreateAuditLogEventPayload", DataSubscriptionCreateAuditLogEventPayload);
function createBaseDataSubscriptionUpdateAuditLogEventPayload() {
    return { subscriptionId: "" };
}
export const DataSubscriptionUpdateAuditLogEventPayload = {
    $type: "devvit.dev_portal.data_api.admin.audit.DataSubscriptionUpdateAuditLogEventPayload",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subscriptionId !== "") {
            writer.uint32(10).string(message.subscriptionId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDataSubscriptionUpdateAuditLogEventPayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subscriptionId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { subscriptionId: isSet(object.subscriptionId) ? String(object.subscriptionId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.subscriptionId !== undefined && (obj.subscriptionId = message.subscriptionId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseDataSubscriptionUpdateAuditLogEventPayload();
        message.subscriptionId = object.subscriptionId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.DataSubscriptionUpdateAuditLogEventPayload", DataSubscriptionUpdateAuditLogEventPayload);
function createBaseDataSubscriptionDeleteAuditLogEventPayload() {
    return { subscriptionId: "" };
}
export const DataSubscriptionDeleteAuditLogEventPayload = {
    $type: "devvit.dev_portal.data_api.admin.audit.DataSubscriptionDeleteAuditLogEventPayload",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subscriptionId !== "") {
            writer.uint32(10).string(message.subscriptionId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDataSubscriptionDeleteAuditLogEventPayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subscriptionId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { subscriptionId: isSet(object.subscriptionId) ? String(object.subscriptionId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.subscriptionId !== undefined && (obj.subscriptionId = message.subscriptionId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseDataSubscriptionDeleteAuditLogEventPayload();
        message.subscriptionId = object.subscriptionId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.DataSubscriptionDeleteAuditLogEventPayload", DataSubscriptionDeleteAuditLogEventPayload);
function createBaseGetAuditLogEventsRequest() {
    return { organizationName: "" };
}
export const GetAuditLogEventsRequest = {
    $type: "devvit.dev_portal.data_api.admin.audit.GetAuditLogEventsRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.organizationName !== "") {
            writer.uint32(10).string(message.organizationName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAuditLogEventsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.organizationName = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { organizationName: isSet(object.organizationName) ? String(object.organizationName) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.organizationName !== undefined && (obj.organizationName = message.organizationName);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetAuditLogEventsRequest();
        message.organizationName = object.organizationName ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.GetAuditLogEventsRequest", GetAuditLogEventsRequest);
function createBaseGetAuditLogEventsResponse() {
    return { events: [] };
}
export const GetAuditLogEventsResponse = {
    $type: "devvit.dev_portal.data_api.admin.audit.GetAuditLogEventsResponse",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.events) {
            AuditLogEvent.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAuditLogEventsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.events.push(AuditLogEvent.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { events: Array.isArray(object?.events) ? object.events.map((e) => AuditLogEvent.fromJSON(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.events) {
            obj.events = message.events.map((e) => e ? AuditLogEvent.toJSON(e) : undefined);
        }
        else {
            obj.events = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetAuditLogEventsResponse();
        message.events = object.events?.map((e) => AuditLogEvent.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.data_api.admin.audit.GetAuditLogEventsResponse", GetAuditLogEventsResponse);
function toTimestamp(date) {
    const seconds = date.getTime() / 1000;
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = t.seconds * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
