// @devvit/protos/types/devvit/plugin/buildpack/filesystem.js
/**
 * #filesystem.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseFile() {
    return { path: "", content: "", isDirectory: false, isBase64: false };
}
export const File = {
    $type: "devvit.plugin.buildpack.File",
    encode(message, writer = _m0.Writer.create()) {
        if (message.path !== "") {
            writer.uint32(10).string(message.path);
        }
        if (message.content !== "") {
            writer.uint32(18).string(message.content);
        }
        if (message.isDirectory === true) {
            writer.uint32(24).bool(message.isDirectory);
        }
        if (message.isBase64 === true) {
            writer.uint32(32).bool(message.isBase64);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFile();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.path = reader.string();
                    break;
                case 2:
                    message.content = reader.string();
                    break;
                case 3:
                    message.isDirectory = reader.bool();
                    break;
                case 4:
                    message.isBase64 = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            path: isSet(object.path) ? String(object.path) : "",
            content: isSet(object.content) ? String(object.content) : "",
            isDirectory: isSet(object.isDirectory) ? Boolean(object.isDirectory) : false,
            isBase64: isSet(object.isBase64) ? Boolean(object.isBase64) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.path !== undefined && (obj.path = message.path);
        message.content !== undefined && (obj.content = message.content);
        message.isDirectory !== undefined && (obj.isDirectory = message.isDirectory);
        message.isBase64 !== undefined && (obj.isBase64 = message.isBase64);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseFile();
        message.path = object.path ?? "";
        message.content = object.content ?? "";
        message.isDirectory = object.isDirectory ?? false;
        message.isBase64 = object.isBase64 ?? false;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.buildpack.File", File);
function createBaseFileSystem() {
    return { files: [] };
}
export const FileSystem = {
    $type: "devvit.plugin.buildpack.FileSystem",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.files) {
            File.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFileSystem();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.files.push(File.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { files: Array.isArray(object?.files) ? object.files.map((e) => File.fromJSON(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.files) {
            obj.files = message.files.map((e) => e ? File.toJSON(e) : undefined);
        }
        else {
            obj.files = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseFileSystem();
        message.files = object.files?.map((e) => File.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.buildpack.FileSystem", FileSystem);
function isSet(value) {
    return value !== null && value !== undefined;
}
