// @devvit/protos/types/devvit/plugin/automodregex/automodregex.js
/**
 * #automodregex.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseAutomodRegexInput() {
    return { pattern: "", input: "", caseInsensitive: false };
}
export const AutomodRegexInput = {
    $type: "devvit.plugin.automodregex.AutomodRegexInput",
    encode(message, writer = _m0.Writer.create()) {
        if (message.pattern !== "") {
            writer.uint32(10).string(message.pattern);
        }
        if (message.input !== "") {
            writer.uint32(18).string(message.input);
        }
        if (message.caseInsensitive === true) {
            writer.uint32(24).bool(message.caseInsensitive);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomodRegexInput();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pattern = reader.string();
                    break;
                case 2:
                    message.input = reader.string();
                    break;
                case 3:
                    message.caseInsensitive = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            pattern: isSet(object.pattern) ? String(object.pattern) : "",
            input: isSet(object.input) ? String(object.input) : "",
            caseInsensitive: isSet(object.caseInsensitive) ? Boolean(object.caseInsensitive) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.pattern !== undefined && (obj.pattern = message.pattern);
        message.input !== undefined && (obj.input = message.input);
        message.caseInsensitive !== undefined && (obj.caseInsensitive = message.caseInsensitive);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAutomodRegexInput();
        message.pattern = object.pattern ?? "";
        message.input = object.input ?? "";
        message.caseInsensitive = object.caseInsensitive ?? false;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.automodregex.AutomodRegexInput", AutomodRegexInput);
function createBaseAutomodRegexMatchOutput() {
    return { groups: [] };
}
export const AutomodRegexMatchOutput = {
    $type: "devvit.plugin.automodregex.AutomodRegexMatchOutput",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.groups) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomodRegexMatchOutput();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.groups.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { groups: Array.isArray(object?.groups) ? object.groups.map((e) => String(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.groups) {
            obj.groups = message.groups.map((e) => e);
        }
        else {
            obj.groups = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAutomodRegexMatchOutput();
        message.groups = object.groups?.map((e) => e) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.automodregex.AutomodRegexMatchOutput", AutomodRegexMatchOutput);
function createBaseAutomodRegexTestOutput() {
    return { match: false };
}
export const AutomodRegexTestOutput = {
    $type: "devvit.plugin.automodregex.AutomodRegexTestOutput",
    encode(message, writer = _m0.Writer.create()) {
        if (message.match === true) {
            writer.uint32(8).bool(message.match);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomodRegexTestOutput();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.match = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { match: isSet(object.match) ? Boolean(object.match) : false };
    },
    toJSON(message) {
        const obj = {};
        message.match !== undefined && (obj.match = message.match);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAutomodRegexTestOutput();
        message.match = object.match ?? false;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.automodregex.AutomodRegexTestOutput", AutomodRegexTestOutput);
export class AutomodRegexClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.automodregex.AutomodRegex";
        this.rpc = rpc;
        this.AutomodRegexMatch = this.AutomodRegexMatch.bind(this);
        this.AutomodRegexTest = this.AutomodRegexTest.bind(this);
    }
    AutomodRegexMatch(request) {
        const data = AutomodRegexInput.encode(request).finish();
        const promise = this.rpc.request(this.service, "AutomodRegexMatch", data);
        return promise.then((data) => AutomodRegexMatchOutput.decode(new _m0.Reader(data)));
    }
    AutomodRegexTest(request) {
        const data = AutomodRegexInput.encode(request).finish();
        const promise = this.rpc.request(this.service, "AutomodRegexTest", data);
        return promise.then((data) => AutomodRegexTestOutput.decode(new _m0.Reader(data)));
    }
}
export const AutomodRegexDefinition = {
    name: "AutomodRegex",
    fullName: "devvit.plugin.automodregex.AutomodRegex",
    methods: {
        automodRegexMatch: {
            name: "AutomodRegexMatch",
            requestType: AutomodRegexInput,
            requestStream: false,
            responseType: AutomodRegexMatchOutput,
            responseStream: false,
            options: {},
        },
        automodRegexTest: {
            name: "AutomodRegexTest",
            requestType: AutomodRegexInput,
            requestStream: false,
            responseType: AutomodRegexTestOutput,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
