// @devvit/protos/types/devvit/plugin/gl/methods.js
/**
 * #methods.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
export var GLMethod;
(function (GLMethod) {
    GLMethod[GLMethod["activeTexture"] = 0] = "activeTexture";
    GLMethod[GLMethod["attachShader"] = 1] = "attachShader";
    GLMethod[GLMethod["bindAttribLocation"] = 2] = "bindAttribLocation";
    GLMethod[GLMethod["bindBuffer"] = 3] = "bindBuffer";
    GLMethod[GLMethod["bindFramebuffer"] = 4] = "bindFramebuffer";
    GLMethod[GLMethod["bindRenderbuffer"] = 5] = "bindRenderbuffer";
    GLMethod[GLMethod["bindTexture"] = 6] = "bindTexture";
    GLMethod[GLMethod["blendColor"] = 7] = "blendColor";
    GLMethod[GLMethod["blendEquation"] = 8] = "blendEquation";
    GLMethod[GLMethod["blendEquationSeparate"] = 9] = "blendEquationSeparate";
    GLMethod[GLMethod["blendFunc"] = 10] = "blendFunc";
    GLMethod[GLMethod["blendFuncSeparate"] = 11] = "blendFuncSeparate";
    GLMethod[GLMethod["bufferData"] = 12] = "bufferData";
    GLMethod[GLMethod["bufferSubData"] = 13] = "bufferSubData";
    GLMethod[GLMethod["checkFramebufferStatus"] = 14] = "checkFramebufferStatus";
    GLMethod[GLMethod["clear"] = 15] = "clear";
    GLMethod[GLMethod["clearColor"] = 16] = "clearColor";
    GLMethod[GLMethod["clearDepth"] = 17] = "clearDepth";
    GLMethod[GLMethod["clearStencil"] = 18] = "clearStencil";
    GLMethod[GLMethod["colorMask"] = 19] = "colorMask";
    GLMethod[GLMethod["compileShader"] = 20] = "compileShader";
    GLMethod[GLMethod["copyTexImage2D"] = 21] = "copyTexImage2D";
    GLMethod[GLMethod["copyTexSubImage2D"] = 22] = "copyTexSubImage2D";
    GLMethod[GLMethod["createBuffer"] = 23] = "createBuffer";
    GLMethod[GLMethod["createFramebuffer"] = 24] = "createFramebuffer";
    GLMethod[GLMethod["createProgram"] = 25] = "createProgram";
    GLMethod[GLMethod["createRenderbuffer"] = 26] = "createRenderbuffer";
    GLMethod[GLMethod["createShader"] = 27] = "createShader";
    GLMethod[GLMethod["createTexture"] = 28] = "createTexture";
    GLMethod[GLMethod["cullFace"] = 29] = "cullFace";
    GLMethod[GLMethod["deleteBuffer"] = 30] = "deleteBuffer";
    GLMethod[GLMethod["deleteFramebuffer"] = 31] = "deleteFramebuffer";
    GLMethod[GLMethod["deleteProgram"] = 32] = "deleteProgram";
    GLMethod[GLMethod["deleteRenderbuffer"] = 33] = "deleteRenderbuffer";
    GLMethod[GLMethod["deleteShader"] = 34] = "deleteShader";
    GLMethod[GLMethod["deleteTexture"] = 35] = "deleteTexture";
    GLMethod[GLMethod["depthFunc"] = 36] = "depthFunc";
    GLMethod[GLMethod["depthMask"] = 37] = "depthMask";
    GLMethod[GLMethod["depthRange"] = 38] = "depthRange";
    GLMethod[GLMethod["detachShader"] = 39] = "detachShader";
    GLMethod[GLMethod["disable"] = 40] = "disable";
    GLMethod[GLMethod["disableVertexAttribArray"] = 41] = "disableVertexAttribArray";
    GLMethod[GLMethod["drawArrays"] = 42] = "drawArrays";
    GLMethod[GLMethod["drawElements"] = 43] = "drawElements";
    GLMethod[GLMethod["enable"] = 44] = "enable";
    GLMethod[GLMethod["enableVertexAttribArray"] = 45] = "enableVertexAttribArray";
    GLMethod[GLMethod["finish"] = 46] = "finish";
    GLMethod[GLMethod["flush"] = 47] = "flush";
    GLMethod[GLMethod["framebufferRenderbuffer"] = 48] = "framebufferRenderbuffer";
    GLMethod[GLMethod["framebufferTexture2D"] = 49] = "framebufferTexture2D";
    GLMethod[GLMethod["frontFace"] = 50] = "frontFace";
    GLMethod[GLMethod["generateMipmap"] = 51] = "generateMipmap";
    GLMethod[GLMethod["getActiveAttrib"] = 52] = "getActiveAttrib";
    GLMethod[GLMethod["getActiveUniform"] = 53] = "getActiveUniform";
    GLMethod[GLMethod["getAttachedShaders"] = 54] = "getAttachedShaders";
    GLMethod[GLMethod["getAttribLocation"] = 55] = "getAttribLocation";
    GLMethod[GLMethod["getBufferParameter"] = 56] = "getBufferParameter";
    GLMethod[GLMethod["getContextAttributes"] = 57] = "getContextAttributes";
    GLMethod[GLMethod["getError"] = 58] = "getError";
    GLMethod[GLMethod["getExtension"] = 59] = "getExtension";
    GLMethod[GLMethod["getFramebufferAttachmentParameter"] = 60] = "getFramebufferAttachmentParameter";
    GLMethod[GLMethod["getParameter"] = 61] = "getParameter";
    GLMethod[GLMethod["getProgramInfoLog"] = 62] = "getProgramInfoLog";
    GLMethod[GLMethod["getProgramParameter"] = 63] = "getProgramParameter";
    GLMethod[GLMethod["getRenderbufferParameter"] = 64] = "getRenderbufferParameter";
    GLMethod[GLMethod["getShaderInfoLog"] = 65] = "getShaderInfoLog";
    GLMethod[GLMethod["getShaderParameter"] = 66] = "getShaderParameter";
    GLMethod[GLMethod["getShaderPrecisionFormat"] = 67] = "getShaderPrecisionFormat";
    GLMethod[GLMethod["getShaderSource"] = 68] = "getShaderSource";
    GLMethod[GLMethod["getSupportedExtensions"] = 69] = "getSupportedExtensions";
    GLMethod[GLMethod["getTexParameter"] = 70] = "getTexParameter";
    GLMethod[GLMethod["getUniform"] = 71] = "getUniform";
    GLMethod[GLMethod["getUniformLocation"] = 72] = "getUniformLocation";
    GLMethod[GLMethod["getVertexAttrib"] = 73] = "getVertexAttrib";
    GLMethod[GLMethod["getVertexAttribOffset"] = 74] = "getVertexAttribOffset";
    GLMethod[GLMethod["hint"] = 75] = "hint";
    GLMethod[GLMethod["isBuffer"] = 76] = "isBuffer";
    GLMethod[GLMethod["isContextLost"] = 77] = "isContextLost";
    GLMethod[GLMethod["isEnabled"] = 78] = "isEnabled";
    GLMethod[GLMethod["isFramebuffer"] = 79] = "isFramebuffer";
    GLMethod[GLMethod["isProgram"] = 80] = "isProgram";
    GLMethod[GLMethod["isRenderbuffer"] = 81] = "isRenderbuffer";
    GLMethod[GLMethod["isShader"] = 82] = "isShader";
    GLMethod[GLMethod["isTexture"] = 83] = "isTexture";
    GLMethod[GLMethod["lineWidth"] = 84] = "lineWidth";
    GLMethod[GLMethod["linkProgram"] = 85] = "linkProgram";
    GLMethod[GLMethod["polygonOffset"] = 86] = "polygonOffset";
    GLMethod[GLMethod["renderbufferStorage"] = 87] = "renderbufferStorage";
    GLMethod[GLMethod["sampleCoverage"] = 88] = "sampleCoverage";
    GLMethod[GLMethod["scissor"] = 89] = "scissor";
    GLMethod[GLMethod["shaderSource"] = 90] = "shaderSource";
    GLMethod[GLMethod["stencilFunc"] = 91] = "stencilFunc";
    GLMethod[GLMethod["stencilFuncSeparate"] = 92] = "stencilFuncSeparate";
    GLMethod[GLMethod["stencilMask"] = 93] = "stencilMask";
    GLMethod[GLMethod["stencilMaskSeparate"] = 94] = "stencilMaskSeparate";
    GLMethod[GLMethod["stencilOp"] = 95] = "stencilOp";
    GLMethod[GLMethod["stencilOpSeparate"] = 96] = "stencilOpSeparate";
    GLMethod[GLMethod["texParameterf"] = 97] = "texParameterf";
    GLMethod[GLMethod["texParameteri"] = 98] = "texParameteri";
    GLMethod[GLMethod["uniform1f"] = 99] = "uniform1f";
    GLMethod[GLMethod["uniform1i"] = 100] = "uniform1i";
    GLMethod[GLMethod["uniform2f"] = 101] = "uniform2f";
    GLMethod[GLMethod["uniform2i"] = 102] = "uniform2i";
    GLMethod[GLMethod["uniform3f"] = 103] = "uniform3f";
    GLMethod[GLMethod["uniform3i"] = 104] = "uniform3i";
    GLMethod[GLMethod["uniform4f"] = 105] = "uniform4f";
    GLMethod[GLMethod["uniform4i"] = 106] = "uniform4i";
    GLMethod[GLMethod["useProgram"] = 107] = "useProgram";
    GLMethod[GLMethod["validateProgram"] = 108] = "validateProgram";
    GLMethod[GLMethod["vertexAttrib1f"] = 109] = "vertexAttrib1f";
    GLMethod[GLMethod["vertexAttrib2f"] = 110] = "vertexAttrib2f";
    GLMethod[GLMethod["vertexAttrib3f"] = 111] = "vertexAttrib3f";
    GLMethod[GLMethod["vertexAttrib4f"] = 112] = "vertexAttrib4f";
    GLMethod[GLMethod["vertexAttribPointer"] = 113] = "vertexAttribPointer";
    GLMethod[GLMethod["viewport"] = 114] = "viewport";
    GLMethod[GLMethod["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(GLMethod || (GLMethod = {}));
export function gLMethodFromJSON(object) {
    switch (object) {
        case 0:
        case "activeTexture":
            return GLMethod.activeTexture;
        case 1:
        case "attachShader":
            return GLMethod.attachShader;
        case 2:
        case "bindAttribLocation":
            return GLMethod.bindAttribLocation;
        case 3:
        case "bindBuffer":
            return GLMethod.bindBuffer;
        case 4:
        case "bindFramebuffer":
            return GLMethod.bindFramebuffer;
        case 5:
        case "bindRenderbuffer":
            return GLMethod.bindRenderbuffer;
        case 6:
        case "bindTexture":
            return GLMethod.bindTexture;
        case 7:
        case "blendColor":
            return GLMethod.blendColor;
        case 8:
        case "blendEquation":
            return GLMethod.blendEquation;
        case 9:
        case "blendEquationSeparate":
            return GLMethod.blendEquationSeparate;
        case 10:
        case "blendFunc":
            return GLMethod.blendFunc;
        case 11:
        case "blendFuncSeparate":
            return GLMethod.blendFuncSeparate;
        case 12:
        case "bufferData":
            return GLMethod.bufferData;
        case 13:
        case "bufferSubData":
            return GLMethod.bufferSubData;
        case 14:
        case "checkFramebufferStatus":
            return GLMethod.checkFramebufferStatus;
        case 15:
        case "clear":
            return GLMethod.clear;
        case 16:
        case "clearColor":
            return GLMethod.clearColor;
        case 17:
        case "clearDepth":
            return GLMethod.clearDepth;
        case 18:
        case "clearStencil":
            return GLMethod.clearStencil;
        case 19:
        case "colorMask":
            return GLMethod.colorMask;
        case 20:
        case "compileShader":
            return GLMethod.compileShader;
        case 21:
        case "copyTexImage2D":
            return GLMethod.copyTexImage2D;
        case 22:
        case "copyTexSubImage2D":
            return GLMethod.copyTexSubImage2D;
        case 23:
        case "createBuffer":
            return GLMethod.createBuffer;
        case 24:
        case "createFramebuffer":
            return GLMethod.createFramebuffer;
        case 25:
        case "createProgram":
            return GLMethod.createProgram;
        case 26:
        case "createRenderbuffer":
            return GLMethod.createRenderbuffer;
        case 27:
        case "createShader":
            return GLMethod.createShader;
        case 28:
        case "createTexture":
            return GLMethod.createTexture;
        case 29:
        case "cullFace":
            return GLMethod.cullFace;
        case 30:
        case "deleteBuffer":
            return GLMethod.deleteBuffer;
        case 31:
        case "deleteFramebuffer":
            return GLMethod.deleteFramebuffer;
        case 32:
        case "deleteProgram":
            return GLMethod.deleteProgram;
        case 33:
        case "deleteRenderbuffer":
            return GLMethod.deleteRenderbuffer;
        case 34:
        case "deleteShader":
            return GLMethod.deleteShader;
        case 35:
        case "deleteTexture":
            return GLMethod.deleteTexture;
        case 36:
        case "depthFunc":
            return GLMethod.depthFunc;
        case 37:
        case "depthMask":
            return GLMethod.depthMask;
        case 38:
        case "depthRange":
            return GLMethod.depthRange;
        case 39:
        case "detachShader":
            return GLMethod.detachShader;
        case 40:
        case "disable":
            return GLMethod.disable;
        case 41:
        case "disableVertexAttribArray":
            return GLMethod.disableVertexAttribArray;
        case 42:
        case "drawArrays":
            return GLMethod.drawArrays;
        case 43:
        case "drawElements":
            return GLMethod.drawElements;
        case 44:
        case "enable":
            return GLMethod.enable;
        case 45:
        case "enableVertexAttribArray":
            return GLMethod.enableVertexAttribArray;
        case 46:
        case "finish":
            return GLMethod.finish;
        case 47:
        case "flush":
            return GLMethod.flush;
        case 48:
        case "framebufferRenderbuffer":
            return GLMethod.framebufferRenderbuffer;
        case 49:
        case "framebufferTexture2D":
            return GLMethod.framebufferTexture2D;
        case 50:
        case "frontFace":
            return GLMethod.frontFace;
        case 51:
        case "generateMipmap":
            return GLMethod.generateMipmap;
        case 52:
        case "getActiveAttrib":
            return GLMethod.getActiveAttrib;
        case 53:
        case "getActiveUniform":
            return GLMethod.getActiveUniform;
        case 54:
        case "getAttachedShaders":
            return GLMethod.getAttachedShaders;
        case 55:
        case "getAttribLocation":
            return GLMethod.getAttribLocation;
        case 56:
        case "getBufferParameter":
            return GLMethod.getBufferParameter;
        case 57:
        case "getContextAttributes":
            return GLMethod.getContextAttributes;
        case 58:
        case "getError":
            return GLMethod.getError;
        case 59:
        case "getExtension":
            return GLMethod.getExtension;
        case 60:
        case "getFramebufferAttachmentParameter":
            return GLMethod.getFramebufferAttachmentParameter;
        case 61:
        case "getParameter":
            return GLMethod.getParameter;
        case 62:
        case "getProgramInfoLog":
            return GLMethod.getProgramInfoLog;
        case 63:
        case "getProgramParameter":
            return GLMethod.getProgramParameter;
        case 64:
        case "getRenderbufferParameter":
            return GLMethod.getRenderbufferParameter;
        case 65:
        case "getShaderInfoLog":
            return GLMethod.getShaderInfoLog;
        case 66:
        case "getShaderParameter":
            return GLMethod.getShaderParameter;
        case 67:
        case "getShaderPrecisionFormat":
            return GLMethod.getShaderPrecisionFormat;
        case 68:
        case "getShaderSource":
            return GLMethod.getShaderSource;
        case 69:
        case "getSupportedExtensions":
            return GLMethod.getSupportedExtensions;
        case 70:
        case "getTexParameter":
            return GLMethod.getTexParameter;
        case 71:
        case "getUniform":
            return GLMethod.getUniform;
        case 72:
        case "getUniformLocation":
            return GLMethod.getUniformLocation;
        case 73:
        case "getVertexAttrib":
            return GLMethod.getVertexAttrib;
        case 74:
        case "getVertexAttribOffset":
            return GLMethod.getVertexAttribOffset;
        case 75:
        case "hint":
            return GLMethod.hint;
        case 76:
        case "isBuffer":
            return GLMethod.isBuffer;
        case 77:
        case "isContextLost":
            return GLMethod.isContextLost;
        case 78:
        case "isEnabled":
            return GLMethod.isEnabled;
        case 79:
        case "isFramebuffer":
            return GLMethod.isFramebuffer;
        case 80:
        case "isProgram":
            return GLMethod.isProgram;
        case 81:
        case "isRenderbuffer":
            return GLMethod.isRenderbuffer;
        case 82:
        case "isShader":
            return GLMethod.isShader;
        case 83:
        case "isTexture":
            return GLMethod.isTexture;
        case 84:
        case "lineWidth":
            return GLMethod.lineWidth;
        case 85:
        case "linkProgram":
            return GLMethod.linkProgram;
        case 86:
        case "polygonOffset":
            return GLMethod.polygonOffset;
        case 87:
        case "renderbufferStorage":
            return GLMethod.renderbufferStorage;
        case 88:
        case "sampleCoverage":
            return GLMethod.sampleCoverage;
        case 89:
        case "scissor":
            return GLMethod.scissor;
        case 90:
        case "shaderSource":
            return GLMethod.shaderSource;
        case 91:
        case "stencilFunc":
            return GLMethod.stencilFunc;
        case 92:
        case "stencilFuncSeparate":
            return GLMethod.stencilFuncSeparate;
        case 93:
        case "stencilMask":
            return GLMethod.stencilMask;
        case 94:
        case "stencilMaskSeparate":
            return GLMethod.stencilMaskSeparate;
        case 95:
        case "stencilOp":
            return GLMethod.stencilOp;
        case 96:
        case "stencilOpSeparate":
            return GLMethod.stencilOpSeparate;
        case 97:
        case "texParameterf":
            return GLMethod.texParameterf;
        case 98:
        case "texParameteri":
            return GLMethod.texParameteri;
        case 99:
        case "uniform1f":
            return GLMethod.uniform1f;
        case 100:
        case "uniform1i":
            return GLMethod.uniform1i;
        case 101:
        case "uniform2f":
            return GLMethod.uniform2f;
        case 102:
        case "uniform2i":
            return GLMethod.uniform2i;
        case 103:
        case "uniform3f":
            return GLMethod.uniform3f;
        case 104:
        case "uniform3i":
            return GLMethod.uniform3i;
        case 105:
        case "uniform4f":
            return GLMethod.uniform4f;
        case 106:
        case "uniform4i":
            return GLMethod.uniform4i;
        case 107:
        case "useProgram":
            return GLMethod.useProgram;
        case 108:
        case "validateProgram":
            return GLMethod.validateProgram;
        case 109:
        case "vertexAttrib1f":
            return GLMethod.vertexAttrib1f;
        case 110:
        case "vertexAttrib2f":
            return GLMethod.vertexAttrib2f;
        case 111:
        case "vertexAttrib3f":
            return GLMethod.vertexAttrib3f;
        case 112:
        case "vertexAttrib4f":
            return GLMethod.vertexAttrib4f;
        case 113:
        case "vertexAttribPointer":
            return GLMethod.vertexAttribPointer;
        case 114:
        case "viewport":
            return GLMethod.viewport;
        case -1:
        case "UNRECOGNIZED":
        default:
            return GLMethod.UNRECOGNIZED;
    }
}
export function gLMethodToJSON(object) {
    switch (object) {
        case GLMethod.activeTexture:
            return 0;
        case GLMethod.attachShader:
            return 1;
        case GLMethod.bindAttribLocation:
            return 2;
        case GLMethod.bindBuffer:
            return 3;
        case GLMethod.bindFramebuffer:
            return 4;
        case GLMethod.bindRenderbuffer:
            return 5;
        case GLMethod.bindTexture:
            return 6;
        case GLMethod.blendColor:
            return 7;
        case GLMethod.blendEquation:
            return 8;
        case GLMethod.blendEquationSeparate:
            return 9;
        case GLMethod.blendFunc:
            return 10;
        case GLMethod.blendFuncSeparate:
            return 11;
        case GLMethod.bufferData:
            return 12;
        case GLMethod.bufferSubData:
            return 13;
        case GLMethod.checkFramebufferStatus:
            return 14;
        case GLMethod.clear:
            return 15;
        case GLMethod.clearColor:
            return 16;
        case GLMethod.clearDepth:
            return 17;
        case GLMethod.clearStencil:
            return 18;
        case GLMethod.colorMask:
            return 19;
        case GLMethod.compileShader:
            return 20;
        case GLMethod.copyTexImage2D:
            return 21;
        case GLMethod.copyTexSubImage2D:
            return 22;
        case GLMethod.createBuffer:
            return 23;
        case GLMethod.createFramebuffer:
            return 24;
        case GLMethod.createProgram:
            return 25;
        case GLMethod.createRenderbuffer:
            return 26;
        case GLMethod.createShader:
            return 27;
        case GLMethod.createTexture:
            return 28;
        case GLMethod.cullFace:
            return 29;
        case GLMethod.deleteBuffer:
            return 30;
        case GLMethod.deleteFramebuffer:
            return 31;
        case GLMethod.deleteProgram:
            return 32;
        case GLMethod.deleteRenderbuffer:
            return 33;
        case GLMethod.deleteShader:
            return 34;
        case GLMethod.deleteTexture:
            return 35;
        case GLMethod.depthFunc:
            return 36;
        case GLMethod.depthMask:
            return 37;
        case GLMethod.depthRange:
            return 38;
        case GLMethod.detachShader:
            return 39;
        case GLMethod.disable:
            return 40;
        case GLMethod.disableVertexAttribArray:
            return 41;
        case GLMethod.drawArrays:
            return 42;
        case GLMethod.drawElements:
            return 43;
        case GLMethod.enable:
            return 44;
        case GLMethod.enableVertexAttribArray:
            return 45;
        case GLMethod.finish:
            return 46;
        case GLMethod.flush:
            return 47;
        case GLMethod.framebufferRenderbuffer:
            return 48;
        case GLMethod.framebufferTexture2D:
            return 49;
        case GLMethod.frontFace:
            return 50;
        case GLMethod.generateMipmap:
            return 51;
        case GLMethod.getActiveAttrib:
            return 52;
        case GLMethod.getActiveUniform:
            return 53;
        case GLMethod.getAttachedShaders:
            return 54;
        case GLMethod.getAttribLocation:
            return 55;
        case GLMethod.getBufferParameter:
            return 56;
        case GLMethod.getContextAttributes:
            return 57;
        case GLMethod.getError:
            return 58;
        case GLMethod.getExtension:
            return 59;
        case GLMethod.getFramebufferAttachmentParameter:
            return 60;
        case GLMethod.getParameter:
            return 61;
        case GLMethod.getProgramInfoLog:
            return 62;
        case GLMethod.getProgramParameter:
            return 63;
        case GLMethod.getRenderbufferParameter:
            return 64;
        case GLMethod.getShaderInfoLog:
            return 65;
        case GLMethod.getShaderParameter:
            return 66;
        case GLMethod.getShaderPrecisionFormat:
            return 67;
        case GLMethod.getShaderSource:
            return 68;
        case GLMethod.getSupportedExtensions:
            return 69;
        case GLMethod.getTexParameter:
            return 70;
        case GLMethod.getUniform:
            return 71;
        case GLMethod.getUniformLocation:
            return 72;
        case GLMethod.getVertexAttrib:
            return 73;
        case GLMethod.getVertexAttribOffset:
            return 74;
        case GLMethod.hint:
            return 75;
        case GLMethod.isBuffer:
            return 76;
        case GLMethod.isContextLost:
            return 77;
        case GLMethod.isEnabled:
            return 78;
        case GLMethod.isFramebuffer:
            return 79;
        case GLMethod.isProgram:
            return 80;
        case GLMethod.isRenderbuffer:
            return 81;
        case GLMethod.isShader:
            return 82;
        case GLMethod.isTexture:
            return 83;
        case GLMethod.lineWidth:
            return 84;
        case GLMethod.linkProgram:
            return 85;
        case GLMethod.polygonOffset:
            return 86;
        case GLMethod.renderbufferStorage:
            return 87;
        case GLMethod.sampleCoverage:
            return 88;
        case GLMethod.scissor:
            return 89;
        case GLMethod.shaderSource:
            return 90;
        case GLMethod.stencilFunc:
            return 91;
        case GLMethod.stencilFuncSeparate:
            return 92;
        case GLMethod.stencilMask:
            return 93;
        case GLMethod.stencilMaskSeparate:
            return 94;
        case GLMethod.stencilOp:
            return 95;
        case GLMethod.stencilOpSeparate:
            return 96;
        case GLMethod.texParameterf:
            return 97;
        case GLMethod.texParameteri:
            return 98;
        case GLMethod.uniform1f:
            return 99;
        case GLMethod.uniform1i:
            return 100;
        case GLMethod.uniform2f:
            return 101;
        case GLMethod.uniform2i:
            return 102;
        case GLMethod.uniform3f:
            return 103;
        case GLMethod.uniform3i:
            return 104;
        case GLMethod.uniform4f:
            return 105;
        case GLMethod.uniform4i:
            return 106;
        case GLMethod.useProgram:
            return 107;
        case GLMethod.validateProgram:
            return 108;
        case GLMethod.vertexAttrib1f:
            return 109;
        case GLMethod.vertexAttrib2f:
            return 110;
        case GLMethod.vertexAttrib3f:
            return 111;
        case GLMethod.vertexAttrib4f:
            return 112;
        case GLMethod.vertexAttribPointer:
            return 113;
        case GLMethod.viewport:
            return 114;
        case GLMethod.UNRECOGNIZED:
        default:
            return -1;
    }
}
