// @devvit/protos/types/devvit/ui/effects/v1alpha/realtime_subscriptions.js
/**
 * #realtime_subscriptions.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
import { RealtimeEvent } from '../../../events/v1alpha/realtime.js';
export var RealtimeSubscriptionStatus;
(function (RealtimeSubscriptionStatus) {
    RealtimeSubscriptionStatus[RealtimeSubscriptionStatus["REALTIME_SUBSCRIBED"] = 0] = "REALTIME_SUBSCRIBED";
    RealtimeSubscriptionStatus[RealtimeSubscriptionStatus["REALTIME_UNSUBSCRIBED"] = 1] = "REALTIME_UNSUBSCRIBED";
    RealtimeSubscriptionStatus[RealtimeSubscriptionStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(RealtimeSubscriptionStatus || (RealtimeSubscriptionStatus = {}));
export function realtimeSubscriptionStatusFromJSON(object) {
    switch (object) {
        case 0:
        case "REALTIME_SUBSCRIBED":
            return RealtimeSubscriptionStatus.REALTIME_SUBSCRIBED;
        case 1:
        case "REALTIME_UNSUBSCRIBED":
            return RealtimeSubscriptionStatus.REALTIME_UNSUBSCRIBED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return RealtimeSubscriptionStatus.UNRECOGNIZED;
    }
}
export function realtimeSubscriptionStatusToJSON(object) {
    switch (object) {
        case RealtimeSubscriptionStatus.REALTIME_SUBSCRIBED:
            return 0;
        case RealtimeSubscriptionStatus.REALTIME_UNSUBSCRIBED:
            return 1;
        case RealtimeSubscriptionStatus.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseRealtimeSubscriptionsEffect() {
    return { subscriptionIds: [] };
}
export const RealtimeSubscriptionsEffect = {
    $type: "devvit.ui.effects.v1alpha.RealtimeSubscriptionsEffect",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.subscriptionIds) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRealtimeSubscriptionsEffect();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subscriptionIds.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subscriptionIds: Array.isArray(object?.subscriptionIds) ? object.subscriptionIds.map((e) => String(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.subscriptionIds) {
            obj.subscriptionIds = message.subscriptionIds.map((e) => e);
        }
        else {
            obj.subscriptionIds = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRealtimeSubscriptionsEffect();
        message.subscriptionIds = object.subscriptionIds?.map((e) => e) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.RealtimeSubscriptionsEffect", RealtimeSubscriptionsEffect);
function createBaseRealtimeSubscriptionEvent() {
    return { event: undefined, status: undefined };
}
export const RealtimeSubscriptionEvent = {
    $type: "devvit.ui.effects.v1alpha.RealtimeSubscriptionEvent",
    encode(message, writer = _m0.Writer.create()) {
        if (message.event !== undefined) {
            RealtimeEvent.encode(message.event, writer.uint32(10).fork()).ldelim();
        }
        if (message.status !== undefined) {
            writer.uint32(16).int32(message.status);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRealtimeSubscriptionEvent();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.event = RealtimeEvent.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.status = reader.int32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            event: isSet(object.event) ? RealtimeEvent.fromJSON(object.event) : undefined,
            status: isSet(object.status) ? realtimeSubscriptionStatusFromJSON(object.status) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.event !== undefined && (obj.event = message.event ? RealtimeEvent.toJSON(message.event) : undefined);
        message.status !== undefined &&
            (obj.status = message.status !== undefined ? realtimeSubscriptionStatusToJSON(message.status) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRealtimeSubscriptionEvent();
        message.event = (object.event !== undefined && object.event !== null)
            ? RealtimeEvent.fromPartial(object.event)
            : undefined;
        message.status = object.status ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.RealtimeSubscriptionEvent", RealtimeSubscriptionEvent);
function isSet(value) {
    return value !== null && value !== undefined;
}
