// @devvit/protos/types/devvit/dev_portal/dev_portal.js
/**
 * #dev_portal.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../google/protobuf/empty.js';
import { Value } from '../../google/protobuf/struct.js';
import { UUID } from '../uuid.js';
import { AppAccountExistsRequest, AppAccountExistsResponse, AppCreationRequest, AppExistsRequest, AppExistsResponse, AppSearchRequest, AppSearchResponse, AppUpdateRequest, CheckIfMediaExistsRequest, CheckIfMediaExistsResponse, CreateAppAccountRequest, CreateAppAccountResponse, FullAppInfo, GetAllWithOwnerRequest, GetAppBySlugRequest, GetPopularAppsRequest, UnpublishAppRequest, UploadNewMediaRequest, UploadNewMediaResponse, } from './app/app.js';
import { AppInfo, MultipleAppInfos } from './app/info/app_info.js';
import { GetAppSettingsFormRequest, GetAppSettingsFormResponse, GetAppSettingsRequest, GetAppSettingsResponse, UpdateAppSettingsRequest, UpdateAppSettingsResponse, ValidateAppSettingsFormRequest, ValidateAppSettingsFormResponse, } from './app_settings/app_settings.js';
import { AppSlugAndVersion, AppVersionCreationRequest, AppVersionUpdateRequest, FullAppVersionInfo, GetAppVersionBundleResponse, ValidInstallLocationsRequest, ValidInstallLocationsResponse, } from './app_version/app_version.js';
import { AppVersionInfo } from './app_version/info/app_version_info.js';
import { GetAuditLogEventsRequest, GetAuditLogEventsResponse } from './data_api/admin/audit/audit.js';
import { CreateDataTokenRequest, CreateDataTokenResponse, GetDataTokenByIdRequest, GetDataTokenByIdResponse, GetDataTokensByOrganizationNameRequest, GetDataTokensByOrganizationNameResponse, RevokeDataTokenByIdRequest, RevokeDataTokenByIdResponse, } from './data_api/admin/auth/auth.js';
import { CreateOrganizationRequest, CreateOrganizationResponse, GetOrganizationByNameRequest, GetOrganizationByNameResponse, GetOrganizationsRequest, GetOrganizationsResponse, } from './data_api/admin/organization/organization.js';
import { CreateDataSubscriptionRequest, CreateDataSubscriptionResponse, DeleteDataSubscriptionRequest, DeleteDataSubscriptionResponse, GetDataSubscriptionByIdRequest, GetDataSubscriptionByIdResponse, GetDataSubscriptionsByOrganizationNameRequest, GetDataSubscriptionsByOrganizationNameResponse, UpdateDataSubscriptionRequest, UpdateDataSubscriptionResponse, } from './data_api/admin/subscription/subscription.js';
import { CreateFeedbackRequest } from './feedback/feedback.js';
import { FullInstallationInfo, GetAllWithAppResponse, GetAllWithInstallerRequest, GetAllWithInstallLocationRequest, GetByAppNameAndInstallLocationRequest, GetInstallationHistoryRequest, GetInstallationHistoryResponse, InstallationCreationRequest, InstallationUpdateRequest, InstallationUpgradeManyRequest, InstallationUpgradeManyResponse, InstallationUpgradeRequest, MultipleInstallationsResponse, } from './installation/installation.js';
import { GetInstallationSettingsFormRequest, GetInstallationSettingsFormResponse, GetInstallationSettingsRequest, GetInstallationSettingsResponse, UpdateInstallationSettingsRequest, UpdateInstallationSettingsResponse, ValidateInstallationSettingsFormRequest, ValidateInstallationSettingsFormResponse, } from './installation_settings/installation_settings.js';
import { GetNutritionByNameVersionRequest, MultipleNutritionCategories } from './nutrition/nutrition.js';
export class AppClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.App";
        this.rpc = rpc;
        this.Create = this.Create.bind(this);
        this.Search = this.Search.bind(this);
        this.GetBySlug = this.GetBySlug.bind(this);
        this.GetByUUID = this.GetByUUID.bind(this);
        this.GetAllWithLocation = this.GetAllWithLocation.bind(this);
        this.GetAllWithInstaller = this.GetAllWithInstaller.bind(this);
        this.GetAllWithOwner = this.GetAllWithOwner.bind(this);
        this.GetPopularApps = this.GetPopularApps.bind(this);
        this.CheckIfMediaExists = this.CheckIfMediaExists.bind(this);
        this.UploadNewMedia = this.UploadNewMedia.bind(this);
        this.Update = this.Update.bind(this);
        this.Unpublish = this.Unpublish.bind(this);
        this.Exists = this.Exists.bind(this);
        this.AppAccountExists = this.AppAccountExists.bind(this);
        this.CreateAppAccount = this.CreateAppAccount.bind(this);
    }
    Create(request) {
        const data = AppCreationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Create", data);
        return promise.then((data) => AppInfo.decode(new _m0.Reader(data)));
    }
    Search(request) {
        const data = AppSearchRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Search", data);
        return promise.then((data) => AppSearchResponse.decode(new _m0.Reader(data)));
    }
    GetBySlug(request) {
        const data = GetAppBySlugRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetBySlug", data);
        return promise.then((data) => FullAppInfo.decode(new _m0.Reader(data)));
    }
    GetByUUID(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetByUUID", data);
        return promise.then((data) => FullAppInfo.decode(new _m0.Reader(data)));
    }
    GetAllWithLocation(request) {
        const data = GetAllWithInstallLocationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAllWithLocation", data);
        return promise.then((data) => MultipleAppInfos.decode(new _m0.Reader(data)));
    }
    GetAllWithInstaller(request) {
        const data = GetAllWithInstallerRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAllWithInstaller", data);
        return promise.then((data) => MultipleAppInfos.decode(new _m0.Reader(data)));
    }
    GetAllWithOwner(request) {
        const data = GetAllWithOwnerRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAllWithOwner", data);
        return promise.then((data) => MultipleAppInfos.decode(new _m0.Reader(data)));
    }
    GetPopularApps(request) {
        const data = GetPopularAppsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetPopularApps", data);
        return promise.then((data) => MultipleAppInfos.decode(new _m0.Reader(data)));
    }
    CheckIfMediaExists(request) {
        const data = CheckIfMediaExistsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CheckIfMediaExists", data);
        return promise.then((data) => CheckIfMediaExistsResponse.decode(new _m0.Reader(data)));
    }
    UploadNewMedia(request) {
        const data = UploadNewMediaRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UploadNewMedia", data);
        return promise.then((data) => UploadNewMediaResponse.decode(new _m0.Reader(data)));
    }
    Update(request) {
        const data = AppUpdateRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Update", data);
        return promise.then((data) => FullAppInfo.decode(new _m0.Reader(data)));
    }
    Unpublish(request) {
        const data = UnpublishAppRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Unpublish", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    Exists(request) {
        const data = AppExistsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Exists", data);
        return promise.then((data) => AppExistsResponse.decode(new _m0.Reader(data)));
    }
    AppAccountExists(request) {
        const data = AppAccountExistsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "AppAccountExists", data);
        return promise.then((data) => AppAccountExistsResponse.decode(new _m0.Reader(data)));
    }
    CreateAppAccount(request) {
        const data = CreateAppAccountRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CreateAppAccount", data);
        return promise.then((data) => CreateAppAccountResponse.decode(new _m0.Reader(data)));
    }
}
export const AppDefinition = {
    name: "App",
    fullName: "devvit.dev_portal.App",
    methods: {
        create: {
            name: "Create",
            requestType: AppCreationRequest,
            requestStream: false,
            responseType: AppInfo,
            responseStream: false,
            options: {},
        },
        search: {
            name: "Search",
            requestType: AppSearchRequest,
            requestStream: false,
            responseType: AppSearchResponse,
            responseStream: false,
            options: {},
        },
        getBySlug: {
            name: "GetBySlug",
            requestType: GetAppBySlugRequest,
            requestStream: false,
            responseType: FullAppInfo,
            responseStream: false,
            options: {},
        },
        getByUUID: {
            name: "GetByUUID",
            requestType: UUID,
            requestStream: false,
            responseType: FullAppInfo,
            responseStream: false,
            options: {},
        },
        getAllWithLocation: {
            name: "GetAllWithLocation",
            requestType: GetAllWithInstallLocationRequest,
            requestStream: false,
            responseType: MultipleAppInfos,
            responseStream: false,
            options: {},
        },
        getAllWithInstaller: {
            name: "GetAllWithInstaller",
            requestType: GetAllWithInstallerRequest,
            requestStream: false,
            responseType: MultipleAppInfos,
            responseStream: false,
            options: {},
        },
        getAllWithOwner: {
            name: "GetAllWithOwner",
            requestType: GetAllWithOwnerRequest,
            requestStream: false,
            responseType: MultipleAppInfos,
            responseStream: false,
            options: {},
        },
        /** Get the X most popular apps, as determined by number of installations. */
        getPopularApps: {
            name: "GetPopularApps",
            requestType: GetPopularAppsRequest,
            requestStream: false,
            responseType: MultipleAppInfos,
            responseStream: false,
            options: {},
        },
        checkIfMediaExists: {
            name: "CheckIfMediaExists",
            requestType: CheckIfMediaExistsRequest,
            requestStream: false,
            responseType: CheckIfMediaExistsResponse,
            responseStream: false,
            options: {},
        },
        uploadNewMedia: {
            name: "UploadNewMedia",
            requestType: UploadNewMediaRequest,
            requestStream: false,
            responseType: UploadNewMediaResponse,
            responseStream: false,
            options: {},
        },
        update: {
            name: "Update",
            requestType: AppUpdateRequest,
            requestStream: false,
            responseType: FullAppInfo,
            responseStream: false,
            options: {},
        },
        unpublish: {
            name: "Unpublish",
            requestType: UnpublishAppRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        exists: {
            name: "Exists",
            requestType: AppExistsRequest,
            requestStream: false,
            responseType: AppExistsResponse,
            responseStream: false,
            options: {},
        },
        appAccountExists: {
            name: "AppAccountExists",
            requestType: AppAccountExistsRequest,
            requestStream: false,
            responseType: AppAccountExistsResponse,
            responseStream: false,
            options: {},
        },
        createAppAccount: {
            name: "CreateAppAccount",
            requestType: CreateAppAccountRequest,
            requestStream: false,
            responseType: CreateAppAccountResponse,
            responseStream: false,
            options: {},
        },
    },
};
export class AppVersionClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.AppVersion";
        this.rpc = rpc;
        this.Create = this.Create.bind(this);
        this.Get = this.Get.bind(this);
        this.GetBySlugAndVersionNumber = this.GetBySlugAndVersionNumber.bind(this);
        this.Update = this.Update.bind(this);
        this.GetValidInstallLocations = this.GetValidInstallLocations.bind(this);
        this.GetAppVersionBundle = this.GetAppVersionBundle.bind(this);
        this.GetNutritionByNameVersion = this.GetNutritionByNameVersion.bind(this);
        this.GetNutritionByAppVersionId = this.GetNutritionByAppVersionId.bind(this);
    }
    Create(request) {
        const data = AppVersionCreationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Create", data);
        return promise.then((data) => AppVersionInfo.decode(new _m0.Reader(data)));
    }
    Get(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "Get", data);
        return promise.then((data) => FullAppVersionInfo.decode(new _m0.Reader(data)));
    }
    GetBySlugAndVersionNumber(request) {
        const data = AppSlugAndVersion.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetBySlugAndVersionNumber", data);
        return promise.then((data) => FullAppVersionInfo.decode(new _m0.Reader(data)));
    }
    Update(request) {
        const data = AppVersionUpdateRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Update", data);
        return promise.then((data) => FullAppVersionInfo.decode(new _m0.Reader(data)));
    }
    GetValidInstallLocations(request) {
        const data = ValidInstallLocationsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetValidInstallLocations", data);
        return promise.then((data) => ValidInstallLocationsResponse.decode(new _m0.Reader(data)));
    }
    GetAppVersionBundle(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAppVersionBundle", data);
        return promise.then((data) => GetAppVersionBundleResponse.decode(new _m0.Reader(data)));
    }
    GetNutritionByNameVersion(request) {
        const data = GetNutritionByNameVersionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetNutritionByNameVersion", data);
        return promise.then((data) => MultipleNutritionCategories.decode(new _m0.Reader(data)));
    }
    GetNutritionByAppVersionId(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetNutritionByAppVersionId", data);
        return promise.then((data) => MultipleNutritionCategories.decode(new _m0.Reader(data)));
    }
}
export const AppVersionDefinition = {
    name: "AppVersion",
    fullName: "devvit.dev_portal.AppVersion",
    methods: {
        create: {
            name: "Create",
            requestType: AppVersionCreationRequest,
            requestStream: false,
            responseType: AppVersionInfo,
            responseStream: false,
            options: {},
        },
        get: {
            name: "Get",
            requestType: UUID,
            requestStream: false,
            responseType: FullAppVersionInfo,
            responseStream: false,
            options: {},
        },
        getBySlugAndVersionNumber: {
            name: "GetBySlugAndVersionNumber",
            requestType: AppSlugAndVersion,
            requestStream: false,
            responseType: FullAppVersionInfo,
            responseStream: false,
            options: {},
        },
        update: {
            name: "Update",
            requestType: AppVersionUpdateRequest,
            requestStream: false,
            responseType: FullAppVersionInfo,
            responseStream: false,
            options: {},
        },
        getValidInstallLocations: {
            name: "GetValidInstallLocations",
            requestType: ValidInstallLocationsRequest,
            requestStream: false,
            responseType: ValidInstallLocationsResponse,
            responseStream: false,
            options: {},
        },
        getAppVersionBundle: {
            name: "GetAppVersionBundle",
            requestType: UUID,
            requestStream: false,
            responseType: GetAppVersionBundleResponse,
            responseStream: false,
            options: {},
        },
        /**
         * Get app nutritional lables by app name (slug) and optional version (1.2.3,
         * 1.2.3.4, latest). Unlike SemVer, latest includes pre-release versions, not
         * just stable.
         */
        getNutritionByNameVersion: {
            name: "GetNutritionByNameVersion",
            requestType: GetNutritionByNameVersionRequest,
            requestStream: false,
            responseType: MultipleNutritionCategories,
            responseStream: false,
            options: {},
        },
        /** Get app nutritional lables by app version id */
        getNutritionByAppVersionId: {
            name: "GetNutritionByAppVersionId",
            requestType: UUID,
            requestStream: false,
            responseType: MultipleNutritionCategories,
            responseStream: false,
            options: {},
        },
    },
};
export class InstallationsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.Installations";
        this.rpc = rpc;
        this.Create = this.Create.bind(this);
        this.GetByUUID = this.GetByUUID.bind(this);
        this.GetByAppNameAndInstallLocation = this.GetByAppNameAndInstallLocation.bind(this);
        this.GetAllWithInstallLocation = this.GetAllWithInstallLocation.bind(this);
        this.GetAllWithInstaller = this.GetAllWithInstaller.bind(this);
        this.GetAllWithVersionUUID = this.GetAllWithVersionUUID.bind(this);
        this.GetAllWithApp = this.GetAllWithApp.bind(this);
        this.GetInstallationHistory = this.GetInstallationHistory.bind(this);
        this.Upgrade = this.Upgrade.bind(this);
        this.UpgradeMany = this.UpgradeMany.bind(this);
        this.Update = this.Update.bind(this);
        this.Remove = this.Remove.bind(this);
    }
    Create(request) {
        const data = InstallationCreationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Create", data);
        return promise.then((data) => FullInstallationInfo.decode(new _m0.Reader(data)));
    }
    GetByUUID(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetByUUID", data);
        return promise.then((data) => FullInstallationInfo.decode(new _m0.Reader(data)));
    }
    GetByAppNameAndInstallLocation(request) {
        const data = GetByAppNameAndInstallLocationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetByAppNameAndInstallLocation", data);
        return promise.then((data) => FullInstallationInfo.decode(new _m0.Reader(data)));
    }
    GetAllWithInstallLocation(request) {
        const data = GetAllWithInstallLocationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAllWithInstallLocation", data);
        return promise.then((data) => MultipleInstallationsResponse.decode(new _m0.Reader(data)));
    }
    GetAllWithInstaller(request) {
        const data = GetAllWithInstallerRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAllWithInstaller", data);
        return promise.then((data) => MultipleInstallationsResponse.decode(new _m0.Reader(data)));
    }
    GetAllWithVersionUUID(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAllWithVersionUUID", data);
        return promise.then((data) => MultipleInstallationsResponse.decode(new _m0.Reader(data)));
    }
    GetAllWithApp(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAllWithApp", data);
        return promise.then((data) => GetAllWithAppResponse.decode(new _m0.Reader(data)));
    }
    GetInstallationHistory(request) {
        const data = GetInstallationHistoryRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetInstallationHistory", data);
        return promise.then((data) => GetInstallationHistoryResponse.decode(new _m0.Reader(data)));
    }
    Upgrade(request) {
        const data = InstallationUpgradeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Upgrade", data);
        return promise.then((data) => FullInstallationInfo.decode(new _m0.Reader(data)));
    }
    UpgradeMany(request) {
        const data = InstallationUpgradeManyRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpgradeMany", data);
        return promise.then((data) => InstallationUpgradeManyResponse.decode(new _m0.Reader(data)));
    }
    Update(request) {
        const data = InstallationUpdateRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Update", data);
        return promise.then((data) => FullInstallationInfo.decode(new _m0.Reader(data)));
    }
    Remove(request) {
        const data = UUID.encode(request).finish();
        const promise = this.rpc.request(this.service, "Remove", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const InstallationsDefinition = {
    name: "Installations",
    fullName: "devvit.dev_portal.Installations",
    methods: {
        create: {
            name: "Create",
            requestType: InstallationCreationRequest,
            requestStream: false,
            responseType: FullInstallationInfo,
            responseStream: false,
            options: {},
        },
        getByUUID: {
            name: "GetByUUID",
            requestType: UUID,
            requestStream: false,
            responseType: FullInstallationInfo,
            responseStream: false,
            options: {},
        },
        getByAppNameAndInstallLocation: {
            name: "GetByAppNameAndInstallLocation",
            requestType: GetByAppNameAndInstallLocationRequest,
            requestStream: false,
            responseType: FullInstallationInfo,
            responseStream: false,
            options: {},
        },
        getAllWithInstallLocation: {
            name: "GetAllWithInstallLocation",
            requestType: GetAllWithInstallLocationRequest,
            requestStream: false,
            responseType: MultipleInstallationsResponse,
            responseStream: false,
            options: {},
        },
        getAllWithInstaller: {
            name: "GetAllWithInstaller",
            requestType: GetAllWithInstallerRequest,
            requestStream: false,
            responseType: MultipleInstallationsResponse,
            responseStream: false,
            options: {},
        },
        getAllWithVersionUUID: {
            name: "GetAllWithVersionUUID",
            requestType: UUID,
            requestStream: false,
            responseType: MultipleInstallationsResponse,
            responseStream: false,
            options: {},
        },
        getAllWithApp: {
            name: "GetAllWithApp",
            requestType: UUID,
            requestStream: false,
            responseType: GetAllWithAppResponse,
            responseStream: false,
            options: {},
        },
        getInstallationHistory: {
            name: "GetInstallationHistory",
            requestType: GetInstallationHistoryRequest,
            requestStream: false,
            responseType: GetInstallationHistoryResponse,
            responseStream: false,
            options: {},
        },
        upgrade: {
            name: "Upgrade",
            requestType: InstallationUpgradeRequest,
            requestStream: false,
            responseType: FullInstallationInfo,
            responseStream: false,
            options: {},
        },
        upgradeMany: {
            name: "UpgradeMany",
            requestType: InstallationUpgradeManyRequest,
            requestStream: false,
            responseType: InstallationUpgradeManyResponse,
            responseStream: false,
            options: {},
        },
        update: {
            name: "Update",
            requestType: InstallationUpdateRequest,
            requestStream: false,
            responseType: FullInstallationInfo,
            responseStream: false,
            options: {},
        },
        remove: {
            name: "Remove",
            requestType: UUID,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
export class FeedbackClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.Feedback";
        this.rpc = rpc;
        this.Create = this.Create.bind(this);
    }
    Create(request) {
        const data = CreateFeedbackRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Create", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const FeedbackDefinition = {
    name: "Feedback",
    fullName: "devvit.dev_portal.Feedback",
    methods: {
        create: {
            name: "Create",
            requestType: CreateFeedbackRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
export class DataApiAdminClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.DataApiAdmin";
        this.rpc = rpc;
        this.CreateDataToken = this.CreateDataToken.bind(this);
        this.GetDataTokenById = this.GetDataTokenById.bind(this);
        this.GetDataTokensByOrganizationName = this.GetDataTokensByOrganizationName.bind(this);
        this.RevokeDataTokenById = this.RevokeDataTokenById.bind(this);
        this.CreateDataSubscription = this.CreateDataSubscription.bind(this);
        this.GetDataSubscriptionById = this.GetDataSubscriptionById.bind(this);
        this.GetDataSubscriptionsByOrganizationName = this.GetDataSubscriptionsByOrganizationName.bind(this);
        this.UpdateDataSubscription = this.UpdateDataSubscription.bind(this);
        this.DeleteDataSubscription = this.DeleteDataSubscription.bind(this);
        this.CreateOrganization = this.CreateOrganization.bind(this);
        this.GetOrganizationByName = this.GetOrganizationByName.bind(this);
        this.GetOrganizations = this.GetOrganizations.bind(this);
        this.GetAuditLogEvents = this.GetAuditLogEvents.bind(this);
    }
    CreateDataToken(request) {
        const data = CreateDataTokenRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CreateDataToken", data);
        return promise.then((data) => CreateDataTokenResponse.decode(new _m0.Reader(data)));
    }
    GetDataTokenById(request) {
        const data = GetDataTokenByIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetDataTokenById", data);
        return promise.then((data) => GetDataTokenByIdResponse.decode(new _m0.Reader(data)));
    }
    GetDataTokensByOrganizationName(request) {
        const data = GetDataTokensByOrganizationNameRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetDataTokensByOrganizationName", data);
        return promise.then((data) => GetDataTokensByOrganizationNameResponse.decode(new _m0.Reader(data)));
    }
    RevokeDataTokenById(request) {
        const data = RevokeDataTokenByIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RevokeDataTokenById", data);
        return promise.then((data) => RevokeDataTokenByIdResponse.decode(new _m0.Reader(data)));
    }
    CreateDataSubscription(request) {
        const data = CreateDataSubscriptionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CreateDataSubscription", data);
        return promise.then((data) => CreateDataSubscriptionResponse.decode(new _m0.Reader(data)));
    }
    GetDataSubscriptionById(request) {
        const data = GetDataSubscriptionByIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetDataSubscriptionById", data);
        return promise.then((data) => GetDataSubscriptionByIdResponse.decode(new _m0.Reader(data)));
    }
    GetDataSubscriptionsByOrganizationName(request) {
        const data = GetDataSubscriptionsByOrganizationNameRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetDataSubscriptionsByOrganizationName", data);
        return promise.then((data) => GetDataSubscriptionsByOrganizationNameResponse.decode(new _m0.Reader(data)));
    }
    UpdateDataSubscription(request) {
        const data = UpdateDataSubscriptionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateDataSubscription", data);
        return promise.then((data) => UpdateDataSubscriptionResponse.decode(new _m0.Reader(data)));
    }
    DeleteDataSubscription(request) {
        const data = DeleteDataSubscriptionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteDataSubscription", data);
        return promise.then((data) => DeleteDataSubscriptionResponse.decode(new _m0.Reader(data)));
    }
    CreateOrganization(request) {
        const data = CreateOrganizationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CreateOrganization", data);
        return promise.then((data) => CreateOrganizationResponse.decode(new _m0.Reader(data)));
    }
    GetOrganizationByName(request) {
        const data = GetOrganizationByNameRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetOrganizationByName", data);
        return promise.then((data) => GetOrganizationByNameResponse.decode(new _m0.Reader(data)));
    }
    GetOrganizations(request) {
        const data = GetOrganizationsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetOrganizations", data);
        return promise.then((data) => GetOrganizationsResponse.decode(new _m0.Reader(data)));
    }
    GetAuditLogEvents(request) {
        const data = GetAuditLogEventsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetAuditLogEvents", data);
        return promise.then((data) => GetAuditLogEventsResponse.decode(new _m0.Reader(data)));
    }
}
export const DataApiAdminDefinition = {
    name: "DataApiAdmin",
    fullName: "devvit.dev_portal.DataApiAdmin",
    methods: {
        /** Data Token */
        createDataToken: {
            name: "CreateDataToken",
            requestType: CreateDataTokenRequest,
            requestStream: false,
            responseType: CreateDataTokenResponse,
            responseStream: false,
            options: {},
        },
        getDataTokenById: {
            name: "GetDataTokenById",
            requestType: GetDataTokenByIdRequest,
            requestStream: false,
            responseType: GetDataTokenByIdResponse,
            responseStream: false,
            options: {},
        },
        getDataTokensByOrganizationName: {
            name: "GetDataTokensByOrganizationName",
            requestType: GetDataTokensByOrganizationNameRequest,
            requestStream: false,
            responseType: GetDataTokensByOrganizationNameResponse,
            responseStream: false,
            options: {},
        },
        revokeDataTokenById: {
            name: "RevokeDataTokenById",
            requestType: RevokeDataTokenByIdRequest,
            requestStream: false,
            responseType: RevokeDataTokenByIdResponse,
            responseStream: false,
            options: {},
        },
        /** Data Subscription */
        createDataSubscription: {
            name: "CreateDataSubscription",
            requestType: CreateDataSubscriptionRequest,
            requestStream: false,
            responseType: CreateDataSubscriptionResponse,
            responseStream: false,
            options: {},
        },
        getDataSubscriptionById: {
            name: "GetDataSubscriptionById",
            requestType: GetDataSubscriptionByIdRequest,
            requestStream: false,
            responseType: GetDataSubscriptionByIdResponse,
            responseStream: false,
            options: {},
        },
        getDataSubscriptionsByOrganizationName: {
            name: "GetDataSubscriptionsByOrganizationName",
            requestType: GetDataSubscriptionsByOrganizationNameRequest,
            requestStream: false,
            responseType: GetDataSubscriptionsByOrganizationNameResponse,
            responseStream: false,
            options: {},
        },
        updateDataSubscription: {
            name: "UpdateDataSubscription",
            requestType: UpdateDataSubscriptionRequest,
            requestStream: false,
            responseType: UpdateDataSubscriptionResponse,
            responseStream: false,
            options: {},
        },
        deleteDataSubscription: {
            name: "DeleteDataSubscription",
            requestType: DeleteDataSubscriptionRequest,
            requestStream: false,
            responseType: DeleteDataSubscriptionResponse,
            responseStream: false,
            options: {},
        },
        /** Organizations */
        createOrganization: {
            name: "CreateOrganization",
            requestType: CreateOrganizationRequest,
            requestStream: false,
            responseType: CreateOrganizationResponse,
            responseStream: false,
            options: {},
        },
        getOrganizationByName: {
            name: "GetOrganizationByName",
            requestType: GetOrganizationByNameRequest,
            requestStream: false,
            responseType: GetOrganizationByNameResponse,
            responseStream: false,
            options: {},
        },
        getOrganizations: {
            name: "GetOrganizations",
            requestType: GetOrganizationsRequest,
            requestStream: false,
            responseType: GetOrganizationsResponse,
            responseStream: false,
            options: {},
        },
        /** Audit Log */
        getAuditLogEvents: {
            name: "GetAuditLogEvents",
            requestType: GetAuditLogEventsRequest,
            requestStream: false,
            responseType: GetAuditLogEventsResponse,
            responseStream: false,
            options: {},
        },
    },
};
export class EventsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.Events";
        this.rpc = rpc;
        this.SendEvent = this.SendEvent.bind(this);
    }
    SendEvent(request) {
        const data = Value.encode(request).finish();
        const promise = this.rpc.request(this.service, "SendEvent", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const EventsDefinition = {
    name: "Events",
    fullName: "devvit.dev_portal.Events",
    methods: {
        /** Fire off a client-side event */
        sendEvent: {
            name: "SendEvent",
            requestType: Value,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
export class DevPortalInstallationSettingsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.DevPortalInstallationSettings";
        this.rpc = rpc;
        this.GetForm = this.GetForm.bind(this);
        this.ValidateForm = this.ValidateForm.bind(this);
        this.GetSettings = this.GetSettings.bind(this);
        this.UpdateSettings = this.UpdateSettings.bind(this);
    }
    GetForm(request) {
        const data = GetInstallationSettingsFormRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetForm", data);
        return promise.then((data) => GetInstallationSettingsFormResponse.decode(new _m0.Reader(data)));
    }
    ValidateForm(request) {
        const data = ValidateInstallationSettingsFormRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ValidateForm", data);
        return promise.then((data) => ValidateInstallationSettingsFormResponse.decode(new _m0.Reader(data)));
    }
    GetSettings(request) {
        const data = GetInstallationSettingsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetSettings", data);
        return promise.then((data) => GetInstallationSettingsResponse.decode(new _m0.Reader(data)));
    }
    UpdateSettings(request) {
        const data = UpdateInstallationSettingsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateSettings", data);
        return promise.then((data) => UpdateInstallationSettingsResponse.decode(new _m0.Reader(data)));
    }
}
export const DevPortalInstallationSettingsDefinition = {
    name: "DevPortalInstallationSettings",
    fullName: "devvit.dev_portal.DevPortalInstallationSettings",
    methods: {
        getForm: {
            name: "GetForm",
            requestType: GetInstallationSettingsFormRequest,
            requestStream: false,
            responseType: GetInstallationSettingsFormResponse,
            responseStream: false,
            options: {},
        },
        validateForm: {
            name: "ValidateForm",
            requestType: ValidateInstallationSettingsFormRequest,
            requestStream: false,
            responseType: ValidateInstallationSettingsFormResponse,
            responseStream: false,
            options: {},
        },
        getSettings: {
            name: "GetSettings",
            requestType: GetInstallationSettingsRequest,
            requestStream: false,
            responseType: GetInstallationSettingsResponse,
            responseStream: false,
            options: {},
        },
        updateSettings: {
            name: "UpdateSettings",
            requestType: UpdateInstallationSettingsRequest,
            requestStream: false,
            responseType: UpdateInstallationSettingsResponse,
            responseStream: false,
            options: {},
        },
    },
};
export class DevPortalAppSettingsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.dev_portal.DevPortalAppSettings";
        this.rpc = rpc;
        this.GetForm = this.GetForm.bind(this);
        this.ValidateForm = this.ValidateForm.bind(this);
        this.GetSettings = this.GetSettings.bind(this);
        this.UpdateSettings = this.UpdateSettings.bind(this);
    }
    GetForm(request) {
        const data = GetAppSettingsFormRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetForm", data);
        return promise.then((data) => GetAppSettingsFormResponse.decode(new _m0.Reader(data)));
    }
    ValidateForm(request) {
        const data = ValidateAppSettingsFormRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ValidateForm", data);
        return promise.then((data) => ValidateAppSettingsFormResponse.decode(new _m0.Reader(data)));
    }
    GetSettings(request) {
        const data = GetAppSettingsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetSettings", data);
        return promise.then((data) => GetAppSettingsResponse.decode(new _m0.Reader(data)));
    }
    UpdateSettings(request) {
        const data = UpdateAppSettingsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateSettings", data);
        return promise.then((data) => UpdateAppSettingsResponse.decode(new _m0.Reader(data)));
    }
}
export const DevPortalAppSettingsDefinition = {
    name: "DevPortalAppSettings",
    fullName: "devvit.dev_portal.DevPortalAppSettings",
    methods: {
        getForm: {
            name: "GetForm",
            requestType: GetAppSettingsFormRequest,
            requestStream: false,
            responseType: GetAppSettingsFormResponse,
            responseStream: false,
            options: {},
        },
        validateForm: {
            name: "ValidateForm",
            requestType: ValidateAppSettingsFormRequest,
            requestStream: false,
            responseType: ValidateAppSettingsFormResponse,
            responseStream: false,
            options: {},
        },
        getSettings: {
            name: "GetSettings",
            requestType: GetAppSettingsRequest,
            requestStream: false,
            responseType: GetAppSettingsResponse,
            responseStream: false,
            options: {},
        },
        updateSettings: {
            name: "UpdateSettings",
            requestType: UpdateAppSettingsRequest,
            requestStream: false,
            responseType: UpdateAppSettingsResponse,
            responseStream: false,
            options: {},
        },
    },
};
