// @devvit/protos/types/devvit/ui/events/v1alpha/handle_ui.js
/**
 * #handle_ui.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Struct } from '../../../../google/protobuf/struct.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
import { Effect } from '../../effects/v1alpha/effect.js';
import { UIEvent } from './event.js';
function createBaseHandleUIEventRequest() {
    return { state: undefined, event: undefined };
}
export const HandleUIEventRequest = {
    $type: "devvit.ui.events.v1alpha.HandleUIEventRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.state !== undefined) {
            Struct.encode(Struct.wrap(message.state), writer.uint32(10).fork()).ldelim();
        }
        if (message.event !== undefined) {
            UIEvent.encode(message.event, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseHandleUIEventRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.state = Struct.unwrap(Struct.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.event = UIEvent.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            state: isObject(object.state) ? object.state : undefined,
            event: isSet(object.event) ? UIEvent.fromJSON(object.event) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.state !== undefined && (obj.state = message.state);
        message.event !== undefined && (obj.event = message.event ? UIEvent.toJSON(message.event) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseHandleUIEventRequest();
        message.state = object.state ?? undefined;
        message.event = (object.event !== undefined && object.event !== null)
            ? UIEvent.fromPartial(object.event)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.events.v1alpha.HandleUIEventRequest", HandleUIEventRequest);
function createBaseHandleUIEventResponse() {
    return { state: undefined, effects: [] };
}
export const HandleUIEventResponse = {
    $type: "devvit.ui.events.v1alpha.HandleUIEventResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.state !== undefined) {
            Struct.encode(Struct.wrap(message.state), writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.effects) {
            Effect.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseHandleUIEventResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.state = Struct.unwrap(Struct.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.effects.push(Effect.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            state: isObject(object.state) ? object.state : undefined,
            effects: Array.isArray(object?.effects) ? object.effects.map((e) => Effect.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.state !== undefined && (obj.state = message.state);
        if (message.effects) {
            obj.effects = message.effects.map((e) => e ? Effect.toJSON(e) : undefined);
        }
        else {
            obj.effects = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseHandleUIEventResponse();
        message.state = object.state ?? undefined;
        message.effects = object.effects?.map((e) => Effect.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.events.v1alpha.HandleUIEventResponse", HandleUIEventResponse);
export class UIEventHandlerClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.ui.events.v1alpha.UIEventHandler";
        this.rpc = rpc;
        this.HandleUIEvent = this.HandleUIEvent.bind(this);
    }
    HandleUIEvent(request) {
        const data = HandleUIEventRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "HandleUIEvent", data);
        return promise.then((data) => HandleUIEventResponse.decode(new _m0.Reader(data)));
    }
}
export const UIEventHandlerDefinition = {
    name: "UIEventHandler",
    fullName: "devvit.ui.events.v1alpha.UIEventHandler",
    methods: {
        handleUIEvent: {
            name: "HandleUIEvent",
            requestType: HandleUIEventRequest,
            requestStream: false,
            responseType: HandleUIEventResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
