// @devvit/protos/types/devvit/dev_portal/feedback/feedback.js
/**
 * #feedback.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseCreateFeedbackRequest() {
    return { content: "" };
}
export const CreateFeedbackRequest = {
    $type: "devvit.dev_portal.feedback.CreateFeedbackRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.content !== "") {
            writer.uint32(10).string(message.content);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateFeedbackRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.content = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { content: isSet(object.content) ? String(object.content) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.content !== undefined && (obj.content = message.content);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCreateFeedbackRequest();
        message.content = object.content ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.feedback.CreateFeedbackRequest", CreateFeedbackRequest);
function isSet(value) {
    return value !== null && value !== undefined;
}
