// @devvit/protos/types/devvit/actor/producer/producer.js
/**
 * #producer.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import Long from "long";
import _m0 from 'protobufjs/minimal.js';
import { map } from "rxjs/operators";
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseArg() {
    return { str: undefined, i32: undefined, i64: undefined };
}
export const Arg = {
    $type: "devvit.actor.producer.Arg",
    encode(message, writer = _m0.Writer.create()) {
        if (message.str !== undefined) {
            writer.uint32(10).string(message.str);
        }
        if (message.i32 !== undefined) {
            writer.uint32(16).int32(message.i32);
        }
        if (message.i64 !== undefined) {
            writer.uint32(24).int64(message.i64);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseArg();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.str = reader.string();
                    break;
                case 2:
                    message.i32 = reader.int32();
                    break;
                case 3:
                    message.i64 = longToNumber(reader.int64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            str: isSet(object.str) ? String(object.str) : undefined,
            i32: isSet(object.i32) ? Number(object.i32) : undefined,
            i64: isSet(object.i64) ? Number(object.i64) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.str !== undefined && (obj.str = message.str);
        message.i32 !== undefined && (obj.i32 = Math.round(message.i32));
        message.i64 !== undefined && (obj.i64 = Math.round(message.i64));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseArg();
        message.str = object.str ?? undefined;
        message.i32 = object.i32 ?? undefined;
        message.i64 = object.i64 ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.producer.Arg", Arg);
function createBaseArgs() {
    return { elements: [] };
}
export const Args = {
    $type: "devvit.actor.producer.Args",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.elements) {
            Arg.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseArgs();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.elements.push(Arg.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { elements: Array.isArray(object?.elements) ? object.elements.map((e) => Arg.fromJSON(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.elements) {
            obj.elements = message.elements.map((e) => e ? Arg.toJSON(e) : undefined);
        }
        else {
            obj.elements = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseArgs();
        message.elements = object.elements?.map((e) => Arg.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.producer.Args", Args);
export class ProducerClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.actor.producer.Producer";
        this.rpc = rpc;
        this.Produce = this.Produce.bind(this);
    }
    Produce(request) {
        const data = Args.encode(request).finish();
        const result = this.rpc.serverStreamingRequest(this.service, "Produce", data);
        return result.pipe(map((data) => Arg.decode(new _m0.Reader(data))));
    }
}
export const ProducerDefinition = {
    name: "Producer",
    fullName: "devvit.actor.producer.Producer",
    methods: {
        produce: {
            name: "Produce",
            requestType: Args,
            requestStream: false,
            responseType: Arg,
            responseStream: true,
            options: {},
        },
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function longToNumber(long) {
    if (long.gt(Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
