// @devvit/protos/types/devvit/reddit/v2alpha/userv2.js
/**
 * #userv2.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import Long from "long";
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { UserFlairV2 } from './flair.js';
function createBaseUserV2() {
    return {
        id: "",
        name: "",
        isGold: false,
        snoovatarImage: "",
        url: "",
        spam: false,
        banned: false,
        flair: undefined,
        karma: 0,
        iconImage: "",
    };
}
export const UserV2 = {
    $type: "devvit.reddit.v2alpha.UserV2",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.isGold !== false) {
            writer.uint32(24).bool(message.isGold);
        }
        if (message.snoovatarImage !== "") {
            writer.uint32(34).string(message.snoovatarImage);
        }
        if (message.url !== "") {
            writer.uint32(42).string(message.url);
        }
        if (message.spam !== false) {
            writer.uint32(48).bool(message.spam);
        }
        if (message.banned !== false) {
            writer.uint32(56).bool(message.banned);
        }
        if (message.flair !== undefined) {
            UserFlairV2.encode(message.flair, writer.uint32(66).fork()).ldelim();
        }
        if (message.karma !== 0) {
            writer.uint32(72).int64(message.karma);
        }
        if (message.iconImage !== "") {
            writer.uint32(82).string(message.iconImage);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUserV2();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.isGold = reader.bool();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.snoovatarImage = reader.string();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.url = reader.string();
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.spam = reader.bool();
                    continue;
                case 7:
                    if (tag !== 56) {
                        break;
                    }
                    message.banned = reader.bool();
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.flair = UserFlairV2.decode(reader, reader.uint32());
                    continue;
                case 9:
                    if (tag !== 72) {
                        break;
                    }
                    message.karma = longToNumber(reader.int64());
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.iconImage = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            isGold: isSet(object.isGold) ? globalThis.Boolean(object.isGold) : false,
            snoovatarImage: isSet(object.snoovatarImage) ? globalThis.String(object.snoovatarImage) : "",
            url: isSet(object.url) ? globalThis.String(object.url) : "",
            spam: isSet(object.spam) ? globalThis.Boolean(object.spam) : false,
            banned: isSet(object.banned) ? globalThis.Boolean(object.banned) : false,
            flair: isSet(object.flair) ? UserFlairV2.fromJSON(object.flair) : undefined,
            karma: isSet(object.karma) ? globalThis.Number(object.karma) : 0,
            iconImage: isSet(object.iconImage) ? globalThis.String(object.iconImage) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.isGold !== false) {
            obj.isGold = message.isGold;
        }
        if (message.snoovatarImage !== "") {
            obj.snoovatarImage = message.snoovatarImage;
        }
        if (message.url !== "") {
            obj.url = message.url;
        }
        if (message.spam !== false) {
            obj.spam = message.spam;
        }
        if (message.banned !== false) {
            obj.banned = message.banned;
        }
        if (message.flair !== undefined) {
            obj.flair = UserFlairV2.toJSON(message.flair);
        }
        if (message.karma !== 0) {
            obj.karma = Math.round(message.karma);
        }
        if (message.iconImage !== "") {
            obj.iconImage = message.iconImage;
        }
        return obj;
    },
    create(base) {
        return UserV2.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseUserV2();
        message.id = object.id ?? "";
        message.name = object.name ?? "";
        message.isGold = object.isGold ?? false;
        message.snoovatarImage = object.snoovatarImage ?? "";
        message.url = object.url ?? "";
        message.spam = object.spam ?? false;
        message.banned = object.banned ?? false;
        message.flair = (object.flair !== undefined && object.flair !== null)
            ? UserFlairV2.fromPartial(object.flair)
            : undefined;
        message.karma = object.karma ?? 0;
        message.iconImage = object.iconImage ?? "";
        return message;
    },
};
messageTypeRegistry.set(UserV2.$type, UserV2);
function longToNumber(long) {
    if (long.gt(globalThis.Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
