// @devvit/protos/types/devvit/plugin/redditapi/subreddits/subreddits_svc.js
/**
 * #subreddits_svc.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../../google/protobuf/empty.js';
import { StringValue } from '../../../../google/protobuf/wrappers.js';
import { JsonStatus, Listing } from '../common/common_msg.js';
import { AboutWhereRequest, BasicSearchRequest, BasicSubredditRequest, BasicWhereRequest, DeleteSrImgRequest, SearchRedditNamesResponse, SearchSubredditsResponse, SiteAdminRequest, StickyRequest, StickyResponse, SubmitTextResponse, SubredditAboutEditRequest, SubredditAboutEditResponse, SubredditAboutResponse, SubredditAboutRulesResponse, SubredditAboutTrafficResponse, SubredditAddRemovalReasonRequest, SubredditAddRemovalReasonResponse, SubredditAutocompleteRequest, SubredditAutocompleteResponse, SubredditGetRemovalReasonsRequest, SubredditGetRemovalReasonsResponse, SubredditPostRequirementsResponse, SubredditsSearchRequest, SubredditsSearchResponse, SubredditStylesheetRequest, SubscribeRequest, UploadSrImgRequest, UploadSrImgResponse, UserSearchResponse, UsersSearchRequest, } from './subreddits_msg.js';
export class SubredditsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.redditapi.subreddits.Subreddits";
        this.rpc = rpc;
        this.AboutWhere = this.AboutWhere.bind(this);
        this.DeleteSrBanner = this.DeleteSrBanner.bind(this);
        this.DeleteSrHeader = this.DeleteSrHeader.bind(this);
        this.DeleteSrIcon = this.DeleteSrIcon.bind(this);
        this.DeleteSrImg = this.DeleteSrImg.bind(this);
        this.SearchRedditNames = this.SearchRedditNames.bind(this);
        this.SearchSubreddits = this.SearchSubreddits.bind(this);
        this.SiteAdmin = this.SiteAdmin.bind(this);
        this.SubmitText = this.SubmitText.bind(this);
        this.SubredditAutocomplete = this.SubredditAutocomplete.bind(this);
        this.SubredditStylesheet = this.SubredditStylesheet.bind(this);
        this.Subscribe = this.Subscribe.bind(this);
        this.UploadSrImg = this.UploadSrImg.bind(this);
        this.SubredditPostRequirements = this.SubredditPostRequirements.bind(this);
        this.SubredditAbout = this.SubredditAbout.bind(this);
        this.SubredditAboutEdit = this.SubredditAboutEdit.bind(this);
        this.SubredditAboutRules = this.SubredditAboutRules.bind(this);
        this.SubredditAboutTraffic = this.SubredditAboutTraffic.bind(this);
        this.Sidebar = this.Sidebar.bind(this);
        this.Sticky = this.Sticky.bind(this);
        this.SubredditsMineWhere = this.SubredditsMineWhere.bind(this);
        this.SubredditsSearch = this.SubredditsSearch.bind(this);
        this.SubredditsWhere = this.SubredditsWhere.bind(this);
        this.UsersSearch = this.UsersSearch.bind(this);
        this.UsersWhere = this.UsersWhere.bind(this);
        this.SubredditAddRemovalReason = this.SubredditAddRemovalReason.bind(this);
        this.SubredditGetRemovalReasons = this.SubredditGetRemovalReasons.bind(this);
    }
    AboutWhere(request) {
        const data = AboutWhereRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "AboutWhere", data);
        return promise.then((data) => Listing.decode(new _m0.Reader(data)));
    }
    DeleteSrBanner(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteSrBanner", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    DeleteSrHeader(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteSrHeader", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    DeleteSrIcon(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteSrIcon", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    DeleteSrImg(request) {
        const data = DeleteSrImgRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteSrImg", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    SearchRedditNames(request) {
        const data = BasicSearchRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SearchRedditNames", data);
        return promise.then((data) => SearchRedditNamesResponse.decode(new _m0.Reader(data)));
    }
    SearchSubreddits(request) {
        const data = BasicSearchRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SearchSubreddits", data);
        return promise.then((data) => SearchSubredditsResponse.decode(new _m0.Reader(data)));
    }
    SiteAdmin(request) {
        const data = SiteAdminRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SiteAdmin", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    SubmitText(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubmitText", data);
        return promise.then((data) => SubmitTextResponse.decode(new _m0.Reader(data)));
    }
    SubredditAutocomplete(request) {
        const data = SubredditAutocompleteRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditAutocomplete", data);
        return promise.then((data) => SubredditAutocompleteResponse.decode(new _m0.Reader(data)));
    }
    SubredditStylesheet(request) {
        const data = SubredditStylesheetRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditStylesheet", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    Subscribe(request) {
        const data = SubscribeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Subscribe", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    UploadSrImg(request) {
        const data = UploadSrImgRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UploadSrImg", data);
        return promise.then((data) => UploadSrImgResponse.decode(new _m0.Reader(data)));
    }
    SubredditPostRequirements(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditPostRequirements", data);
        return promise.then((data) => SubredditPostRequirementsResponse.decode(new _m0.Reader(data)));
    }
    SubredditAbout(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditAbout", data);
        return promise.then((data) => SubredditAboutResponse.decode(new _m0.Reader(data)));
    }
    SubredditAboutEdit(request) {
        const data = SubredditAboutEditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditAboutEdit", data);
        return promise.then((data) => SubredditAboutEditResponse.decode(new _m0.Reader(data)));
    }
    SubredditAboutRules(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditAboutRules", data);
        return promise.then((data) => SubredditAboutRulesResponse.decode(new _m0.Reader(data)));
    }
    SubredditAboutTraffic(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditAboutTraffic", data);
        return promise.then((data) => SubredditAboutTrafficResponse.decode(new _m0.Reader(data)));
    }
    Sidebar(request) {
        const data = BasicSubredditRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Sidebar", data);
        return promise.then((data) => StringValue.decode(new _m0.Reader(data)));
    }
    Sticky(request) {
        const data = StickyRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Sticky", data);
        return promise.then((data) => StickyResponse.decode(new _m0.Reader(data)));
    }
    SubredditsMineWhere(request) {
        const data = BasicWhereRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditsMineWhere", data);
        return promise.then((data) => Listing.decode(new _m0.Reader(data)));
    }
    SubredditsSearch(request) {
        const data = SubredditsSearchRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditsSearch", data);
        return promise.then((data) => SubredditsSearchResponse.decode(new _m0.Reader(data)));
    }
    SubredditsWhere(request) {
        const data = BasicWhereRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditsWhere", data);
        return promise.then((data) => SubredditsSearchResponse.decode(new _m0.Reader(data)));
    }
    UsersSearch(request) {
        const data = UsersSearchRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UsersSearch", data);
        return promise.then((data) => UserSearchResponse.decode(new _m0.Reader(data)));
    }
    UsersWhere(request) {
        const data = BasicWhereRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UsersWhere", data);
        return promise.then((data) => SubredditsSearchResponse.decode(new _m0.Reader(data)));
    }
    SubredditAddRemovalReason(request) {
        const data = SubredditAddRemovalReasonRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditAddRemovalReason", data);
        return promise.then((data) => SubredditAddRemovalReasonResponse.decode(new _m0.Reader(data)));
    }
    SubredditGetRemovalReasons(request) {
        const data = SubredditGetRemovalReasonsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditGetRemovalReasons", data);
        return promise.then((data) => SubredditGetRemovalReasonsResponse.decode(new _m0.Reader(data)));
    }
}
export const SubredditsDefinition = {
    name: "Subreddits",
    fullName: "devvit.plugin.redditapi.subreddits.Subreddits",
    methods: {
        /** https://www.reddit.com/dev/api#GET_about_\{where\} */
        aboutWhere: {
            name: "AboutWhere",
            requestType: AboutWhereRequest,
            requestStream: false,
            responseType: Listing,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_delete_sr_banner */
        deleteSrBanner: {
            name: "DeleteSrBanner",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_delete_sr_header */
        deleteSrHeader: {
            name: "DeleteSrHeader",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_delete_sr_icon */
        deleteSrIcon: {
            name: "DeleteSrIcon",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_delete_sr_img */
        deleteSrImg: {
            name: "DeleteSrImg",
            requestType: DeleteSrImgRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_search_reddit_names */
        searchRedditNames: {
            name: "SearchRedditNames",
            requestType: BasicSearchRequest,
            requestStream: false,
            responseType: SearchRedditNamesResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_search_subreddits */
        searchSubreddits: {
            name: "SearchSubreddits",
            requestType: BasicSearchRequest,
            requestStream: false,
            responseType: SearchSubredditsResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_site_admin */
        siteAdmin: {
            name: "SiteAdmin",
            requestType: SiteAdminRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_api_submit_text */
        submitText: {
            name: "SubmitText",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: SubmitTextResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_api_subreddit_autocomplete_v2 */
        subredditAutocomplete: {
            name: "SubredditAutocomplete",
            requestType: SubredditAutocompleteRequest,
            requestStream: false,
            responseType: SubredditAutocompleteResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_subreddit_stylesheet */
        subredditStylesheet: {
            name: "SubredditStylesheet",
            requestType: SubredditStylesheetRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_subscribe */
        subscribe: {
            name: "Subscribe",
            requestType: SubscribeRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_upload_sr_img */
        uploadSrImg: {
            name: "UploadSrImg",
            requestType: UploadSrImgRequest,
            requestStream: false,
            responseType: UploadSrImgResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_api_v1_\{subreddit\}_post_requirements */
        subredditPostRequirements: {
            name: "SubredditPostRequirements",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: SubredditPostRequirementsResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_r_\{subreddit\}_about */
        subredditAbout: {
            name: "SubredditAbout",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: SubredditAboutResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_r_\{subreddit\}_about_edit */
        subredditAboutEdit: {
            name: "SubredditAboutEdit",
            requestType: SubredditAboutEditRequest,
            requestStream: false,
            responseType: SubredditAboutEditResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_r_\{subreddit\}_about_rules */
        subredditAboutRules: {
            name: "SubredditAboutRules",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: SubredditAboutRulesResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_r_\{subreddit\}_about_traffic */
        subredditAboutTraffic: {
            name: "SubredditAboutTraffic",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: SubredditAboutTrafficResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_sidebar */
        sidebar: {
            name: "Sidebar",
            requestType: BasicSubredditRequest,
            requestStream: false,
            responseType: StringValue,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_sticky */
        sticky: {
            name: "Sticky",
            requestType: StickyRequest,
            requestStream: false,
            responseType: StickyResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_subreddits_mine_\{where\} */
        subredditsMineWhere: {
            name: "SubredditsMineWhere",
            requestType: BasicWhereRequest,
            requestStream: false,
            responseType: Listing,
            responseStream: false,
            options: {},
        },
        /**
         * https://www.reddit.com/dev/api#GET_subreddits_search
         * Make Wrapped
         */
        subredditsSearch: {
            name: "SubredditsSearch",
            requestType: SubredditsSearchRequest,
            requestStream: false,
            responseType: SubredditsSearchResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_subreddits_\{where\} */
        subredditsWhere: {
            name: "SubredditsWhere",
            requestType: BasicWhereRequest,
            requestStream: false,
            responseType: SubredditsSearchResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_users_search */
        usersSearch: {
            name: "UsersSearch",
            requestType: UsersSearchRequest,
            requestStream: false,
            responseType: UserSearchResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_users_\{where\} */
        usersWhere: {
            name: "UsersWhere",
            requestType: BasicWhereRequest,
            requestStream: false,
            responseType: SubredditsSearchResponse,
            responseStream: false,
            options: {},
        },
        /**
         * Add a removal reason to a subreddit.
         *
         * POST /api/v1/{subreddit}/removal_reasons
         */
        subredditAddRemovalReason: {
            name: "SubredditAddRemovalReason",
            requestType: SubredditAddRemovalReasonRequest,
            requestStream: false,
            responseType: SubredditAddRemovalReasonResponse,
            responseStream: false,
            options: {},
        },
        /**
         * Get the list of subreddit's removal reasons.
         *
         * GET /api/v1/{subreddit}/removal_reasons
         */
        subredditGetRemovalReasons: {
            name: "SubredditGetRemovalReasons",
            requestType: SubredditGetRemovalReasonsRequest,
            requestStream: false,
            responseType: SubredditGetRemovalReasonsResponse,
            responseStream: false,
            options: {},
        },
    },
};
