// @devvit/protos/types/devvit/plugin/redditapi/postcollections/postcollections_svc.js
/**
 * #postcollections_svc.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../../google/protobuf/empty.js';
import { AddPostToCollectionRequest, CollectionRequest, CollectionResponse, CreateCollectionRequest, DeleteCollectionRequest, FollowCollectionRequest, RemovePostInCollectionRequest, ReorderCollectionRequest, SubredditCollectionsRequest, SubredditCollectionsResponse, UpdateCollectionDescriptionRequest, UpdateCollectionDisplayLayoutRequest, UpdateCollectionTitleRequest, } from './postcollections_msg.js';
export class PostCollectionsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.redditapi.postcollections.PostCollections";
        this.rpc = rpc;
        this.Collection = this.Collection.bind(this);
        this.SubredditCollections = this.SubredditCollections.bind(this);
        this.Create = this.Create.bind(this);
        this.Delete = this.Delete.bind(this);
        this.AddPost = this.AddPost.bind(this);
        this.RemovePost = this.RemovePost.bind(this);
        this.Follow = this.Follow.bind(this);
        this.Reorder = this.Reorder.bind(this);
        this.UpdateTitle = this.UpdateTitle.bind(this);
        this.UpdateDescription = this.UpdateDescription.bind(this);
        this.UpdateDisplayLayout = this.UpdateDisplayLayout.bind(this);
    }
    Collection(request) {
        const data = CollectionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Collection", data);
        return promise.then((data) => CollectionResponse.decode(new _m0.Reader(data)));
    }
    SubredditCollections(request) {
        const data = SubredditCollectionsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SubredditCollections", data);
        return promise.then((data) => SubredditCollectionsResponse.decode(new _m0.Reader(data)));
    }
    Create(request) {
        const data = CreateCollectionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Create", data);
        return promise.then((data) => CollectionResponse.decode(new _m0.Reader(data)));
    }
    Delete(request) {
        const data = DeleteCollectionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Delete", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    AddPost(request) {
        const data = AddPostToCollectionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "AddPost", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    RemovePost(request) {
        const data = RemovePostInCollectionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RemovePost", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    Follow(request) {
        const data = FollowCollectionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Follow", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    Reorder(request) {
        const data = ReorderCollectionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Reorder", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    UpdateTitle(request) {
        const data = UpdateCollectionTitleRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateTitle", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    UpdateDescription(request) {
        const data = UpdateCollectionDescriptionRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateDescription", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    UpdateDisplayLayout(request) {
        const data = UpdateCollectionDisplayLayoutRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateDisplayLayout", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const PostCollectionsDefinition = {
    name: "PostCollections",
    fullName: "devvit.plugin.redditapi.postcollections.PostCollections",
    methods: {
        /** https://www.reddit.com/dev/api/#GET_api_v1_collections_collection */
        collection: {
            name: "Collection",
            requestType: CollectionRequest,
            requestStream: false,
            responseType: CollectionResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#GET_api_v1_collections_subreddit_collections */
        subredditCollections: {
            name: "SubredditCollections",
            requestType: SubredditCollectionsRequest,
            requestStream: false,
            responseType: SubredditCollectionsResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_create_collection */
        create: {
            name: "Create",
            requestType: CreateCollectionRequest,
            requestStream: false,
            responseType: CollectionResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_delete_collection */
        delete: {
            name: "Delete",
            requestType: DeleteCollectionRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_add_post_to_collection */
        addPost: {
            name: "AddPost",
            requestType: AddPostToCollectionRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_remove_post_in_collection */
        removePost: {
            name: "RemovePost",
            requestType: RemovePostInCollectionRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_follow_collection */
        follow: {
            name: "Follow",
            requestType: FollowCollectionRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_reorder_collection */
        reorder: {
            name: "Reorder",
            requestType: ReorderCollectionRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_update_collection_title */
        updateTitle: {
            name: "UpdateTitle",
            requestType: UpdateCollectionTitleRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_update_collection_description */
        updateDescription: {
            name: "UpdateDescription",
            requestType: UpdateCollectionDescriptionRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api/#POST_api_v1_collections_update_collection_display_layout */
        updateDisplayLayout: {
            name: "UpdateDisplayLayout",
            requestType: UpdateCollectionDisplayLayoutRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
