// @devvit/protos/types/devvit/plugin/gl/glrenderer.js
/**
 * #glrenderer.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import Long from "long";
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { Value } from '../../../google/protobuf/struct.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { gLMethodFromJSON, gLMethodToJSON } from './methods.js';
import { gLMethodTypeFromJSON, gLMethodTypeToJSON } from './methodtype.js';
function createBaseCommandArgument() {
    return { null: undefined, scalar: undefined, blob: undefined, referenceId: undefined };
}
export const CommandArgument = {
    $type: "devvit.plugin.gl.CommandArgument",
    encode(message, writer = _m0.Writer.create()) {
        if (message.null !== undefined) {
            writer.uint32(8).bool(message.null);
        }
        if (message.scalar !== undefined) {
            Value.encode(Value.wrap(message.scalar), writer.uint32(18).fork()).ldelim();
        }
        if (message.blob !== undefined) {
            writer.uint32(26).bytes(message.blob);
        }
        if (message.referenceId !== undefined) {
            writer.uint32(32).uint64(message.referenceId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommandArgument();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.null = reader.bool();
                    break;
                case 2:
                    message.scalar = Value.unwrap(Value.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.blob = reader.bytes();
                    break;
                case 4:
                    message.referenceId = longToNumber(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            null: isSet(object.null) ? Boolean(object.null) : undefined,
            scalar: isSet(object?.scalar) ? object.scalar : undefined,
            blob: isSet(object.blob) ? bytesFromBase64(object.blob) : undefined,
            referenceId: isSet(object.referenceId) ? Number(object.referenceId) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.null !== undefined && (obj.null = message.null);
        message.scalar !== undefined && (obj.scalar = message.scalar);
        message.blob !== undefined && (obj.blob = message.blob !== undefined ? base64FromBytes(message.blob) : undefined);
        message.referenceId !== undefined && (obj.referenceId = Math.round(message.referenceId));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCommandArgument();
        message.null = object.null ?? undefined;
        message.scalar = object.scalar ?? undefined;
        message.blob = object.blob ?? undefined;
        message.referenceId = object.referenceId ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.gl.CommandArgument", CommandArgument);
function createBaseGLCommand() {
    return { method: 0, type: 0, args: [], referenceId: 0 };
}
export const GLCommand = {
    $type: "devvit.plugin.gl.GLCommand",
    encode(message, writer = _m0.Writer.create()) {
        if (message.method !== 0) {
            writer.uint32(8).int32(message.method);
        }
        if (message.type !== 0) {
            writer.uint32(16).int32(message.type);
        }
        for (const v of message.args) {
            CommandArgument.encode(v, writer.uint32(26).fork()).ldelim();
        }
        if (message.referenceId !== 0) {
            writer.uint32(32).uint64(message.referenceId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGLCommand();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.method = reader.int32();
                    break;
                case 2:
                    message.type = reader.int32();
                    break;
                case 3:
                    message.args.push(CommandArgument.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.referenceId = longToNumber(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            method: isSet(object.method) ? gLMethodFromJSON(object.method) : 0,
            type: isSet(object.type) ? gLMethodTypeFromJSON(object.type) : 0,
            args: Array.isArray(object?.args) ? object.args.map((e) => CommandArgument.fromJSON(e)) : [],
            referenceId: isSet(object.referenceId) ? Number(object.referenceId) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.method !== undefined && (obj.method = gLMethodToJSON(message.method));
        message.type !== undefined && (obj.type = gLMethodTypeToJSON(message.type));
        if (message.args) {
            obj.args = message.args.map((e) => e ? CommandArgument.toJSON(e) : undefined);
        }
        else {
            obj.args = [];
        }
        message.referenceId !== undefined && (obj.referenceId = Math.round(message.referenceId));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGLCommand();
        message.method = object.method ?? 0;
        message.type = object.type ?? 0;
        message.args = object.args?.map((e) => CommandArgument.fromPartial(e)) || [];
        message.referenceId = object.referenceId ?? 0;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.gl.GLCommand", GLCommand);
function createBaseExecuteMessage() {
    return { commands: [] };
}
export const ExecuteMessage = {
    $type: "devvit.plugin.gl.ExecuteMessage",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.commands) {
            GLCommand.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseExecuteMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.commands.push(GLCommand.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { commands: Array.isArray(object?.commands) ? object.commands.map((e) => GLCommand.fromJSON(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.commands) {
            obj.commands = message.commands.map((e) => e ? GLCommand.toJSON(e) : undefined);
        }
        else {
            obj.commands = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseExecuteMessage();
        message.commands = object.commands?.map((e) => GLCommand.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.gl.ExecuteMessage", ExecuteMessage);
export class GLRendererClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.gl.GLRenderer";
        this.rpc = rpc;
        this.Execute = this.Execute.bind(this);
    }
    Execute(request) {
        const data = ExecuteMessage.encode(request).finish();
        const promise = this.rpc.request(this.service, "Execute", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const GLRendererDefinition = {
    name: "GLRenderer",
    fullName: "devvit.plugin.gl.GLRenderer",
    methods: {
        execute: {
            name: "Execute",
            requestType: ExecuteMessage,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function longToNumber(long) {
    if (long.gt(Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
