// @devvit/protos/types/devvit/gateway/resolvers.js
/**
 * #resolvers.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../typeRegistry.js';
import { DevvitSubredditMetadata } from '../gql/subreddit_metadata.js';
function createBaseSubredditMetadataRequest() {
    return { subreddit: "" };
}
export const SubredditMetadataRequest = {
    $type: "devvit.gateway.SubredditMetadataRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== "") {
            writer.uint32(10).string(message.subreddit);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubredditMetadataRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subreddit = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { subreddit: isSet(object.subreddit) ? String(object.subreddit) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.subreddit !== undefined && (obj.subreddit = message.subreddit);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseSubredditMetadataRequest();
        message.subreddit = object.subreddit ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.SubredditMetadataRequest", SubredditMetadataRequest);
export class SubredditMetadataResolverClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.gateway.SubredditMetadataResolver";
        this.rpc = rpc;
        this.GetSubredditMetadata = this.GetSubredditMetadata.bind(this);
    }
    GetSubredditMetadata(request) {
        const data = SubredditMetadataRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetSubredditMetadata", data);
        return promise.then((data) => DevvitSubredditMetadata.decode(new _m0.Reader(data)));
    }
}
export const SubredditMetadataResolverDefinition = {
    name: "SubredditMetadataResolver",
    fullName: "devvit.gateway.SubredditMetadataResolver",
    methods: {
        getSubredditMetadata: {
            name: "GetSubredditMetadata",
            requestType: SubredditMetadataRequest,
            requestStream: false,
            responseType: DevvitSubredditMetadata,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
