// @devvit/protos/types/devvit/plugin/devserver/devserver.js
/**
 * #devserver.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { map } from "rxjs/operators";
import { Empty } from '../../../google/protobuf/empty.js';
import { Timestamp } from '../../../google/protobuf/timestamp.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { LinkedBundle } from '../../runtime/bundle.js';
import { CompileLog } from '../buildpack/buildpack_common.js';
import { File, FileSystem } from '../buildpack/filesystem.js';
import { TestCasesList } from './testcases.js';
function createBaseProjectListRequest() {
    return { projectRootDirectory: "" };
}
export const ProjectListRequest = {
    $type: "devvit.plugin.devserver.ProjectListRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.projectRootDirectory !== "") {
            writer.uint32(10).string(message.projectRootDirectory);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProjectListRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.projectRootDirectory = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { projectRootDirectory: isSet(object.projectRootDirectory) ? String(object.projectRootDirectory) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.projectRootDirectory !== undefined && (obj.projectRootDirectory = message.projectRootDirectory);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseProjectListRequest();
        message.projectRootDirectory = object.projectRootDirectory ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ProjectListRequest", ProjectListRequest);
function createBaseProjectBuildProblems() {
    return { warnings: [], errors: [] };
}
export const ProjectBuildProblems = {
    $type: "devvit.plugin.devserver.ProjectBuildProblems",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.warnings) {
            CompileLog.encode(v, writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.errors) {
            CompileLog.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProjectBuildProblems();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.warnings.push(CompileLog.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.errors.push(CompileLog.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            warnings: Array.isArray(object?.warnings) ? object.warnings.map((e) => CompileLog.fromJSON(e)) : [],
            errors: Array.isArray(object?.errors) ? object.errors.map((e) => CompileLog.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.warnings) {
            obj.warnings = message.warnings.map((e) => e ? CompileLog.toJSON(e) : undefined);
        }
        else {
            obj.warnings = [];
        }
        if (message.errors) {
            obj.errors = message.errors.map((e) => e ? CompileLog.toJSON(e) : undefined);
        }
        else {
            obj.errors = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseProjectBuildProblems();
        message.warnings = object.warnings?.map((e) => CompileLog.fromPartial(e)) || [];
        message.errors = object.errors?.map((e) => CompileLog.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ProjectBuildProblems", ProjectBuildProblems);
function createBaseProjectListingResponse() {
    return { projectRootDirectory: "", projectFileSystem: undefined, buildProblems: {} };
}
export const ProjectListingResponse = {
    $type: "devvit.plugin.devserver.ProjectListingResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.projectRootDirectory !== "") {
            writer.uint32(10).string(message.projectRootDirectory);
        }
        if (message.projectFileSystem !== undefined) {
            FileSystem.encode(message.projectFileSystem, writer.uint32(18).fork()).ldelim();
        }
        Object.entries(message.buildProblems).forEach(([key, value]) => {
            ProjectListingResponse_BuildProblemsEntry.encode({ key: key, value }, writer.uint32(26).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProjectListingResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.projectRootDirectory = reader.string();
                    break;
                case 2:
                    message.projectFileSystem = FileSystem.decode(reader, reader.uint32());
                    break;
                case 3:
                    const entry3 = ProjectListingResponse_BuildProblemsEntry.decode(reader, reader.uint32());
                    if (entry3.value !== undefined) {
                        message.buildProblems[entry3.key] = entry3.value;
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            projectRootDirectory: isSet(object.projectRootDirectory) ? String(object.projectRootDirectory) : "",
            projectFileSystem: isSet(object.projectFileSystem) ? FileSystem.fromJSON(object.projectFileSystem) : undefined,
            buildProblems: isObject(object.buildProblems)
                ? Object.entries(object.buildProblems).reduce((acc, [key, value]) => {
                    acc[key] = ProjectBuildProblems.fromJSON(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        message.projectRootDirectory !== undefined && (obj.projectRootDirectory = message.projectRootDirectory);
        message.projectFileSystem !== undefined &&
            (obj.projectFileSystem = message.projectFileSystem ? FileSystem.toJSON(message.projectFileSystem) : undefined);
        obj.buildProblems = {};
        if (message.buildProblems) {
            Object.entries(message.buildProblems).forEach(([k, v]) => {
                obj.buildProblems[k] = ProjectBuildProblems.toJSON(v);
            });
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseProjectListingResponse();
        message.projectRootDirectory = object.projectRootDirectory ?? "";
        message.projectFileSystem = (object.projectFileSystem !== undefined && object.projectFileSystem !== null)
            ? FileSystem.fromPartial(object.projectFileSystem)
            : undefined;
        message.buildProblems = Object.entries(object.buildProblems ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = ProjectBuildProblems.fromPartial(value);
            }
            return acc;
        }, {});
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ProjectListingResponse", ProjectListingResponse);
function createBaseProjectListingResponse_BuildProblemsEntry() {
    return { key: "", value: undefined };
}
export const ProjectListingResponse_BuildProblemsEntry = {
    $type: "devvit.plugin.devserver.ProjectListingResponse.BuildProblemsEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== undefined) {
            ProjectBuildProblems.encode(message.value, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProjectListingResponse_BuildProblemsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.string();
                    break;
                case 2:
                    message.value = ProjectBuildProblems.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? String(object.key) : "",
            value: isSet(object.value) ? ProjectBuildProblems.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined && (obj.value = message.value ? ProjectBuildProblems.toJSON(message.value) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseProjectListingResponse_BuildProblemsEntry();
        message.key = object.key ?? "";
        message.value = (object.value !== undefined && object.value !== null)
            ? ProjectBuildProblems.fromPartial(object.value)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ProjectListingResponse.BuildProblemsEntry", ProjectListingResponse_BuildProblemsEntry);
function createBaseProjectUpdatesRequest() {
    return { projectRootDirectory: "" };
}
export const ProjectUpdatesRequest = {
    $type: "devvit.plugin.devserver.ProjectUpdatesRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.projectRootDirectory !== "") {
            writer.uint32(10).string(message.projectRootDirectory);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProjectUpdatesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.projectRootDirectory = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { projectRootDirectory: isSet(object.projectRootDirectory) ? String(object.projectRootDirectory) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.projectRootDirectory !== undefined && (obj.projectRootDirectory = message.projectRootDirectory);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseProjectUpdatesRequest();
        message.projectRootDirectory = object.projectRootDirectory ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ProjectUpdatesRequest", ProjectUpdatesRequest);
function createBaseProjectUpdateResponse() {
    return { projectRootDirectory: "", file: undefined, linkedBundle: undefined, actorPath: "", buildProblems: {} };
}
export const ProjectUpdateResponse = {
    $type: "devvit.plugin.devserver.ProjectUpdateResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.projectRootDirectory !== "") {
            writer.uint32(10).string(message.projectRootDirectory);
        }
        if (message.file !== undefined) {
            File.encode(message.file, writer.uint32(18).fork()).ldelim();
        }
        if (message.linkedBundle !== undefined) {
            LinkedBundle.encode(message.linkedBundle, writer.uint32(26).fork()).ldelim();
        }
        if (message.actorPath !== "") {
            writer.uint32(34).string(message.actorPath);
        }
        Object.entries(message.buildProblems).forEach(([key, value]) => {
            ProjectUpdateResponse_BuildProblemsEntry.encode({ key: key, value }, writer.uint32(42).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProjectUpdateResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.projectRootDirectory = reader.string();
                    break;
                case 2:
                    message.file = File.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.linkedBundle = LinkedBundle.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.actorPath = reader.string();
                    break;
                case 5:
                    const entry5 = ProjectUpdateResponse_BuildProblemsEntry.decode(reader, reader.uint32());
                    if (entry5.value !== undefined) {
                        message.buildProblems[entry5.key] = entry5.value;
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            projectRootDirectory: isSet(object.projectRootDirectory) ? String(object.projectRootDirectory) : "",
            file: isSet(object.file) ? File.fromJSON(object.file) : undefined,
            linkedBundle: isSet(object.linkedBundle) ? LinkedBundle.fromJSON(object.linkedBundle) : undefined,
            actorPath: isSet(object.actorPath) ? String(object.actorPath) : "",
            buildProblems: isObject(object.buildProblems)
                ? Object.entries(object.buildProblems).reduce((acc, [key, value]) => {
                    acc[key] = ProjectBuildProblems.fromJSON(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        message.projectRootDirectory !== undefined && (obj.projectRootDirectory = message.projectRootDirectory);
        message.file !== undefined && (obj.file = message.file ? File.toJSON(message.file) : undefined);
        message.linkedBundle !== undefined &&
            (obj.linkedBundle = message.linkedBundle ? LinkedBundle.toJSON(message.linkedBundle) : undefined);
        message.actorPath !== undefined && (obj.actorPath = message.actorPath);
        obj.buildProblems = {};
        if (message.buildProblems) {
            Object.entries(message.buildProblems).forEach(([k, v]) => {
                obj.buildProblems[k] = ProjectBuildProblems.toJSON(v);
            });
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseProjectUpdateResponse();
        message.projectRootDirectory = object.projectRootDirectory ?? "";
        message.file = (object.file !== undefined && object.file !== null) ? File.fromPartial(object.file) : undefined;
        message.linkedBundle = (object.linkedBundle !== undefined && object.linkedBundle !== null)
            ? LinkedBundle.fromPartial(object.linkedBundle)
            : undefined;
        message.actorPath = object.actorPath ?? "";
        message.buildProblems = Object.entries(object.buildProblems ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = ProjectBuildProblems.fromPartial(value);
            }
            return acc;
        }, {});
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ProjectUpdateResponse", ProjectUpdateResponse);
function createBaseProjectUpdateResponse_BuildProblemsEntry() {
    return { key: "", value: undefined };
}
export const ProjectUpdateResponse_BuildProblemsEntry = {
    $type: "devvit.plugin.devserver.ProjectUpdateResponse.BuildProblemsEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== undefined) {
            ProjectBuildProblems.encode(message.value, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProjectUpdateResponse_BuildProblemsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.string();
                    break;
                case 2:
                    message.value = ProjectBuildProblems.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? String(object.key) : "",
            value: isSet(object.value) ? ProjectBuildProblems.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined && (obj.value = message.value ? ProjectBuildProblems.toJSON(message.value) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseProjectUpdateResponse_BuildProblemsEntry();
        message.key = object.key ?? "";
        message.value = (object.value !== undefined && object.value !== null)
            ? ProjectBuildProblems.fromPartial(object.value)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ProjectUpdateResponse.BuildProblemsEntry", ProjectUpdateResponse_BuildProblemsEntry);
function createBaseTestCasesRequest() {
    return { actorName: "" };
}
export const TestCasesRequest = {
    $type: "devvit.plugin.devserver.TestCasesRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.actorName !== "") {
            writer.uint32(10).string(message.actorName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestCasesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.actorName = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { actorName: isSet(object.actorName) ? String(object.actorName) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.actorName !== undefined && (obj.actorName = message.actorName);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseTestCasesRequest();
        message.actorName = object.actorName ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.TestCasesRequest", TestCasesRequest);
function createBaseAuthenticationRequest() {
    return {};
}
export const AuthenticationRequest = {
    $type: "devvit.plugin.devserver.AuthenticationRequest",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAuthenticationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseAuthenticationRequest();
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.AuthenticationRequest", AuthenticationRequest);
function createBaseAuthenticationResponse() {
    return { storedToken: undefined };
}
export const AuthenticationResponse = {
    $type: "devvit.plugin.devserver.AuthenticationResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.storedToken !== undefined) {
            StoredToken.encode(message.storedToken, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAuthenticationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.storedToken = StoredToken.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { storedToken: isSet(object.storedToken) ? StoredToken.fromJSON(object.storedToken) : undefined };
    },
    toJSON(message) {
        const obj = {};
        message.storedToken !== undefined &&
            (obj.storedToken = message.storedToken ? StoredToken.toJSON(message.storedToken) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAuthenticationResponse();
        message.storedToken = (object.storedToken !== undefined && object.storedToken !== null)
            ? StoredToken.fromPartial(object.storedToken)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.AuthenticationResponse", AuthenticationResponse);
function createBaseValidationRequest() {
    return {};
}
export const ValidationRequest = {
    $type: "devvit.plugin.devserver.ValidationRequest",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseValidationRequest();
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.ValidationRequest", ValidationRequest);
function createBaseStoredToken() {
    return { refreshToken: "", accessToken: "", expiresAt: undefined, scope: "", tokenType: "" };
}
export const StoredToken = {
    $type: "devvit.plugin.devserver.StoredToken",
    encode(message, writer = _m0.Writer.create()) {
        if (message.refreshToken !== "") {
            writer.uint32(10).string(message.refreshToken);
        }
        if (message.accessToken !== "") {
            writer.uint32(18).string(message.accessToken);
        }
        if (message.expiresAt !== undefined) {
            Timestamp.encode(toTimestamp(message.expiresAt), writer.uint32(26).fork()).ldelim();
        }
        if (message.scope !== "") {
            writer.uint32(34).string(message.scope);
        }
        if (message.tokenType !== "") {
            writer.uint32(42).string(message.tokenType);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStoredToken();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.refreshToken = reader.string();
                    break;
                case 2:
                    message.accessToken = reader.string();
                    break;
                case 3:
                    message.expiresAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.scope = reader.string();
                    break;
                case 5:
                    message.tokenType = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            refreshToken: isSet(object.refreshToken) ? String(object.refreshToken) : "",
            accessToken: isSet(object.accessToken) ? String(object.accessToken) : "",
            expiresAt: isSet(object.expiresAt) ? fromJsonTimestamp(object.expiresAt) : undefined,
            scope: isSet(object.scope) ? String(object.scope) : "",
            tokenType: isSet(object.tokenType) ? String(object.tokenType) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.refreshToken !== undefined && (obj.refreshToken = message.refreshToken);
        message.accessToken !== undefined && (obj.accessToken = message.accessToken);
        message.expiresAt !== undefined && (obj.expiresAt = message.expiresAt.toISOString());
        message.scope !== undefined && (obj.scope = message.scope);
        message.tokenType !== undefined && (obj.tokenType = message.tokenType);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseStoredToken();
        message.refreshToken = object.refreshToken ?? "";
        message.accessToken = object.accessToken ?? "";
        message.expiresAt = object.expiresAt ?? undefined;
        message.scope = object.scope ?? "";
        message.tokenType = object.tokenType ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.StoredToken", StoredToken);
export class DevServerClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.devserver.DevServer";
        this.rpc = rpc;
        this.ListProject = this.ListProject.bind(this);
        this.NotifyBundleUpdates = this.NotifyBundleUpdates.bind(this);
        this.GetTestCases = this.GetTestCases.bind(this);
        this.UpdateTestCases = this.UpdateTestCases.bind(this);
    }
    ListProject(request) {
        const data = ProjectListRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ListProject", data);
        return promise.then((data) => ProjectListingResponse.decode(new _m0.Reader(data)));
    }
    NotifyBundleUpdates(request) {
        const data = ProjectUpdatesRequest.encode(request).finish();
        const result = this.rpc.serverStreamingRequest(this.service, "NotifyBundleUpdates", data);
        return result.pipe(map((data) => ProjectUpdateResponse.decode(new _m0.Reader(data))));
    }
    GetTestCases(request) {
        const data = TestCasesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetTestCases", data);
        return promise.then((data) => TestCasesList.decode(new _m0.Reader(data)));
    }
    UpdateTestCases(request) {
        const data = TestCasesList.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateTestCases", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const DevServerDefinition = {
    name: "DevServer",
    fullName: "devvit.plugin.devserver.DevServer",
    methods: {
        /**
         * Used when the client connects initially to populate the sidebar.
         * This could be updated to take in target project root directory
         * if we support one server powering multiple projects
         */
        listProject: {
            name: "ListProject",
            requestType: ProjectListRequest,
            requestStream: false,
            responseType: ProjectListingResponse,
            responseStream: false,
            options: {},
        },
        /** For now the webUI will request linked bundles over http request caching */
        notifyBundleUpdates: {
            name: "NotifyBundleUpdates",
            requestType: ProjectUpdatesRequest,
            requestStream: false,
            responseType: ProjectUpdateResponse,
            responseStream: true,
            options: {},
        },
        /** Get the list of test cases saved with a given actor */
        getTestCases: {
            name: "GetTestCases",
            requestType: TestCasesRequest,
            requestStream: false,
            responseType: TestCasesList,
            responseStream: false,
            options: {},
        },
        /** Update the list of test cases saved with a given actor */
        updateTestCases: {
            name: "UpdateTestCases",
            requestType: TestCasesList,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
export class OAuthProviderClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.devserver.OAuthProvider";
        this.rpc = rpc;
        this.Authenticate = this.Authenticate.bind(this);
        this.Validate = this.Validate.bind(this);
        this.NotifyAuthenticationUpdates = this.NotifyAuthenticationUpdates.bind(this);
        this.Logout = this.Logout.bind(this);
    }
    Authenticate(request) {
        const data = AuthenticationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Authenticate", data);
        return promise.then((data) => AuthenticationResponse.decode(new _m0.Reader(data)));
    }
    Validate(request) {
        const data = ValidationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Validate", data);
        return promise.then((data) => AuthenticationResponse.decode(new _m0.Reader(data)));
    }
    NotifyAuthenticationUpdates(request) {
        const data = Empty.encode(request).finish();
        const result = this.rpc.serverStreamingRequest(this.service, "NotifyAuthenticationUpdates", data);
        return result.pipe(map((data) => AuthenticationResponse.decode(new _m0.Reader(data))));
    }
    Logout(request) {
        const data = Empty.encode(request).finish();
        const promise = this.rpc.request(this.service, "Logout", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
}
export const OAuthProviderDefinition = {
    name: "OAuthProvider",
    fullName: "devvit.plugin.devserver.OAuthProvider",
    methods: {
        authenticate: {
            name: "Authenticate",
            requestType: AuthenticationRequest,
            requestStream: false,
            responseType: AuthenticationResponse,
            responseStream: false,
            options: {},
        },
        validate: {
            name: "Validate",
            requestType: ValidationRequest,
            requestStream: false,
            responseType: AuthenticationResponse,
            responseStream: false,
            options: {},
        },
        notifyAuthenticationUpdates: {
            name: "NotifyAuthenticationUpdates",
            requestType: Empty,
            requestStream: false,
            responseType: AuthenticationResponse,
            responseStream: true,
            options: {},
        },
        logout: {
            name: "Logout",
            requestType: Empty,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
function toTimestamp(date) {
    const seconds = date.getTime() / 1000;
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = t.seconds * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
