// @devvit/protos/types/devvit/plugin/redditapi/flair/flair_svc.js
/**
 * #flair_svc.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../../google/protobuf/empty.js';
import { JsonStatus } from '../common/common_msg.js';
import { ClearFlairTemplatesRequest, DeleteFlairRequest, DeleteFlairTemplateRequest, FlairArray, FlairConfigRequest, FlairCsvRequest, FlairCsvResponse, FlairListRequest, FlairListResponse, FlairObject, FlairRequest, FlairSelectorRequest, FlairSelectorResponse, FlairTemplateOrderRequest, FlairTemplateRequest, LinkFlairRequest, SelectFlairRequest, SetFlairEnabledRequest, } from './flair_msg.js';
export class FlairClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.redditapi.flair.Flair";
        this.rpc = rpc;
        this.ClearFlairTemplates = this.ClearFlairTemplates.bind(this);
        this.DeleteFlair = this.DeleteFlair.bind(this);
        this.DeleteFlairTemplate = this.DeleteFlairTemplate.bind(this);
        this.Flair = this.Flair.bind(this);
        this.FlairTemplateOrder = this.FlairTemplateOrder.bind(this);
        this.FlairConfig = this.FlairConfig.bind(this);
        this.FlairCsv = this.FlairCsv.bind(this);
        this.FlairList = this.FlairList.bind(this);
        this.FlairSelector = this.FlairSelector.bind(this);
        this.FlairTemplate = this.FlairTemplate.bind(this);
        this.LinkFlair = this.LinkFlair.bind(this);
        this.SelectFlair = this.SelectFlair.bind(this);
        this.SetFlairEnabled = this.SetFlairEnabled.bind(this);
        this.UserFlair = this.UserFlair.bind(this);
    }
    ClearFlairTemplates(request) {
        const data = ClearFlairTemplatesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ClearFlairTemplates", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    DeleteFlair(request) {
        const data = DeleteFlairRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteFlair", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    DeleteFlairTemplate(request) {
        const data = DeleteFlairTemplateRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteFlairTemplate", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    Flair(request) {
        const data = FlairRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Flair", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    FlairTemplateOrder(request) {
        const data = FlairTemplateOrderRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FlairTemplateOrder", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    FlairConfig(request) {
        const data = FlairConfigRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FlairConfig", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    FlairCsv(request) {
        const data = FlairCsvRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FlairCsv", data);
        return promise.then((data) => FlairCsvResponse.decode(new _m0.Reader(data)));
    }
    FlairList(request) {
        const data = FlairListRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FlairList", data);
        return promise.then((data) => FlairListResponse.decode(new _m0.Reader(data)));
    }
    FlairSelector(request) {
        const data = FlairSelectorRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FlairSelector", data);
        return promise.then((data) => FlairSelectorResponse.decode(new _m0.Reader(data)));
    }
    FlairTemplate(request) {
        const data = FlairTemplateRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FlairTemplate", data);
        return promise.then((data) => FlairObject.decode(new _m0.Reader(data)));
    }
    LinkFlair(request) {
        const data = LinkFlairRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "LinkFlair", data);
        return promise.then((data) => FlairArray.decode(new _m0.Reader(data)));
    }
    SelectFlair(request) {
        const data = SelectFlairRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SelectFlair", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    SetFlairEnabled(request) {
        const data = SetFlairEnabledRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SetFlairEnabled", data);
        return promise.then((data) => JsonStatus.decode(new _m0.Reader(data)));
    }
    UserFlair(request) {
        const data = LinkFlairRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UserFlair", data);
        return promise.then((data) => FlairArray.decode(new _m0.Reader(data)));
    }
}
export const FlairDefinition = {
    name: "Flair",
    fullName: "devvit.plugin.redditapi.flair.Flair",
    methods: {
        /** https://www.reddit.com/dev/api#POST_api_clearflairtemplates */
        clearFlairTemplates: {
            name: "ClearFlairTemplates",
            requestType: ClearFlairTemplatesRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_deleteflair */
        deleteFlair: {
            name: "DeleteFlair",
            requestType: DeleteFlairRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_deleteflairtemplate */
        deleteFlairTemplate: {
            name: "DeleteFlairTemplate",
            requestType: DeleteFlairTemplateRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_flair */
        flair: {
            name: "Flair",
            requestType: FlairRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /**
         * https://www.reddit.com/dev/api#PATCH_api_flair_template_order
         * Note: Endpoint expects body as raw array of flair id's
         */
        flairTemplateOrder: {
            name: "FlairTemplateOrder",
            requestType: FlairTemplateOrderRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_flairconfig */
        flairConfig: {
            name: "FlairConfig",
            requestType: FlairConfigRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_flaircsv */
        flairCsv: {
            name: "FlairCsv",
            requestType: FlairCsvRequest,
            requestStream: false,
            responseType: FlairCsvResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#GET_api_flairlist */
        flairList: {
            name: "FlairList",
            requestType: FlairListRequest,
            requestStream: false,
            responseType: FlairListResponse,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_flairselector */
        flairSelector: {
            name: "FlairSelector",
            requestType: FlairSelectorRequest,
            requestStream: false,
            responseType: FlairSelectorResponse,
            responseStream: false,
            options: {},
        },
        /**
         * https://www.reddit.com/dev/api#POST_api_flairtemplate_v2
         * Note: We ignore the "v1" implementation of this method
         * Beware: The JSON response to this endpoint is a FlairObject in camelCase, not snake_case
         * like the other responses.
         */
        flairTemplate: {
            name: "FlairTemplate",
            requestType: FlairTemplateRequest,
            requestStream: false,
            responseType: FlairObject,
            responseStream: false,
            options: {},
        },
        /**
         * https://www.reddit.com/dev/api#GET_api_link_flair_v2
         * Note: We ignore the "v1" implementation of this method
         */
        linkFlair: {
            name: "LinkFlair",
            requestType: LinkFlairRequest,
            requestStream: false,
            responseType: FlairArray,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_selectflair */
        selectFlair: {
            name: "SelectFlair",
            requestType: SelectFlairRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /** https://www.reddit.com/dev/api#POST_api_setflairenabled */
        setFlairEnabled: {
            name: "SetFlairEnabled",
            requestType: SetFlairEnabledRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /**
         * https://www.reddit.com/dev/api#GET_api_user_flair_v2
         * Note: We ignore the "v1" implementation of this method
         */
        userFlair: {
            name: "UserFlair",
            requestType: LinkFlairRequest,
            requestStream: false,
            responseType: FlairArray,
            responseStream: false,
            options: {},
        },
    },
};
