// @devvit/protos/types/devvit/reddit/thing_type.js
/**
 * #thing_type.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
export var ThingType;
(function (ThingType) {
    ThingType[ThingType["UNKNOWN"] = 0] = "UNKNOWN";
    ThingType[ThingType["COMMENT"] = 1] = "COMMENT";
    ThingType[ThingType["ACCOUNT"] = 2] = "ACCOUNT";
    ThingType[ThingType["POST"] = 3] = "POST";
    ThingType[ThingType["MESSAGE"] = 4] = "MESSAGE";
    ThingType[ThingType["SUBREDDIT"] = 5] = "SUBREDDIT";
    ThingType[ThingType["AWARD"] = 6] = "AWARD";
    ThingType[ThingType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(ThingType || (ThingType = {}));
export function thingTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return ThingType.UNKNOWN;
        case 1:
        case "COMMENT":
            return ThingType.COMMENT;
        case 2:
        case "ACCOUNT":
            return ThingType.ACCOUNT;
        case 3:
        case "POST":
            return ThingType.POST;
        case 4:
        case "MESSAGE":
            return ThingType.MESSAGE;
        case 5:
        case "SUBREDDIT":
            return ThingType.SUBREDDIT;
        case 6:
        case "AWARD":
            return ThingType.AWARD;
        case -1:
        case "UNRECOGNIZED":
        default:
            return ThingType.UNRECOGNIZED;
    }
}
export function thingTypeToJSON(object) {
    switch (object) {
        case ThingType.UNKNOWN:
            return 0;
        case ThingType.COMMENT:
            return 1;
        case ThingType.ACCOUNT:
            return 2;
        case ThingType.POST:
            return 3;
        case ThingType.MESSAGE:
            return 4;
        case ThingType.SUBREDDIT:
            return 5;
        case ThingType.AWARD:
            return 6;
        case ThingType.UNRECOGNIZED:
        default:
            return -1;
    }
}
