// @devvit/protos/types/devvit/ui/effects/v1alpha/navigate_to_url.js
/**
 * #navigate_to_url.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
function createBaseNavigateToUrlEffect() {
    return { url: "", target: undefined };
}
export const NavigateToUrlEffect = {
    $type: "devvit.ui.effects.v1alpha.NavigateToUrlEffect",
    encode(message, writer = _m0.Writer.create()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        if (message.target !== undefined) {
            writer.uint32(18).string(message.target);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseNavigateToUrlEffect();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.url = reader.string();
                    break;
                case 2:
                    message.target = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            url: isSet(object.url) ? String(object.url) : "",
            target: isSet(object.target) ? String(object.target) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.url !== undefined && (obj.url = message.url);
        message.target !== undefined && (obj.target = message.target);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseNavigateToUrlEffect();
        message.url = object.url ?? "";
        message.target = object.target ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.NavigateToUrlEffect", NavigateToUrlEffect);
function isSet(value) {
    return value !== null && value !== undefined;
}
