// @devvit/protos/types/devvit/ui/effects/v1alpha/rerender_ui.js
/**
 * #rerender_ui.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
function createBaseRerenderEffect() {
    return { delaySeconds: undefined };
}
export const RerenderEffect = {
    $type: "devvit.ui.effects.v1alpha.RerenderEffect",
    encode(message, writer = _m0.Writer.create()) {
        if (message.delaySeconds !== undefined) {
            writer.uint32(13).float(message.delaySeconds);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRerenderEffect();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.delaySeconds = reader.float();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { delaySeconds: isSet(object.delaySeconds) ? Number(object.delaySeconds) : undefined };
    },
    toJSON(message) {
        const obj = {};
        message.delaySeconds !== undefined && (obj.delaySeconds = message.delaySeconds);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRerenderEffect();
        message.delaySeconds = object.delaySeconds ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.RerenderEffect", RerenderEffect);
function isSet(value) {
    return value !== null && value !== undefined;
}
