// @devvit/protos/types/devvit/ui/effects/v1alpha/send_event.js
/**
 * #send_event.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
import { UIEvent } from '../../events/v1alpha/event.js';
function createBaseSendEventEffect() {
    return { event: undefined };
}
export const SendEventEffect = {
    $type: "devvit.ui.effects.v1alpha.SendEventEffect",
    encode(message, writer = _m0.Writer.create()) {
        if (message.event !== undefined) {
            UIEvent.encode(message.event, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSendEventEffect();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.event = UIEvent.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { event: isSet(object.event) ? UIEvent.fromJSON(object.event) : undefined };
    },
    toJSON(message) {
        const obj = {};
        message.event !== undefined && (obj.event = message.event ? UIEvent.toJSON(message.event) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseSendEventEffect();
        message.event = (object.event !== undefined && object.event !== null)
            ? UIEvent.fromPartial(object.event)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.SendEventEffect", SendEventEffect);
function isSet(value) {
    return value !== null && value !== undefined;
}
