// @devvit/protos/types/devvit/plugin/inspector/inspector.js
/**
 * #inspector.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { map } from "rxjs/operators";
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { SerializableServiceDefinition } from '../../runtime/serializable.js';
export var InspectorUpdateType;
(function (InspectorUpdateType) {
    InspectorUpdateType[InspectorUpdateType["ADDED"] = 0] = "ADDED";
    InspectorUpdateType[InspectorUpdateType["REMOVED"] = 1] = "REMOVED";
    InspectorUpdateType[InspectorUpdateType["CHANGED"] = 2] = "CHANGED";
    /** SUBSCRIBED - The client is now a registered subscriber and will receive updates. */
    InspectorUpdateType[InspectorUpdateType["SUBSCRIBED"] = 3] = "SUBSCRIBED";
    InspectorUpdateType[InspectorUpdateType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(InspectorUpdateType || (InspectorUpdateType = {}));
export function inspectorUpdateTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "ADDED":
            return InspectorUpdateType.ADDED;
        case 1:
        case "REMOVED":
            return InspectorUpdateType.REMOVED;
        case 2:
        case "CHANGED":
            return InspectorUpdateType.CHANGED;
        case 3:
        case "SUBSCRIBED":
            return InspectorUpdateType.SUBSCRIBED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return InspectorUpdateType.UNRECOGNIZED;
    }
}
export function inspectorUpdateTypeToJSON(object) {
    switch (object) {
        case InspectorUpdateType.ADDED:
            return 0;
        case InspectorUpdateType.REMOVED:
            return 1;
        case InspectorUpdateType.CHANGED:
            return 2;
        case InspectorUpdateType.SUBSCRIBED:
            return 3;
        case InspectorUpdateType.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseInspectorQuery() {
    return { type: "", hostname: "" };
}
export const InspectorQuery = {
    $type: "devvit.plugin.inspector.InspectorQuery",
    encode(message, writer = _m0.Writer.create()) {
        if (message.type !== "") {
            writer.uint32(10).string(message.type);
        }
        if (message.hostname !== "") {
            writer.uint32(18).string(message.hostname);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInspectorQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.type = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.hostname = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            type: isSet(object.type) ? globalThis.String(object.type) : "",
            hostname: isSet(object.hostname) ? globalThis.String(object.hostname) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.type !== "") {
            obj.type = message.type;
        }
        if (message.hostname !== "") {
            obj.hostname = message.hostname;
        }
        return obj;
    },
    create(base) {
        return InspectorQuery.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseInspectorQuery();
        message.type = object.type ?? "";
        message.hostname = object.hostname ?? "";
        return message;
    },
};
messageTypeRegistry.set(InspectorQuery.$type, InspectorQuery);
function createBaseInspectorDescription() {
    return { provides: [], hostname: "" };
}
export const InspectorDescription = {
    $type: "devvit.plugin.inspector.InspectorDescription",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.provides) {
            SerializableServiceDefinition.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.hostname !== "") {
            writer.uint32(18).string(message.hostname);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInspectorDescription();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.provides.push(SerializableServiceDefinition.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.hostname = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            provides: globalThis.Array.isArray(object?.provides)
                ? object.provides.map((e) => SerializableServiceDefinition.fromJSON(e))
                : [],
            hostname: isSet(object.hostname) ? globalThis.String(object.hostname) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.provides?.length) {
            obj.provides = message.provides.map((e) => SerializableServiceDefinition.toJSON(e));
        }
        if (message.hostname !== "") {
            obj.hostname = message.hostname;
        }
        return obj;
    },
    create(base) {
        return InspectorDescription.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseInspectorDescription();
        message.provides = object.provides?.map((e) => SerializableServiceDefinition.fromPartial(e)) || [];
        message.hostname = object.hostname ?? "";
        return message;
    },
};
messageTypeRegistry.set(InspectorDescription.$type, InspectorDescription);
function createBaseInspectorDescriptions() {
    return { descriptions: [] };
}
export const InspectorDescriptions = {
    $type: "devvit.plugin.inspector.InspectorDescriptions",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.descriptions) {
            InspectorDescription.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInspectorDescriptions();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.descriptions.push(InspectorDescription.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            descriptions: globalThis.Array.isArray(object?.descriptions)
                ? object.descriptions.map((e) => InspectorDescription.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.descriptions?.length) {
            obj.descriptions = message.descriptions.map((e) => InspectorDescription.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return InspectorDescriptions.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseInspectorDescriptions();
        message.descriptions = object.descriptions?.map((e) => InspectorDescription.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set(InspectorDescriptions.$type, InspectorDescriptions);
function createBaseInspectorUpdate() {
    return { description: undefined, type: 0 };
}
export const InspectorUpdate = {
    $type: "devvit.plugin.inspector.InspectorUpdate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.description !== undefined) {
            InspectorDescription.encode(message.description, writer.uint32(10).fork()).ldelim();
        }
        if (message.type !== 0) {
            writer.uint32(16).int32(message.type);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInspectorUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.description = InspectorDescription.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.type = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            description: isSet(object.description) ? InspectorDescription.fromJSON(object.description) : undefined,
            type: isSet(object.type) ? inspectorUpdateTypeFromJSON(object.type) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.description !== undefined) {
            obj.description = InspectorDescription.toJSON(message.description);
        }
        if (message.type !== 0) {
            obj.type = inspectorUpdateTypeToJSON(message.type);
        }
        return obj;
    },
    create(base) {
        return InspectorUpdate.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseInspectorUpdate();
        message.description = (object.description !== undefined && object.description !== null)
            ? InspectorDescription.fromPartial(object.description)
            : undefined;
        message.type = object.type ?? 0;
        return message;
    },
};
messageTypeRegistry.set(InspectorUpdate.$type, InspectorUpdate);
export const InspectorServiceName = "devvit.plugin.inspector.Inspector";
export class InspectorClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || InspectorServiceName;
        this.rpc = rpc;
        this.ListActors = this.ListActors.bind(this);
        this.ActorChanges = this.ActorChanges.bind(this);
    }
    ListActors(request, metadata) {
        const data = InspectorQuery.encode(request).finish();
        const promise = this.rpc.request(this.service, "ListActors", data, metadata);
        return promise.then((data) => InspectorDescriptions.decode(_m0.Reader.create(data)));
    }
    ActorChanges(request, metadata) {
        const data = Empty.encode(request).finish();
        const result = this.rpc.serverStreamingRequest(this.service, "ActorChanges", data, metadata);
        return result.pipe(map((data) => InspectorUpdate.decode(_m0.Reader.create(data))));
    }
}
export const InspectorDefinition = {
    name: "Inspector",
    fullName: "devvit.plugin.inspector.Inspector",
    methods: {
        listActors: {
            name: "ListActors",
            requestType: InspectorQuery,
            requestStream: false,
            responseType: InspectorDescriptions,
            responseStream: false,
            options: {},
        },
        actorChanges: {
            name: "ActorChanges",
            requestType: Empty,
            requestStream: false,
            responseType: InspectorUpdate,
            responseStream: true,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
