// @devvit/shared-types/Assets.js
export const ASSET_DIRNAME = 'assets';
// List of allowed asset extensions. Note that this should not be used as definitive proof that
// an asset is valid, as it is possible to spoof the extension of a file. This is just a quick
// check to make sure the file is probably an image.
export const ALLOWED_ASSET_EXTENSIONS = ['.png', '.jpg', '.jpeg', '.gif', '.webp'];
export const KILOBYTE = 1024;
export const MEGABYTE = KILOBYTE * KILOBYTE;
export const GIGABYTE = MEGABYTE * KILOBYTE;
export const MAX_ASSET_FOLDER_SIZE_BYTES = GIGABYTE;
export const MAX_ASSET_GIF_SIZE = 40 * MEGABYTE;
export const MAX_ASSET_NON_GIF_SIZE = 20 * MEGABYTE;
export function prettyPrintSize(bytes) {
    const sizeNames = ['B', 'KB', 'MB', 'GB', 'TB'];
    let curSizeName = 0;
    let remainingBytes = bytes;
    while (remainingBytes >= KILOBYTE && curSizeName < sizeNames.length - 1) {
        curSizeName++;
        remainingBytes = remainingBytes / KILOBYTE;
    }
    const prettyByteCount = Math.round(Math.floor(remainingBytes * 100)) / 100;
    return `${prettyByteCount}${sizeNames[curSizeName]}`;
}
