// @devvit/protos/types/devvit/plugin/redditapi/linksandcomments/linksandcomments_svc.js
/**
 * #linksandcomments_svc.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../../google/protobuf/empty.js';
import { JsonStatus, Listing } from '../common/common_msg.js';
import { BasicIdRequest, CommentRequest, FollowPostRequest, InfoRequest, JsonWrappedComment, MoreChildrenRequest, ReportAwardRequest, ReportRequest, SaveRequest, SendRepliesRequest, SetContestModeRequest, SetSubredditStickyRequest, SetSuggestedSortRequest, SubmitRequest, SubmitResponse, VoteRequest, } from './linksandcomments_msg.js';
export const LinksAndCommentsServiceName = "devvit.plugin.redditapi.linksandcomments.LinksAndComments";
export class LinksAndCommentsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || LinksAndCommentsServiceName;
        this.rpc = rpc;
        this.Comment = this.Comment.bind(this);
        this.Del = this.Del.bind(this);
        this.EditUserText = this.EditUserText.bind(this);
        this.FollowPost = this.FollowPost.bind(this);
        this.Hide = this.Hide.bind(this);
        this.Info = this.Info.bind(this);
        this.Lock = this.Lock.bind(this);
        this.MarkNSFW = this.MarkNSFW.bind(this);
        this.MoreChildren = this.MoreChildren.bind(this);
        this.Report = this.Report.bind(this);
        this.ReportAward = this.ReportAward.bind(this);
        this.Save = this.Save.bind(this);
        this.SendReplies = this.SendReplies.bind(this);
        this.SetContestMode = this.SetContestMode.bind(this);
        this.SetSubredditSticky = this.SetSubredditSticky.bind(this);
        this.SetSuggestedSort = this.SetSuggestedSort.bind(this);
        this.Spoiler = this.Spoiler.bind(this);
        this.Submit = this.Submit.bind(this);
        this.Unhide = this.Unhide.bind(this);
        this.Unlock = this.Unlock.bind(this);
        this.UnmarkNSFW = this.UnmarkNSFW.bind(this);
        this.Unsave = this.Unsave.bind(this);
        this.Unspoiler = this.Unspoiler.bind(this);
        this.Vote = this.Vote.bind(this);
    }
    Comment(request, metadata) {
        const data = CommentRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Comment", data, metadata);
        return promise.then((data) => JsonWrappedComment.decode(_m0.Reader.create(data)));
    }
    Del(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Del", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    EditUserText(request, metadata) {
        const data = CommentRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "EditUserText", data, metadata);
        return promise.then((data) => JsonWrappedComment.decode(_m0.Reader.create(data)));
    }
    FollowPost(request, metadata) {
        const data = FollowPostRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FollowPost", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Hide(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Hide", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Info(request, metadata) {
        const data = InfoRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Info", data, metadata);
        return promise.then((data) => Listing.decode(_m0.Reader.create(data)));
    }
    Lock(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Lock", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    MarkNSFW(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "MarkNSFW", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    MoreChildren(request, metadata) {
        const data = MoreChildrenRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "MoreChildren", data, metadata);
        return promise.then((data) => JsonWrappedComment.decode(_m0.Reader.create(data)));
    }
    Report(request, metadata) {
        const data = ReportRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Report", data, metadata);
        return promise.then((data) => JsonStatus.decode(_m0.Reader.create(data)));
    }
    ReportAward(request, metadata) {
        const data = ReportAwardRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ReportAward", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Save(request, metadata) {
        const data = SaveRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Save", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    SendReplies(request, metadata) {
        const data = SendRepliesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SendReplies", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    SetContestMode(request, metadata) {
        const data = SetContestModeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SetContestMode", data, metadata);
        return promise.then((data) => JsonStatus.decode(_m0.Reader.create(data)));
    }
    SetSubredditSticky(request, metadata) {
        const data = SetSubredditStickyRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SetSubredditSticky", data, metadata);
        return promise.then((data) => JsonStatus.decode(_m0.Reader.create(data)));
    }
    SetSuggestedSort(request, metadata) {
        const data = SetSuggestedSortRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SetSuggestedSort", data, metadata);
        return promise.then((data) => JsonStatus.decode(_m0.Reader.create(data)));
    }
    Spoiler(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Spoiler", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Submit(request, metadata) {
        const data = SubmitRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Submit", data, metadata);
        return promise.then((data) => SubmitResponse.decode(_m0.Reader.create(data)));
    }
    Unhide(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Unhide", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Unlock(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Unlock", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    UnmarkNSFW(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UnmarkNSFW", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Unsave(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Unsave", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Unspoiler(request, metadata) {
        const data = BasicIdRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Unspoiler", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Vote(request, metadata) {
        const data = VoteRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Vote", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
}
export const LinksAndCommentsDefinition = {
    name: "LinksAndComments",
    fullName: "devvit.plugin.redditapi.linksandcomments.LinksAndComments",
    methods: {
        /**
         * Submit a new Comment
         * Note: `thing_id` refers to the Post or Comment this comment is replying to
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_comment}
         */
        comment: {
            name: "Comment",
            requestType: CommentRequest,
            requestStream: false,
            responseType: JsonWrappedComment,
            responseStream: false,
            options: {},
        },
        /**
         * Delete a Post or Comment
         * Note: `id` must be a single thing ID for a Post or Comment
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_del}
         */
        del: {
            name: "Del",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Edit the body text of a Comment or Post
         * Note: `thing_id` refers to the Post or Comment being edited
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_editusertext}
         */
        editUserText: {
            name: "EditUserText",
            requestType: CommentRequest,
            requestStream: false,
            responseType: JsonWrappedComment,
            responseStream: false,
            options: {},
        },
        /**
         * Follow or unfollow a Post
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_follow_post}
         */
        followPost: {
            name: "FollowPost",
            requestType: FollowPostRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Hide one or more Posts for the user
         * Note: `id` must be one or more comma-separated Post thing IDs
         *
         * @see {@link Unhide}
         * @see {@link https://www.reddit.com/dev/api#POST_api_hide}
         */
        hide: {
            name: "Hide",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Get one or more things by ID, Subreddit name, or Link Post URL
         *
         * @see {@link https://www.reddit.com/dev/api#GET_api_info}
         */
        info: {
            name: "Info",
            requestType: InfoRequest,
            requestStream: false,
            responseType: Listing,
            responseStream: false,
            options: {},
        },
        /**
         * Disable comments for a Post or Comment
         * Note: `id` must be a single thing ID for a Post or Comment
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_lock}
         */
        lock: {
            name: "Lock",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Marks a Post as Not-Safe-For-Work
         *
         * @see {@link UnmarkNSFW}
         * @see {@link https://www.reddit.com/dev/api#POST_api_marknsfw}
         */
        markNSFW: {
            name: "MarkNSFW",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Get more comments from a truncated Comment tree
         *
         * A comment tree has been truncated if it ends with an entry of kind "more"
         * instead of "t1".  Use this API to fetch the additional comments listed in
         * the "children" array of the "more" entry.
         *
         * @see {@link https://www.reddit.com/dev/api#GET_api_morechildren}
         */
        moreChildren: {
            name: "MoreChildren",
            requestType: MoreChildrenRequest,
            requestStream: false,
            responseType: JsonWrappedComment,
            responseStream: false,
            options: {},
        },
        /**
         * Report a Post, Comment, or Private Message
         *
         * If reporting a Post or Comment the report is sent to the moderators of the subreddit for review.
         * If reporting a Private Message the report is sent to Reddit for review.
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_report}
         */
        report: {
            name: "Report",
            requestType: ReportRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /**
         * Report an award to Reddit for review
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_report_award}
         */
        reportAward: {
            name: "ReportAward",
            requestType: ReportAwardRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Save a Post or Comment to the user's Saved list
         *
         * @see {@link Unsave}
         * @see {@link https://www.reddit.com/dev/api#POST_api_save}
         */
        save: {
            name: "Save",
            requestType: SaveRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Enable or disable reply notifications for a Post or Comment
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_sendreplies}
         */
        sendReplies: {
            name: "SendReplies",
            requestType: SendRepliesRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Toggle Contest Mode for a Post's comments
         *
         * Contest Mode randomizes the sort order and hides the vote score for comments
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_set_contest_mode}
         */
        setContestMode: {
            name: "SetContestMode",
            requestType: SetContestModeRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /**
         * Set the "sticky" or "pinned" status of a Post
         *
         * If the post was created by the current user they can pin the post
         * to the top of their user profile.  Otherwise they must be a moderator
         * of the subreddit to stick it to the top of the subreddit it was posted in.
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_set_subreddit_sticky}
         */
        setSubredditSticky: {
            name: "SetSubredditSticky",
            requestType: SetSubredditStickyRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /**
         * Set the default sort order for comments in the given Post
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_set_suggested_sort}
         */
        setSuggestedSort: {
            name: "SetSuggestedSort",
            requestType: SetSuggestedSortRequest,
            requestStream: false,
            responseType: JsonStatus,
            responseStream: false,
            options: {},
        },
        /**
         * Sets the spoiler tag on a given Post to hide its contents by default
         * Note: `id` must be a single thing ID for a Post
         *
         * @see {@link Unspoiler}
         * @see {@link https://www.reddit.com/dev/api#POST_api_spoiler}
         */
        spoiler: {
            name: "Spoiler",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Create a new Post
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_submit}
         */
        submit: {
            name: "Submit",
            requestType: SubmitRequest,
            requestStream: false,
            responseType: SubmitResponse,
            responseStream: false,
            options: {},
        },
        /**
         * Unhide one or more Posts for the user
         * Note: `id` must be one or more comma-separated Post thing IDs
         *
         * @see {@link Hide}
         * @see {@link https://www.reddit.com/dev/api#POST_api_unhide}
         */
        unhide: {
            name: "Unhide",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Enable comments for a Post or Comment
         * Note: `id` must be a single thing ID for a Post or Comment
         *
         * @see {@link Lock}
         * @see {@link https://www.reddit.com/dev/api#POST_api_unlock}
         */
        unlock: {
            name: "Unlock",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Clears the Not-Safe-For-Work status from a Post
         *
         * @see {@link MarkNSFW}
         * @see {@link https://www.reddit.com/dev/api#POST_api_unmarknsfw}
         */
        unmarkNSFW: {
            name: "UnmarkNSFW",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Remove a Post or Comment from the user's Saved list
         *
         * @see {@link Save}
         * @see {@link https://www.reddit.com/dev/api#POST_api_unsave}
         */
        unsave: {
            name: "Unsave",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Clears the spoiler tag on a given Post
         *
         * @see {@link Spoiler}
         * @see {@link https://www.reddit.com/dev/api#POST_api_unspoiler}
         */
        unspoiler: {
            name: "Unspoiler",
            requestType: BasicIdRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /**
         * Vote or clear a vote on a Post or Comment
         *
         * @see {@link https://www.reddit.com/dev/api#POST_api_vote}
         */
        vote: {
            name: "Vote",
            requestType: VoteRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
