// @devvit/protos/types/devvit/cli/playtest.js
/**
 * #playtest.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../typeRegistry.js';
import { LogMessage } from '../plugin/logger/logger.js';
function createBasePlaytestProviderMessage() {
    return { callState: undefined, connect: undefined };
}
export const PlaytestProviderMessage = {
    $type: "devvit.cli.PlaytestProviderMessage",
    encode(message, writer = _m0.Writer.create()) {
        if (message.callState !== undefined) {
            PlaytestCallState.encode(message.callState, writer.uint32(10).fork()).ldelim();
        }
        if (message.connect !== undefined) {
            PlaytestConnect.encode(message.connect, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePlaytestProviderMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.callState = PlaytestCallState.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.connect = PlaytestConnect.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            callState: isSet(object.callState) ? PlaytestCallState.fromJSON(object.callState) : undefined,
            connect: isSet(object.connect) ? PlaytestConnect.fromJSON(object.connect) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.callState !== undefined &&
            (obj.callState = message.callState ? PlaytestCallState.toJSON(message.callState) : undefined);
        message.connect !== undefined &&
            (obj.connect = message.connect ? PlaytestConnect.toJSON(message.connect) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePlaytestProviderMessage();
        message.callState = (object.callState !== undefined && object.callState !== null)
            ? PlaytestCallState.fromPartial(object.callState)
            : undefined;
        message.connect = (object.connect !== undefined && object.connect !== null)
            ? PlaytestConnect.fromPartial(object.connect)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.cli.PlaytestProviderMessage", PlaytestProviderMessage);
function createBasePlaytestSubscriberMessage() {
    return { appInstalled: undefined };
}
export const PlaytestSubscriberMessage = {
    $type: "devvit.cli.PlaytestSubscriberMessage",
    encode(message, writer = _m0.Writer.create()) {
        if (message.appInstalled !== undefined) {
            PlaytestAppInstalled.encode(message.appInstalled, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePlaytestSubscriberMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.appInstalled = PlaytestAppInstalled.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            appInstalled: isSet(object.appInstalled) ? PlaytestAppInstalled.fromJSON(object.appInstalled) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.appInstalled !== undefined &&
            (obj.appInstalled = message.appInstalled ? PlaytestAppInstalled.toJSON(message.appInstalled) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePlaytestSubscriberMessage();
        message.appInstalled = (object.appInstalled !== undefined && object.appInstalled !== null)
            ? PlaytestAppInstalled.fromPartial(object.appInstalled)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.cli.PlaytestSubscriberMessage", PlaytestSubscriberMessage);
function createBasePlaytestAppInstalled() {
    return {};
}
export const PlaytestAppInstalled = {
    $type: "devvit.cli.PlaytestAppInstalled",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePlaytestAppInstalled();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBasePlaytestAppInstalled();
        return message;
    },
};
messageTypeRegistry.set("devvit.cli.PlaytestAppInstalled", PlaytestAppInstalled);
function createBasePlaytestCallState() {
    return { logs: [] };
}
export const PlaytestCallState = {
    $type: "devvit.cli.PlaytestCallState",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.logs) {
            LogMessage.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePlaytestCallState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.logs.push(LogMessage.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { logs: Array.isArray(object?.logs) ? object.logs.map((e) => LogMessage.fromJSON(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.logs) {
            obj.logs = message.logs.map((e) => e ? LogMessage.toJSON(e) : undefined);
        }
        else {
            obj.logs = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBasePlaytestCallState();
        message.logs = object.logs?.map((e) => LogMessage.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.cli.PlaytestCallState", PlaytestCallState);
function createBasePlaytestConnect() {
    return { url: "" };
}
export const PlaytestConnect = {
    $type: "devvit.cli.PlaytestConnect",
    encode(message, writer = _m0.Writer.create()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePlaytestConnect();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.url = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { url: isSet(object.url) ? String(object.url) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.url !== undefined && (obj.url = message.url);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePlaytestConnect();
        message.url = object.url ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.cli.PlaytestConnect", PlaytestConnect);
function isSet(value) {
    return value !== null && value !== undefined;
}
