// @devvit/protos/types/devvit/dev_portal/data_api/admin/auth/auth.js
/**
 * #auth.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Duration } from '../../../../../google/protobuf/duration.js';
import { Timestamp } from '../../../../../google/protobuf/timestamp.js';
import { messageTypeRegistry } from '../../../../../typeRegistry.js';
import { DataToken } from '../../../../data/api/v1alpha/admin.js';
function createBaseCreateDataTokenRequest() {
    return { organizationName: "", expiresIn: undefined };
}
export const CreateDataTokenRequest = {
    $type: "devvit.dev_portal.data_api.admin.auth.CreateDataTokenRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.organizationName !== "") {
            writer.uint32(10).string(message.organizationName);
        }
        if (message.expiresIn !== undefined) {
            Duration.encode(message.expiresIn, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateDataTokenRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.organizationName = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.expiresIn = Duration.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            organizationName: isSet(object.organizationName) ? globalThis.String(object.organizationName) : "",
            expiresIn: isSet(object.expiresIn) ? Duration.fromJSON(object.expiresIn) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.organizationName !== "") {
            obj.organizationName = message.organizationName;
        }
        if (message.expiresIn !== undefined) {
            obj.expiresIn = Duration.toJSON(message.expiresIn);
        }
        return obj;
    },
    create(base) {
        return CreateDataTokenRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCreateDataTokenRequest();
        message.organizationName = object.organizationName ?? "";
        message.expiresIn = (object.expiresIn !== undefined && object.expiresIn !== null)
            ? Duration.fromPartial(object.expiresIn)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set(CreateDataTokenRequest.$type, CreateDataTokenRequest);
function createBaseCreateDataTokenResponse() {
    return { token: "", expiresAt: undefined };
}
export const CreateDataTokenResponse = {
    $type: "devvit.dev_portal.data_api.admin.auth.CreateDataTokenResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.token !== "") {
            writer.uint32(10).string(message.token);
        }
        if (message.expiresAt !== undefined) {
            Timestamp.encode(toTimestamp(message.expiresAt), writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateDataTokenResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.token = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.expiresAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            token: isSet(object.token) ? globalThis.String(object.token) : "",
            expiresAt: isSet(object.expiresAt) ? fromJsonTimestamp(object.expiresAt) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.token !== "") {
            obj.token = message.token;
        }
        if (message.expiresAt !== undefined) {
            obj.expiresAt = message.expiresAt.toISOString();
        }
        return obj;
    },
    create(base) {
        return CreateDataTokenResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCreateDataTokenResponse();
        message.token = object.token ?? "";
        message.expiresAt = object.expiresAt ?? undefined;
        return message;
    },
};
messageTypeRegistry.set(CreateDataTokenResponse.$type, CreateDataTokenResponse);
function createBaseGetDataTokenByIdRequest() {
    return { id: "" };
}
export const GetDataTokenByIdRequest = {
    $type: "devvit.dev_portal.data_api.admin.auth.GetDataTokenByIdRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetDataTokenByIdRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { id: isSet(object.id) ? globalThis.String(object.id) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        return obj;
    },
    create(base) {
        return GetDataTokenByIdRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetDataTokenByIdRequest();
        message.id = object.id ?? "";
        return message;
    },
};
messageTypeRegistry.set(GetDataTokenByIdRequest.$type, GetDataTokenByIdRequest);
function createBaseGetDataTokenByIdResponse() {
    return { token: undefined };
}
export const GetDataTokenByIdResponse = {
    $type: "devvit.dev_portal.data_api.admin.auth.GetDataTokenByIdResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.token !== undefined) {
            DataToken.encode(message.token, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetDataTokenByIdResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.token = DataToken.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { token: isSet(object.token) ? DataToken.fromJSON(object.token) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.token !== undefined) {
            obj.token = DataToken.toJSON(message.token);
        }
        return obj;
    },
    create(base) {
        return GetDataTokenByIdResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetDataTokenByIdResponse();
        message.token = (object.token !== undefined && object.token !== null)
            ? DataToken.fromPartial(object.token)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set(GetDataTokenByIdResponse.$type, GetDataTokenByIdResponse);
function createBaseGetDataTokensByOrganizationNameRequest() {
    return { organizationName: "" };
}
export const GetDataTokensByOrganizationNameRequest = {
    $type: "devvit.dev_portal.data_api.admin.auth.GetDataTokensByOrganizationNameRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.organizationName !== "") {
            writer.uint32(10).string(message.organizationName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetDataTokensByOrganizationNameRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.organizationName = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { organizationName: isSet(object.organizationName) ? globalThis.String(object.organizationName) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.organizationName !== "") {
            obj.organizationName = message.organizationName;
        }
        return obj;
    },
    create(base) {
        return GetDataTokensByOrganizationNameRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetDataTokensByOrganizationNameRequest();
        message.organizationName = object.organizationName ?? "";
        return message;
    },
};
messageTypeRegistry.set(GetDataTokensByOrganizationNameRequest.$type, GetDataTokensByOrganizationNameRequest);
function createBaseGetDataTokensByOrganizationNameResponse() {
    return { tokens: [] };
}
export const GetDataTokensByOrganizationNameResponse = {
    $type: "devvit.dev_portal.data_api.admin.auth.GetDataTokensByOrganizationNameResponse",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.tokens) {
            DataToken.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetDataTokensByOrganizationNameResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.tokens.push(DataToken.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            tokens: globalThis.Array.isArray(object?.tokens) ? object.tokens.map((e) => DataToken.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.tokens?.length) {
            obj.tokens = message.tokens.map((e) => DataToken.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return GetDataTokensByOrganizationNameResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetDataTokensByOrganizationNameResponse();
        message.tokens = object.tokens?.map((e) => DataToken.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set(GetDataTokensByOrganizationNameResponse.$type, GetDataTokensByOrganizationNameResponse);
function createBaseRevokeDataTokenByIdRequest() {
    return { tokenId: "" };
}
export const RevokeDataTokenByIdRequest = {
    $type: "devvit.dev_portal.data_api.admin.auth.RevokeDataTokenByIdRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.tokenId !== "") {
            writer.uint32(10).string(message.tokenId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRevokeDataTokenByIdRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.tokenId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { tokenId: isSet(object.tokenId) ? globalThis.String(object.tokenId) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.tokenId !== "") {
            obj.tokenId = message.tokenId;
        }
        return obj;
    },
    create(base) {
        return RevokeDataTokenByIdRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRevokeDataTokenByIdRequest();
        message.tokenId = object.tokenId ?? "";
        return message;
    },
};
messageTypeRegistry.set(RevokeDataTokenByIdRequest.$type, RevokeDataTokenByIdRequest);
function createBaseRevokeDataTokenByIdResponse() {
    return {};
}
export const RevokeDataTokenByIdResponse = {
    $type: "devvit.dev_portal.data_api.admin.auth.RevokeDataTokenByIdResponse",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRevokeDataTokenByIdResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return RevokeDataTokenByIdResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseRevokeDataTokenByIdResponse();
        return message;
    },
};
messageTypeRegistry.set(RevokeDataTokenByIdResponse.$type, RevokeDataTokenByIdResponse);
function toTimestamp(date) {
    const seconds = Math.trunc(date.getTime() / 1000);
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = (t.seconds || 0) * 1000;
    millis += (t.nanos || 0) / 1000000;
    return new globalThis.Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof globalThis.Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new globalThis.Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
