// @devvit/runtimes/plugins/redditapi/PostCollections.js
import { CollectionResponse, Empty, SubredditCollectionsResponse } from '@devvit/protos';
import { findAndThrowRedditAPIError } from './errors.js';
import { RedditAPI } from './RedditAPI.js';
import { deepSnakeToCamel } from './snakeToCamel.js';
import { Buffer } from '../../lib/Buffer.js';
export class PostCollections extends RedditAPI {
    async Collection(request, metadata) {
        const response = await this._fetch('Collection', request, metadata);
        return CollectionResponse.fromJSON(this._decodeAndValidateResponseJSON(response));
    }
    async SubredditCollections(request, metadata) {
        const response = await this._fetch('SubredditCollections', request, metadata);
        // Must wrap response.body in Buffer.from
        const body = JSON.parse(new TextDecoder().decode(Buffer.from(response.body)));
        findAndThrowRedditAPIError(response.status, body);
        return SubredditCollectionsResponse.fromJSON({ collections: deepSnakeToCamel(body) });
    }
    async Create(request, metadata) {
        const response = await this._fetch('CreateCollection', request, metadata);
        return CollectionResponse.fromJSON(this._decodeAndValidateResponseJSON(response));
    }
    async AddPost(request, metadata) {
        await this._fetch('AddPostToCollection', request, metadata);
        return Empty.fromJSON({});
    }
    async Delete(request, metadata) {
        await this._fetch('DeleteCollection', request, metadata);
        return Empty.fromJSON({});
    }
    async Follow(request, metadata) {
        await this._fetch('FollowCollection', request, metadata);
        return Empty.fromJSON({});
    }
    async RemovePost(request, metadata) {
        await this._fetch('RemovePostInCollection', request, metadata);
        return Empty.fromJSON({});
    }
    async Reorder(request, metadata) {
        await this._fetch('ReorderCollection', request, metadata);
        return Empty.fromJSON({});
    }
    async UpdateDescription(request, metadata) {
        await this._fetch('UpdateCollectionDescription', request, metadata);
        return Empty.fromJSON({});
    }
    async UpdateDisplayLayout(request, metadata) {
        await this._fetch('UpdateCollectionDisplayLayout', request, metadata);
        return Empty.fromJSON({});
    }
    async UpdateTitle(request, metadata) {
        await this._fetch('UpdateCollectionTitle', request, metadata);
        return Empty.fromJSON({});
    }
}
