// @devvit/protos/types/devvit/plugin/streambroker/streambroker.js
/**
 * #streambroker.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import Long from "long";
import _m0 from 'protobufjs/minimal.js';
import { map } from "rxjs/operators";
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseMessage() {
    return { topic: "", value: "" };
}
export const Message = {
    $type: "devvit.plugin.streambroker.Message",
    encode(message, writer = _m0.Writer.create()) {
        if (message.topic !== "") {
            writer.uint32(10).string(message.topic);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.topic = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            topic: isSet(object.topic) ? globalThis.String(object.topic) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.topic !== "") {
            obj.topic = message.topic;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
    create(base) {
        return Message.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseMessage();
        message.topic = object.topic ?? "";
        message.value = object.value ?? "";
        return message;
    },
};
messageTypeRegistry.set(Message.$type, Message);
function createBaseCursor() {
    return { topic: "", offset: 0, closeAtEnd: false };
}
export const Cursor = {
    $type: "devvit.plugin.streambroker.Cursor",
    encode(message, writer = _m0.Writer.create()) {
        if (message.topic !== "") {
            writer.uint32(10).string(message.topic);
        }
        if (message.offset !== 0) {
            writer.uint32(16).int64(message.offset);
        }
        if (message.closeAtEnd !== false) {
            writer.uint32(24).bool(message.closeAtEnd);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCursor();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.topic = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.offset = longToNumber(reader.int64());
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.closeAtEnd = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            topic: isSet(object.topic) ? globalThis.String(object.topic) : "",
            offset: isSet(object.offset) ? globalThis.Number(object.offset) : 0,
            closeAtEnd: isSet(object.closeAtEnd) ? globalThis.Boolean(object.closeAtEnd) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.topic !== "") {
            obj.topic = message.topic;
        }
        if (message.offset !== 0) {
            obj.offset = Math.round(message.offset);
        }
        if (message.closeAtEnd !== false) {
            obj.closeAtEnd = message.closeAtEnd;
        }
        return obj;
    },
    create(base) {
        return Cursor.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCursor();
        message.topic = object.topic ?? "";
        message.offset = object.offset ?? 0;
        message.closeAtEnd = object.closeAtEnd ?? false;
        return message;
    },
};
messageTypeRegistry.set(Cursor.$type, Cursor);
export const StreamBrokerServiceName = "devvit.plugin.streambroker.StreamBroker";
export class StreamBrokerClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || StreamBrokerServiceName;
        this.rpc = rpc;
        this.Put = this.Put.bind(this);
        this.Get = this.Get.bind(this);
    }
    Put(request, metadata) {
        const data = Message.encode(request).finish();
        const promise = this.rpc.request(this.service, "Put", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Get(request, metadata) {
        const data = Cursor.encode(request).finish();
        const result = this.rpc.serverStreamingRequest(this.service, "Get", data, metadata);
        return result.pipe(map((data) => Message.decode(_m0.Reader.create(data))));
    }
}
export const StreamBrokerDefinition = {
    name: "StreamBroker",
    fullName: "devvit.plugin.streambroker.StreamBroker",
    methods: {
        put: {
            name: "Put",
            requestType: Message,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        get: {
            name: "Get",
            requestType: Cursor,
            requestStream: false,
            responseType: Message,
            responseStream: true,
            options: {},
        },
    },
};
function longToNumber(long) {
    if (long.gt(globalThis.Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
