// @devvit/protos/types/devvit/runtime/actor/supervisor.js
/**
 * #supervisor.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { LinkedBundle } from '../bundle.js';
import { Envelope } from '../envelope.js';
function createBaseUnloadRequest() {
    return { force: false, hostname: "" };
}
export const UnloadRequest = {
    $type: "devvit.runtime.actor.UnloadRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.force !== false) {
            writer.uint32(8).bool(message.force);
        }
        if (message.hostname !== "") {
            writer.uint32(18).string(message.hostname);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUnloadRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.force = reader.bool();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.hostname = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            force: isSet(object.force) ? globalThis.Boolean(object.force) : false,
            hostname: isSet(object.hostname) ? globalThis.String(object.hostname) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.force !== false) {
            obj.force = message.force;
        }
        if (message.hostname !== "") {
            obj.hostname = message.hostname;
        }
        return obj;
    },
    create(base) {
        return UnloadRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseUnloadRequest();
        message.force = object.force ?? false;
        message.hostname = object.hostname ?? "";
        return message;
    },
};
messageTypeRegistry.set(UnloadRequest.$type, UnloadRequest);
export const SupervisorServiceName = "devvit.runtime.actor.Supervisor";
export class SupervisorClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || SupervisorServiceName;
        this.rpc = rpc;
        this.Filter = this.Filter.bind(this);
        this.Load = this.Load.bind(this);
        this.Unload = this.Unload.bind(this);
    }
    Filter(request, metadata) {
        const data = Envelope.encode(request).finish();
        const promise = this.rpc.request(this.service, "Filter", data, metadata);
        return promise.then((data) => Envelope.decode(_m0.Reader.create(data)));
    }
    Load(request, metadata) {
        const data = LinkedBundle.encode(request).finish();
        const promise = this.rpc.request(this.service, "Load", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    Unload(request, metadata) {
        const data = UnloadRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Unload", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
}
export const SupervisorDefinition = {
    name: "Supervisor",
    fullName: "devvit.runtime.actor.Supervisor",
    methods: {
        /**
         * Transform an incoming envelope.
         *
         * Because the input and outputs are Envelopes, they are actually Envelopes
         * wrapped in Envelopes. That is, Envelope.message is an Envelope. The outer
         * Envelope.success flag indicates filter status.
         */
        filter: {
            name: "Filter",
            requestType: Envelope,
            requestStream: false,
            responseType: Envelope,
            responseStream: false,
            options: {},
        },
        /** Unpacks and loads each bundled Actor into their own Workers */
        load: {
            name: "Load",
            requestType: LinkedBundle,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** Unloads the actor at the provided URI, and unloads dependencies if needed */
        unload: {
            name: "Unload",
            requestType: UnloadRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
