// @devvit/shared-types/Result.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Result_error, _Result_value;
export class Result {
    constructor(success, valueOrError) {
        _Result_error.set(this, void 0);
        _Result_value.set(this, void 0);
        this.ok = success;
        if (success) {
            __classPrivateFieldSet(this, _Result_value, valueOrError, "f");
        }
        else {
            if (valueOrError != null) {
                __classPrivateFieldSet(this, _Result_error, valueOrError, "f");
            }
        }
    }
    get value() {
        if (!this.ok) {
            throw new Error('Cannot get value from failed result');
        }
        return __classPrivateFieldGet(this, _Result_value, "f");
    }
    get error() {
        if (this.ok) {
            throw new Error('Cannot get error from non-failed result');
        }
        return __classPrivateFieldGet(this, _Result_error, "f");
    }
    static Ok(value) {
        return new Result(true, value);
    }
    static Err(error) {
        return new Result(false, error);
    }
}
_Result_error = new WeakMap(), _Result_value = new WeakMap();
