// @devvit/protos/types/devvit/runtime/actor/bootstrap.js
/**
 * #bootstrap.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { LinkedBundle } from '../bundle.js';
export const BootstrapActorServiceName = "devvit.runtime.actor.BootstrapActor";
export class BootstrapActorClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || BootstrapActorServiceName;
        this.rpc = rpc;
        this.LoadActor = this.LoadActor.bind(this);
        this.IsResponding = this.IsResponding.bind(this);
    }
    LoadActor(request, metadata) {
        const data = LinkedBundle.encode(request).finish();
        const promise = this.rpc.request(this.service, "LoadActor", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    IsResponding(request, metadata) {
        const data = Empty.encode(request).finish();
        const promise = this.rpc.request(this.service, "IsResponding", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
}
export const BootstrapActorDefinition = {
    name: "BootstrapActor",
    fullName: "devvit.runtime.actor.BootstrapActor",
    methods: {
        /** Loads an Actor into the Worker and binds its service to a URL */
        loadActor: {
            name: "LoadActor",
            requestType: LinkedBundle,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        isResponding: {
            name: "IsResponding",
            requestType: Empty,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
