// @devvit/protos/types/devvit/dev_portal/actor/actor_type.js
/**
 * #actor_type.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseActorTypeInfo() {
    return { id: "", name: "", isEnabled: false };
}
export const ActorTypeInfo = {
    $type: "devvit.dev_portal.actor.ActorTypeInfo",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.isEnabled === true) {
            writer.uint32(24).bool(message.isEnabled);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseActorTypeInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.string();
                    break;
                case 2:
                    message.name = reader.string();
                    break;
                case 3:
                    message.isEnabled = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? String(object.id) : "",
            name: isSet(object.name) ? String(object.name) : "",
            isEnabled: isSet(object.isEnabled) ? Boolean(object.isEnabled) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        message.name !== undefined && (obj.name = message.name);
        message.isEnabled !== undefined && (obj.isEnabled = message.isEnabled);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseActorTypeInfo();
        message.id = object.id ?? "";
        message.name = object.name ?? "";
        message.isEnabled = object.isEnabled ?? false;
        return message;
    },
};
messageTypeRegistry.set("devvit.dev_portal.actor.ActorTypeInfo", ActorTypeInfo);
function isSet(value) {
    return value !== null && value !== undefined;
}
