// @devvit/protos/types/devvit/plugin/redditapi/modnote/modnote_svc.js
/**
 * #modnote_svc.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../../google/protobuf/empty.js';
import { DeleteNotesRequest, DeleteNotesResponse, GetNotesRequest, ModNotesResponse, PostModNotesResponse, PostNotesRequest, PostRemovalNoteRequest, RecentNotesRequest, } from './modnote_msg.js';
export class ModNoteClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.redditapi.modnote.ModNote";
        this.rpc = rpc;
        this.GetNotes = this.GetNotes.bind(this);
        this.DeleteNotes = this.DeleteNotes.bind(this);
        this.PostNotes = this.PostNotes.bind(this);
        this.PostRemovalNote = this.PostRemovalNote.bind(this);
        this.RecentNotes = this.RecentNotes.bind(this);
    }
    GetNotes(request) {
        const data = GetNotesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetNotes", data);
        return promise.then((data) => ModNotesResponse.decode(new _m0.Reader(data)));
    }
    DeleteNotes(request) {
        const data = DeleteNotesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteNotes", data);
        return promise.then((data) => DeleteNotesResponse.decode(new _m0.Reader(data)));
    }
    PostNotes(request) {
        const data = PostNotesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PostNotes", data);
        return promise.then((data) => PostModNotesResponse.decode(new _m0.Reader(data)));
    }
    PostRemovalNote(request) {
        const data = PostRemovalNoteRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PostRemovalNote", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    RecentNotes(request) {
        const data = RecentNotesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RecentNotes", data);
        return promise.then((data) => ModNotesResponse.decode(new _m0.Reader(data)));
    }
}
export const ModNoteDefinition = {
    name: "ModNote",
    fullName: "devvit.plugin.redditapi.modnote.ModNote",
    methods: {
        /** @see {@link https://www.reddit.com/dev/api#GET_api_mod_notes} */
        getNotes: {
            name: "GetNotes",
            requestType: GetNotesRequest,
            requestStream: false,
            responseType: ModNotesResponse,
            responseStream: false,
            options: {},
        },
        /** @see {@link https://www.reddit.com/dev/api/#DELETE_api_mod_notes} */
        deleteNotes: {
            name: "DeleteNotes",
            requestType: DeleteNotesRequest,
            requestStream: false,
            responseType: DeleteNotesResponse,
            responseStream: false,
            options: {},
        },
        /** @see {@link https://www.reddit.com/dev/api#POST_api_mod_notes} */
        postNotes: {
            name: "PostNotes",
            requestType: PostNotesRequest,
            requestStream: false,
            responseType: PostModNotesResponse,
            responseStream: false,
            options: {},
        },
        /** Add a mod note for why a post or comment was removed */
        postRemovalNote: {
            name: "PostRemovalNote",
            requestType: PostRemovalNoteRequest,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        /** @see {@link https://www.reddit.com/dev/api#GET_api_mod_notes_recent} */
        recentNotes: {
            name: "RecentNotes",
            requestType: RecentNotesRequest,
            requestStream: false,
            responseType: ModNotesResponse,
            responseStream: false,
            options: {},
        },
    },
};
