// @devvit/protos/types/devvit/gateway/v1alpha/triggers.js
/**
 * #triggers.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { TriggerEvent } from '../../data/triggers/v1alpha/triggers.js';
import { InstallationInfo } from '../../dev_portal/installation/installation.js';
function createBaseTriggerResponse() {
    return { total: 0, success: [], errors: {} };
}
export const TriggerResponse = {
    $type: "devvit.gateway.v1alpha.TriggerResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.total !== 0) {
            writer.uint32(8).int32(message.total);
        }
        for (const v of message.success) {
            writer.uint32(18).string(v);
        }
        Object.entries(message.errors).forEach(([key, value]) => {
            TriggerResponse_ErrorsEntry.encode({ key: key, value }, writer.uint32(26).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTriggerResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.total = reader.int32();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.success.push(reader.string());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    const entry3 = TriggerResponse_ErrorsEntry.decode(reader, reader.uint32());
                    if (entry3.value !== undefined) {
                        message.errors[entry3.key] = entry3.value;
                    }
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            total: isSet(object.total) ? globalThis.Number(object.total) : 0,
            success: globalThis.Array.isArray(object?.success) ? object.success.map((e) => globalThis.String(e)) : [],
            errors: isObject(object.errors)
                ? Object.entries(object.errors).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.total !== 0) {
            obj.total = Math.round(message.total);
        }
        if (message.success?.length) {
            obj.success = message.success;
        }
        if (message.errors) {
            const entries = Object.entries(message.errors);
            if (entries.length > 0) {
                obj.errors = {};
                entries.forEach(([k, v]) => {
                    obj.errors[k] = v;
                });
            }
        }
        return obj;
    },
    create(base) {
        return TriggerResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseTriggerResponse();
        message.total = object.total ?? 0;
        message.success = object.success?.map((e) => e) || [];
        message.errors = Object.entries(object.errors ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = globalThis.String(value);
            }
            return acc;
        }, {});
        return message;
    },
};
messageTypeRegistry.set(TriggerResponse.$type, TriggerResponse);
function createBaseTriggerResponse_ErrorsEntry() {
    return { key: "", value: "" };
}
export const TriggerResponse_ErrorsEntry = {
    $type: "devvit.gateway.v1alpha.TriggerResponse.ErrorsEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTriggerResponse_ErrorsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
    create(base) {
        return TriggerResponse_ErrorsEntry.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseTriggerResponse_ErrorsEntry();
        message.key = object.key ?? "";
        message.value = object.value ?? "";
        return message;
    },
};
messageTypeRegistry.set(TriggerResponse_ErrorsEntry.$type, TriggerResponse_ErrorsEntry);
function createBaseTriggerFanOutResponse() {
    return { installationIds: [] };
}
export const TriggerFanOutResponse = {
    $type: "devvit.gateway.v1alpha.TriggerFanOutResponse",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.installationIds) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTriggerFanOutResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.installationIds.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            installationIds: globalThis.Array.isArray(object?.installationIds)
                ? object.installationIds.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.installationIds?.length) {
            obj.installationIds = message.installationIds;
        }
        return obj;
    },
    create(base) {
        return TriggerFanOutResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseTriggerFanOutResponse();
        message.installationIds = object.installationIds?.map((e) => e) || [];
        return message;
    },
};
messageTypeRegistry.set(TriggerFanOutResponse.$type, TriggerFanOutResponse);
function createBaseTriggerInvocation() {
    return { installationId: "", evt: undefined };
}
export const TriggerInvocation = {
    $type: "devvit.gateway.v1alpha.TriggerInvocation",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installationId !== "") {
            writer.uint32(10).string(message.installationId);
        }
        if (message.evt !== undefined) {
            TriggerEvent.encode(message.evt, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTriggerInvocation();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.installationId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.evt = TriggerEvent.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            installationId: isSet(object.installationId) ? globalThis.String(object.installationId) : "",
            evt: isSet(object.evt) ? TriggerEvent.fromJSON(object.evt) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.installationId !== "") {
            obj.installationId = message.installationId;
        }
        if (message.evt !== undefined) {
            obj.evt = TriggerEvent.toJSON(message.evt);
        }
        return obj;
    },
    create(base) {
        return TriggerInvocation.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseTriggerInvocation();
        message.installationId = object.installationId ?? "";
        message.evt = (object.evt !== undefined && object.evt !== null) ? TriggerEvent.fromPartial(object.evt) : undefined;
        return message;
    },
};
messageTypeRegistry.set(TriggerInvocation.$type, TriggerInvocation);
export const TriggersServiceName = "devvit.gateway.v1alpha.Triggers";
export class TriggersClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || TriggersServiceName;
        this.rpc = rpc;
        this.OnTrigger = this.OnTrigger.bind(this);
        this.OnFanOut = this.OnFanOut.bind(this);
        this.OnInvoke = this.OnInvoke.bind(this);
        this.OnInstall = this.OnInstall.bind(this);
        this.OnUninstall = this.OnUninstall.bind(this);
        this.OnInstallChanged = this.OnInstallChanged.bind(this);
    }
    OnTrigger(request, metadata) {
        const data = TriggerEvent.encode(request).finish();
        const promise = this.rpc.request(this.service, "OnTrigger", data, metadata);
        return promise.then((data) => TriggerResponse.decode(_m0.Reader.create(data)));
    }
    OnFanOut(request, metadata) {
        const data = TriggerEvent.encode(request).finish();
        const promise = this.rpc.request(this.service, "OnFanOut", data, metadata);
        return promise.then((data) => TriggerFanOutResponse.decode(_m0.Reader.create(data)));
    }
    OnInvoke(request, metadata) {
        const data = TriggerInvocation.encode(request).finish();
        const promise = this.rpc.request(this.service, "OnInvoke", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    OnInstall(request, metadata) {
        const data = InstallationInfo.encode(request).finish();
        const promise = this.rpc.request(this.service, "OnInstall", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    OnUninstall(request, metadata) {
        const data = InstallationInfo.encode(request).finish();
        const promise = this.rpc.request(this.service, "OnUninstall", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    OnInstallChanged(request, metadata) {
        const data = InstallationInfo.encode(request).finish();
        const promise = this.rpc.request(this.service, "OnInstallChanged", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
}
export const TriggersDefinition = {
    name: "Triggers",
    fullName: "devvit.gateway.v1alpha.Triggers",
    methods: {
        onTrigger: {
            name: "OnTrigger",
            requestType: TriggerEvent,
            requestStream: false,
            responseType: TriggerResponse,
            responseStream: false,
            options: {},
        },
        onFanOut: {
            name: "OnFanOut",
            requestType: TriggerEvent,
            requestStream: false,
            responseType: TriggerFanOutResponse,
            responseStream: false,
            options: {},
        },
        onInvoke: {
            name: "OnInvoke",
            requestType: TriggerInvocation,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        onInstall: {
            name: "OnInstall",
            requestType: InstallationInfo,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        onUninstall: {
            name: "OnUninstall",
            requestType: InstallationInfo,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        onInstallChanged: {
            name: "OnInstallChanged",
            requestType: InstallationInfo,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
