// @devvit/protos/types/devvit/plugin/clock/clock.js
/**
 * #clock.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { Timestamp } from '../../../google/protobuf/timestamp.js';
export const ClockServiceName = "devvit.plugin.clock.Clock";
export class ClockClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || ClockServiceName;
        this.rpc = rpc;
        this.Now = this.Now.bind(this);
    }
    Now(request, metadata) {
        const data = Empty.encode(request).finish();
        const promise = this.rpc.request(this.service, "Now", data, metadata);
        return promise.then((data) => Timestamp.decode(_m0.Reader.create(data)));
    }
}
export const ClockDefinition = {
    name: "Clock",
    fullName: "devvit.plugin.clock.Clock",
    methods: {
        now: {
            name: "Now",
            requestType: Empty,
            requestStream: false,
            responseType: Timestamp,
            responseStream: false,
            options: {},
        },
    },
};
