// @devvit/protos/types/devvit/gateway/utils/v1alpha/utilities.js
/**
 * #utilities.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import Long from "long";
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
function createBaseAccessKeyPurgeRequest() {
    return { userId: "", includeRefresh: false };
}
export const AccessKeyPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.AccessKeyPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.userId !== "") {
            writer.uint32(10).string(message.userId);
        }
        if (message.includeRefresh === true) {
            writer.uint32(16).bool(message.includeRefresh);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessKeyPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.userId = reader.string();
                    break;
                case 2:
                    message.includeRefresh = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            userId: isSet(object.userId) ? String(object.userId) : "",
            includeRefresh: isSet(object.includeRefresh) ? Boolean(object.includeRefresh) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.userId !== undefined && (obj.userId = message.userId);
        message.includeRefresh !== undefined && (obj.includeRefresh = message.includeRefresh);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAccessKeyPurgeRequest();
        message.userId = object.userId ?? "";
        message.includeRefresh = object.includeRefresh ?? false;
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.utils.v1alpha.AccessKeyPurgeRequest", AccessKeyPurgeRequest);
function createBaseInstallationsPurgeRequest() {
    return { subreddit: "" };
}
export const InstallationsPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.InstallationsPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== "") {
            writer.uint32(10).string(message.subreddit);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInstallationsPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subreddit = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { subreddit: isSet(object.subreddit) ? String(object.subreddit) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.subreddit !== undefined && (obj.subreddit = message.subreddit);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseInstallationsPurgeRequest();
        message.subreddit = object.subreddit ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.utils.v1alpha.InstallationsPurgeRequest", InstallationsPurgeRequest);
function createBaseInstallationPurgeRequest() {
    return { installationId: "" };
}
export const InstallationPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.InstallationPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installationId !== "") {
            writer.uint32(10).string(message.installationId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInstallationPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.installationId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { installationId: isSet(object.installationId) ? String(object.installationId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.installationId !== undefined && (obj.installationId = message.installationId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseInstallationPurgeRequest();
        message.installationId = object.installationId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.utils.v1alpha.InstallationPurgeRequest", InstallationPurgeRequest);
function createBaseAppVersionPurgeRequest() {
    return { appVersionId: "" };
}
export const AppVersionPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.AppVersionPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.appVersionId !== "") {
            writer.uint32(10).string(message.appVersionId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAppVersionPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.appVersionId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { appVersionId: isSet(object.appVersionId) ? String(object.appVersionId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.appVersionId !== undefined && (obj.appVersionId = message.appVersionId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAppVersionPurgeRequest();
        message.appVersionId = object.appVersionId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.utils.v1alpha.AppVersionPurgeRequest", AppVersionPurgeRequest);
function createBasePurgeResult() {
    return { count: 0 };
}
export const PurgeResult = {
    $type: "devvit.gateway.utils.v1alpha.PurgeResult",
    encode(message, writer = _m0.Writer.create()) {
        if (message.count !== 0) {
            writer.uint32(8).int64(message.count);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePurgeResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.count = longToNumber(reader.int64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { count: isSet(object.count) ? Number(object.count) : 0 };
    },
    toJSON(message) {
        const obj = {};
        message.count !== undefined && (obj.count = Math.round(message.count));
        return obj;
    },
    fromPartial(object) {
        const message = createBasePurgeResult();
        message.count = object.count ?? 0;
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.utils.v1alpha.PurgeResult", PurgeResult);
function createBaseAcceptInviteRequest() {
    return { subredditName: "", appAccountId: "" };
}
export const AcceptInviteRequest = {
    $type: "devvit.gateway.utils.v1alpha.AcceptInviteRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subredditName !== "") {
            writer.uint32(10).string(message.subredditName);
        }
        if (message.appAccountId !== "") {
            writer.uint32(18).string(message.appAccountId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAcceptInviteRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subredditName = reader.string();
                    break;
                case 2:
                    message.appAccountId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subredditName: isSet(object.subredditName) ? String(object.subredditName) : "",
            appAccountId: isSet(object.appAccountId) ? String(object.appAccountId) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.subredditName !== undefined && (obj.subredditName = message.subredditName);
        message.appAccountId !== undefined && (obj.appAccountId = message.appAccountId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAcceptInviteRequest();
        message.subredditName = object.subredditName ?? "";
        message.appAccountId = object.appAccountId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.utils.v1alpha.AcceptInviteRequest", AcceptInviteRequest);
function createBaseAcceptInviteResponse() {
    return {};
}
export const AcceptInviteResponse = {
    $type: "devvit.gateway.utils.v1alpha.AcceptInviteResponse",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAcceptInviteResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseAcceptInviteResponse();
        return message;
    },
};
messageTypeRegistry.set("devvit.gateway.utils.v1alpha.AcceptInviteResponse", AcceptInviteResponse);
export class CacheUtilsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.gateway.utils.v1alpha.CacheUtils";
        this.rpc = rpc;
        this.PurgeAccessKey = this.PurgeAccessKey.bind(this);
        this.PurgeInstallations = this.PurgeInstallations.bind(this);
        this.PurgeInstallation = this.PurgeInstallation.bind(this);
        this.PurgeAppVersion = this.PurgeAppVersion.bind(this);
    }
    PurgeAccessKey(request) {
        const data = AccessKeyPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeAccessKey", data);
        return promise.then((data) => PurgeResult.decode(new _m0.Reader(data)));
    }
    PurgeInstallations(request) {
        const data = InstallationsPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeInstallations", data);
        return promise.then((data) => PurgeResult.decode(new _m0.Reader(data)));
    }
    PurgeInstallation(request) {
        const data = InstallationPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeInstallation", data);
        return promise.then((data) => PurgeResult.decode(new _m0.Reader(data)));
    }
    PurgeAppVersion(request) {
        const data = AppVersionPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeAppVersion", data);
        return promise.then((data) => PurgeResult.decode(new _m0.Reader(data)));
    }
}
export const CacheUtilsDefinition = {
    name: "CacheUtils",
    fullName: "devvit.gateway.utils.v1alpha.CacheUtils",
    methods: {
        purgeAccessKey: {
            name: "PurgeAccessKey",
            requestType: AccessKeyPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
        purgeInstallations: {
            name: "PurgeInstallations",
            requestType: InstallationsPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
        purgeInstallation: {
            name: "PurgeInstallation",
            requestType: InstallationPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
        purgeAppVersion: {
            name: "PurgeAppVersion",
            requestType: AppVersionPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
    },
};
export class ModInviteUtilsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.gateway.utils.v1alpha.ModInviteUtils";
        this.rpc = rpc;
        this.AcceptInvite = this.AcceptInvite.bind(this);
    }
    AcceptInvite(request) {
        const data = AcceptInviteRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "AcceptInvite", data);
        return promise.then((data) => AcceptInviteResponse.decode(new _m0.Reader(data)));
    }
}
export const ModInviteUtilsDefinition = {
    name: "ModInviteUtils",
    fullName: "devvit.gateway.utils.v1alpha.ModInviteUtils",
    methods: {
        acceptInvite: {
            name: "AcceptInvite",
            requestType: AcceptInviteRequest,
            requestStream: false,
            responseType: AcceptInviteResponse,
            responseStream: false,
            options: {},
        },
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function longToNumber(long) {
    if (long.gt(Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
