// @devvit/protos/types/devvit/remote_logger/remote_logger.js
/**
 * #remote_logger.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { map } from "rxjs/operators";
import { Timestamp } from '../../google/protobuf/timestamp.js';
import { messageTypeRegistry } from '../../typeRegistry.js';
import { LogErrorMessage, LogEventMessage, LogMessage } from '../plugin/logger/logger.js';
export var RemoteLogType;
(function (RemoteLogType) {
    RemoteLogType[RemoteLogType["LOG"] = 0] = "LOG";
    RemoteLogType[RemoteLogType["ERROR"] = 1] = "ERROR";
    RemoteLogType[RemoteLogType["EVENT"] = 2] = "EVENT";
    RemoteLogType[RemoteLogType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(RemoteLogType || (RemoteLogType = {}));
export function remoteLogTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "LOG":
            return RemoteLogType.LOG;
        case 1:
        case "ERROR":
            return RemoteLogType.ERROR;
        case 2:
        case "EVENT":
            return RemoteLogType.EVENT;
        case -1:
        case "UNRECOGNIZED":
        default:
            return RemoteLogType.UNRECOGNIZED;
    }
}
export function remoteLogTypeToJSON(object) {
    switch (object) {
        case RemoteLogType.LOG:
            return 0;
        case RemoteLogType.ERROR:
            return 1;
        case RemoteLogType.EVENT:
            return 2;
        case RemoteLogType.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseRemoteLogPutResponse() {
    return {};
}
export const RemoteLogPutResponse = {
    $type: "devvit.remote_logger.RemoteLogPutResponse",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoteLogPutResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = createBaseRemoteLogPutResponse();
        return message;
    },
};
messageTypeRegistry.set("devvit.remote_logger.RemoteLogPutResponse", RemoteLogPutResponse);
function createBaseKeepAliveMessage() {
    return { timestamp: undefined };
}
export const KeepAliveMessage = {
    $type: "devvit.remote_logger.KeepAliveMessage",
    encode(message, writer = _m0.Writer.create()) {
        if (message.timestamp !== undefined) {
            Timestamp.encode(toTimestamp(message.timestamp), writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseKeepAliveMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.timestamp = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { timestamp: isSet(object.timestamp) ? fromJsonTimestamp(object.timestamp) : undefined };
    },
    toJSON(message) {
        const obj = {};
        message.timestamp !== undefined && (obj.timestamp = message.timestamp.toISOString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseKeepAliveMessage();
        message.timestamp = object.timestamp ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.remote_logger.KeepAliveMessage", KeepAliveMessage);
function createBaseRemoteLogMessage() {
    return { log: undefined, error: undefined, keepalive: undefined, event: undefined };
}
export const RemoteLogMessage = {
    $type: "devvit.remote_logger.RemoteLogMessage",
    encode(message, writer = _m0.Writer.create()) {
        if (message.log !== undefined) {
            LogMessage.encode(message.log, writer.uint32(10).fork()).ldelim();
        }
        if (message.error !== undefined) {
            LogErrorMessage.encode(message.error, writer.uint32(18).fork()).ldelim();
        }
        if (message.keepalive !== undefined) {
            KeepAliveMessage.encode(message.keepalive, writer.uint32(26).fork()).ldelim();
        }
        if (message.event !== undefined) {
            LogEventMessage.encode(message.event, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoteLogMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.log = LogMessage.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.error = LogErrorMessage.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.keepalive = KeepAliveMessage.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.event = LogEventMessage.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            log: isSet(object.log) ? LogMessage.fromJSON(object.log) : undefined,
            error: isSet(object.error) ? LogErrorMessage.fromJSON(object.error) : undefined,
            keepalive: isSet(object.keepalive) ? KeepAliveMessage.fromJSON(object.keepalive) : undefined,
            event: isSet(object.event) ? LogEventMessage.fromJSON(object.event) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.log !== undefined && (obj.log = message.log ? LogMessage.toJSON(message.log) : undefined);
        message.error !== undefined && (obj.error = message.error ? LogErrorMessage.toJSON(message.error) : undefined);
        message.keepalive !== undefined &&
            (obj.keepalive = message.keepalive ? KeepAliveMessage.toJSON(message.keepalive) : undefined);
        message.event !== undefined && (obj.event = message.event ? LogEventMessage.toJSON(message.event) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRemoteLogMessage();
        message.log = (object.log !== undefined && object.log !== null) ? LogMessage.fromPartial(object.log) : undefined;
        message.error = (object.error !== undefined && object.error !== null)
            ? LogErrorMessage.fromPartial(object.error)
            : undefined;
        message.keepalive = (object.keepalive !== undefined && object.keepalive !== null)
            ? KeepAliveMessage.fromPartial(object.keepalive)
            : undefined;
        message.event = (object.event !== undefined && object.event !== null)
            ? LogEventMessage.fromPartial(object.event)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.remote_logger.RemoteLogMessage", RemoteLogMessage);
function createBaseRemoteLogInstallationFilter() {
    return { installationId: "" };
}
export const RemoteLogInstallationFilter = {
    $type: "devvit.remote_logger.RemoteLogInstallationFilter",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installationId !== "") {
            writer.uint32(10).string(message.installationId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoteLogInstallationFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.installationId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { installationId: isSet(object.installationId) ? String(object.installationId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.installationId !== undefined && (obj.installationId = message.installationId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRemoteLogInstallationFilter();
        message.installationId = object.installationId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.remote_logger.RemoteLogInstallationFilter", RemoteLogInstallationFilter);
function createBaseRemoteLogSubredditAppNameFilter() {
    return { subreddit: "", appName: "" };
}
export const RemoteLogSubredditAppNameFilter = {
    $type: "devvit.remote_logger.RemoteLogSubredditAppNameFilter",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== "") {
            writer.uint32(18).string(message.subreddit);
        }
        if (message.appName !== "") {
            writer.uint32(26).string(message.appName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoteLogSubredditAppNameFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 2:
                    message.subreddit = reader.string();
                    break;
                case 3:
                    message.appName = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subreddit: isSet(object.subreddit) ? String(object.subreddit) : "",
            appName: isSet(object.appName) ? String(object.appName) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.subreddit !== undefined && (obj.subreddit = message.subreddit);
        message.appName !== undefined && (obj.appName = message.appName);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRemoteLogSubredditAppNameFilter();
        message.subreddit = object.subreddit ?? "";
        message.appName = object.appName ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.remote_logger.RemoteLogSubredditAppNameFilter", RemoteLogSubredditAppNameFilter);
function createBaseRemoteLogQuery() {
    return { installation: undefined, subredditAppName: undefined, type: 0, since: undefined };
}
export const RemoteLogQuery = {
    $type: "devvit.remote_logger.RemoteLogQuery",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installation !== undefined) {
            RemoteLogInstallationFilter.encode(message.installation, writer.uint32(10).fork()).ldelim();
        }
        if (message.subredditAppName !== undefined) {
            RemoteLogSubredditAppNameFilter.encode(message.subredditAppName, writer.uint32(18).fork()).ldelim();
        }
        if (message.type !== 0) {
            writer.uint32(24).int32(message.type);
        }
        if (message.since !== undefined) {
            Timestamp.encode(toTimestamp(message.since), writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoteLogQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.installation = RemoteLogInstallationFilter.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.subredditAppName = RemoteLogSubredditAppNameFilter.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.type = reader.int32();
                    break;
                case 4:
                    message.since = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            installation: isSet(object.installation) ? RemoteLogInstallationFilter.fromJSON(object.installation) : undefined,
            subredditAppName: isSet(object.subredditAppName)
                ? RemoteLogSubredditAppNameFilter.fromJSON(object.subredditAppName)
                : undefined,
            type: isSet(object.type) ? remoteLogTypeFromJSON(object.type) : 0,
            since: isSet(object.since) ? fromJsonTimestamp(object.since) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.installation !== undefined &&
            (obj.installation = message.installation ? RemoteLogInstallationFilter.toJSON(message.installation) : undefined);
        message.subredditAppName !== undefined && (obj.subredditAppName = message.subredditAppName
            ? RemoteLogSubredditAppNameFilter.toJSON(message.subredditAppName)
            : undefined);
        message.type !== undefined && (obj.type = remoteLogTypeToJSON(message.type));
        message.since !== undefined && (obj.since = message.since.toISOString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseRemoteLogQuery();
        message.installation = (object.installation !== undefined && object.installation !== null)
            ? RemoteLogInstallationFilter.fromPartial(object.installation)
            : undefined;
        message.subredditAppName = (object.subredditAppName !== undefined && object.subredditAppName !== null)
            ? RemoteLogSubredditAppNameFilter.fromPartial(object.subredditAppName)
            : undefined;
        message.type = object.type ?? 0;
        message.since = object.since ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.remote_logger.RemoteLogQuery", RemoteLogQuery);
export class RemoteLogPublisherClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.remote_logger.RemoteLogPublisher";
        this.rpc = rpc;
        this.Put = this.Put.bind(this);
    }
    Put(request) {
        const data = RemoteLogMessage.encode(request).finish();
        const promise = this.rpc.request(this.service, "Put", data);
        return promise.then((data) => RemoteLogPutResponse.decode(new _m0.Reader(data)));
    }
}
export const RemoteLogPublisherDefinition = {
    name: "RemoteLogPublisher",
    fullName: "devvit.remote_logger.RemoteLogPublisher",
    methods: {
        put: {
            name: "Put",
            requestType: RemoteLogMessage,
            requestStream: false,
            responseType: RemoteLogPutResponse,
            responseStream: false,
            options: {},
        },
    },
};
export class RemoteLogConsumerClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.remote_logger.RemoteLogConsumer";
        this.rpc = rpc;
        this.Tail = this.Tail.bind(this);
    }
    Tail(request) {
        const data = RemoteLogQuery.encode(request).finish();
        const result = this.rpc.serverStreamingRequest(this.service, "Tail", data);
        return result.pipe(map((data) => RemoteLogMessage.decode(new _m0.Reader(data))));
    }
}
export const RemoteLogConsumerDefinition = {
    name: "RemoteLogConsumer",
    fullName: "devvit.remote_logger.RemoteLogConsumer",
    methods: {
        tail: {
            name: "Tail",
            requestType: RemoteLogQuery,
            requestStream: false,
            responseType: RemoteLogMessage,
            responseStream: true,
            options: {},
        },
    },
};
function toTimestamp(date) {
    const seconds = date.getTime() / 1000;
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = t.seconds * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
