// @devvit/protos/types/devvit/runtime/serializable.js
/**
 * #serializable.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../typeRegistry.js';
function createBaseSerializableServiceDefinition() {
    return { fullName: "", methods: [], name: "", version: "" };
}
export const SerializableServiceDefinition = {
    $type: "devvit.runtime.SerializableServiceDefinition",
    encode(message, writer = _m0.Writer.create()) {
        if (message.fullName !== "") {
            writer.uint32(10).string(message.fullName);
        }
        for (const v of message.methods) {
            SerializableMethodDefinition.encode(v, writer.uint32(18).fork()).ldelim();
        }
        if (message.name !== "") {
            writer.uint32(26).string(message.name);
        }
        if (message.version !== "") {
            writer.uint32(34).string(message.version);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSerializableServiceDefinition();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.fullName = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.methods.push(SerializableMethodDefinition.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.version = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            fullName: isSet(object.fullName) ? globalThis.String(object.fullName) : "",
            methods: globalThis.Array.isArray(object?.methods)
                ? object.methods.map((e) => SerializableMethodDefinition.fromJSON(e))
                : [],
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            version: isSet(object.version) ? globalThis.String(object.version) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.fullName !== "") {
            obj.fullName = message.fullName;
        }
        if (message.methods?.length) {
            obj.methods = message.methods.map((e) => SerializableMethodDefinition.toJSON(e));
        }
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.version !== "") {
            obj.version = message.version;
        }
        return obj;
    },
    create(base) {
        return SerializableServiceDefinition.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSerializableServiceDefinition();
        message.fullName = object.fullName ?? "";
        message.methods = object.methods?.map((e) => SerializableMethodDefinition.fromPartial(e)) || [];
        message.name = object.name ?? "";
        message.version = object.version ?? "";
        return message;
    },
};
messageTypeRegistry.set(SerializableServiceDefinition.$type, SerializableServiceDefinition);
function createBaseSerializableMethodDefinition() {
    return { fullName: "", name: "", requestStream: false, responseStream: false, requestType: "", responseType: "" };
}
export const SerializableMethodDefinition = {
    $type: "devvit.runtime.SerializableMethodDefinition",
    encode(message, writer = _m0.Writer.create()) {
        if (message.fullName !== "") {
            writer.uint32(10).string(message.fullName);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.requestStream !== false) {
            writer.uint32(24).bool(message.requestStream);
        }
        if (message.responseStream !== false) {
            writer.uint32(32).bool(message.responseStream);
        }
        if (message.requestType !== "") {
            writer.uint32(42).string(message.requestType);
        }
        if (message.responseType !== "") {
            writer.uint32(50).string(message.responseType);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSerializableMethodDefinition();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.fullName = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.requestStream = reader.bool();
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.responseStream = reader.bool();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.requestType = reader.string();
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.responseType = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            fullName: isSet(object.fullName) ? globalThis.String(object.fullName) : "",
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            requestStream: isSet(object.requestStream) ? globalThis.Boolean(object.requestStream) : false,
            responseStream: isSet(object.responseStream) ? globalThis.Boolean(object.responseStream) : false,
            requestType: isSet(object.requestType) ? globalThis.String(object.requestType) : "",
            responseType: isSet(object.responseType) ? globalThis.String(object.responseType) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.fullName !== "") {
            obj.fullName = message.fullName;
        }
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.requestStream !== false) {
            obj.requestStream = message.requestStream;
        }
        if (message.responseStream !== false) {
            obj.responseStream = message.responseStream;
        }
        if (message.requestType !== "") {
            obj.requestType = message.requestType;
        }
        if (message.responseType !== "") {
            obj.responseType = message.responseType;
        }
        return obj;
    },
    create(base) {
        return SerializableMethodDefinition.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSerializableMethodDefinition();
        message.fullName = object.fullName ?? "";
        message.name = object.name ?? "";
        message.requestStream = object.requestStream ?? false;
        message.responseStream = object.responseStream ?? false;
        message.requestType = object.requestType ?? "";
        message.responseType = object.responseType ?? "";
        return message;
    },
};
messageTypeRegistry.set(SerializableMethodDefinition.$type, SerializableMethodDefinition);
function isSet(value) {
    return value !== null && value !== undefined;
}
