// @devvit/ui-renderer/client/ui-app.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RemoteApp_customPost, _RemoteApp_uiEventHandler;
import { CustomPostDefinition, UIEventHandlerDefinition } from '@devvit/protos';
/**
 * Adapts an ActorRef (usually provided by the deprecated BrowserRuntime) to the
 * APIs implemented by a UIApp.
 * @deprecated
 */
export class RemoteApp {
    constructor(ref) {
        _RemoteApp_customPost.set(this, void 0);
        _RemoteApp_uiEventHandler.set(this, void 0);
        __classPrivateFieldSet(this, _RemoteApp_customPost, ref.As(CustomPostDefinition), "f");
        __classPrivateFieldSet(this, _RemoteApp_uiEventHandler, ref.As(UIEventHandlerDefinition), "f");
    }
    HandleUIEvent(req, meta) {
        return __classPrivateFieldGet(this, _RemoteApp_uiEventHandler, "f").HandleUIEvent(req, meta);
    }
    RenderPost(req, meta) {
        return __classPrivateFieldGet(this, _RemoteApp_customPost, "f").RenderPost(req, meta);
    }
    RenderPostContent(req, meta) {
        return __classPrivateFieldGet(this, _RemoteApp_customPost, "f").RenderPostContent(req, meta);
    }
    RenderPostComposer(req, meta) {
        return __classPrivateFieldGet(this, _RemoteApp_customPost, "f").RenderPostComposer(req, meta);
    }
}
_RemoteApp_customPost = new WeakMap(), _RemoteApp_uiEventHandler = new WeakMap();
