// @devvit/protos/types/devvit/gateway/utils/v1alpha/utilities.js
/**
 * #utilities.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import Long from "long";
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
function createBaseAccessKeyPurgeRequest() {
    return { userId: "", includeRefresh: false };
}
export const AccessKeyPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.AccessKeyPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.userId !== "") {
            writer.uint32(10).string(message.userId);
        }
        if (message.includeRefresh !== false) {
            writer.uint32(16).bool(message.includeRefresh);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessKeyPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.userId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.includeRefresh = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            userId: isSet(object.userId) ? globalThis.String(object.userId) : "",
            includeRefresh: isSet(object.includeRefresh) ? globalThis.Boolean(object.includeRefresh) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.userId !== "") {
            obj.userId = message.userId;
        }
        if (message.includeRefresh !== false) {
            obj.includeRefresh = message.includeRefresh;
        }
        return obj;
    },
    create(base) {
        return AccessKeyPurgeRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAccessKeyPurgeRequest();
        message.userId = object.userId ?? "";
        message.includeRefresh = object.includeRefresh ?? false;
        return message;
    },
};
messageTypeRegistry.set(AccessKeyPurgeRequest.$type, AccessKeyPurgeRequest);
function createBaseInstallationsPurgeRequest() {
    return { subreddit: "" };
}
export const InstallationsPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.InstallationsPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== "") {
            writer.uint32(10).string(message.subreddit);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInstallationsPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.subreddit = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { subreddit: isSet(object.subreddit) ? globalThis.String(object.subreddit) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.subreddit !== "") {
            obj.subreddit = message.subreddit;
        }
        return obj;
    },
    create(base) {
        return InstallationsPurgeRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseInstallationsPurgeRequest();
        message.subreddit = object.subreddit ?? "";
        return message;
    },
};
messageTypeRegistry.set(InstallationsPurgeRequest.$type, InstallationsPurgeRequest);
function createBaseInstallationPurgeRequest() {
    return { installationId: "" };
}
export const InstallationPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.InstallationPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installationId !== "") {
            writer.uint32(10).string(message.installationId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInstallationPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.installationId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { installationId: isSet(object.installationId) ? globalThis.String(object.installationId) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.installationId !== "") {
            obj.installationId = message.installationId;
        }
        return obj;
    },
    create(base) {
        return InstallationPurgeRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseInstallationPurgeRequest();
        message.installationId = object.installationId ?? "";
        return message;
    },
};
messageTypeRegistry.set(InstallationPurgeRequest.$type, InstallationPurgeRequest);
function createBaseAppVersionPurgeRequest() {
    return { appVersionId: "" };
}
export const AppVersionPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.AppVersionPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.appVersionId !== "") {
            writer.uint32(10).string(message.appVersionId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAppVersionPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.appVersionId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { appVersionId: isSet(object.appVersionId) ? globalThis.String(object.appVersionId) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.appVersionId !== "") {
            obj.appVersionId = message.appVersionId;
        }
        return obj;
    },
    create(base) {
        return AppVersionPurgeRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAppVersionPurgeRequest();
        message.appVersionId = object.appVersionId ?? "";
        return message;
    },
};
messageTypeRegistry.set(AppVersionPurgeRequest.$type, AppVersionPurgeRequest);
function createBaseAppPurgeRequest() {
    return { appId: "" };
}
export const AppPurgeRequest = {
    $type: "devvit.gateway.utils.v1alpha.AppPurgeRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.appId !== "") {
            writer.uint32(10).string(message.appId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAppPurgeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.appId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { appId: isSet(object.appId) ? globalThis.String(object.appId) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.appId !== "") {
            obj.appId = message.appId;
        }
        return obj;
    },
    create(base) {
        return AppPurgeRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAppPurgeRequest();
        message.appId = object.appId ?? "";
        return message;
    },
};
messageTypeRegistry.set(AppPurgeRequest.$type, AppPurgeRequest);
function createBasePurgeResult() {
    return { count: 0 };
}
export const PurgeResult = {
    $type: "devvit.gateway.utils.v1alpha.PurgeResult",
    encode(message, writer = _m0.Writer.create()) {
        if (message.count !== 0) {
            writer.uint32(8).int64(message.count);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePurgeResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.count = longToNumber(reader.int64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { count: isSet(object.count) ? globalThis.Number(object.count) : 0 };
    },
    toJSON(message) {
        const obj = {};
        if (message.count !== 0) {
            obj.count = Math.round(message.count);
        }
        return obj;
    },
    create(base) {
        return PurgeResult.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePurgeResult();
        message.count = object.count ?? 0;
        return message;
    },
};
messageTypeRegistry.set(PurgeResult.$type, PurgeResult);
function createBaseAcceptInviteRequest() {
    return { subredditName: "", appAccountId: "" };
}
export const AcceptInviteRequest = {
    $type: "devvit.gateway.utils.v1alpha.AcceptInviteRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subredditName !== "") {
            writer.uint32(10).string(message.subredditName);
        }
        if (message.appAccountId !== "") {
            writer.uint32(18).string(message.appAccountId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAcceptInviteRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.subredditName = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.appAccountId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            subredditName: isSet(object.subredditName) ? globalThis.String(object.subredditName) : "",
            appAccountId: isSet(object.appAccountId) ? globalThis.String(object.appAccountId) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.subredditName !== "") {
            obj.subredditName = message.subredditName;
        }
        if (message.appAccountId !== "") {
            obj.appAccountId = message.appAccountId;
        }
        return obj;
    },
    create(base) {
        return AcceptInviteRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAcceptInviteRequest();
        message.subredditName = object.subredditName ?? "";
        message.appAccountId = object.appAccountId ?? "";
        return message;
    },
};
messageTypeRegistry.set(AcceptInviteRequest.$type, AcceptInviteRequest);
function createBaseAcceptInviteResponse() {
    return {};
}
export const AcceptInviteResponse = {
    $type: "devvit.gateway.utils.v1alpha.AcceptInviteResponse",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAcceptInviteResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return AcceptInviteResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseAcceptInviteResponse();
        return message;
    },
};
messageTypeRegistry.set(AcceptInviteResponse.$type, AcceptInviteResponse);
export const CacheUtilsServiceName = "devvit.gateway.utils.v1alpha.CacheUtils";
export class CacheUtilsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || CacheUtilsServiceName;
        this.rpc = rpc;
        this.PurgeAccessKey = this.PurgeAccessKey.bind(this);
        this.PurgeInstallations = this.PurgeInstallations.bind(this);
        this.PurgeInstallation = this.PurgeInstallation.bind(this);
        this.PurgeAppVersion = this.PurgeAppVersion.bind(this);
        this.PurgeApp = this.PurgeApp.bind(this);
    }
    PurgeAccessKey(request, metadata) {
        const data = AccessKeyPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeAccessKey", data, metadata);
        return promise.then((data) => PurgeResult.decode(_m0.Reader.create(data)));
    }
    PurgeInstallations(request, metadata) {
        const data = InstallationsPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeInstallations", data, metadata);
        return promise.then((data) => PurgeResult.decode(_m0.Reader.create(data)));
    }
    PurgeInstallation(request, metadata) {
        const data = InstallationPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeInstallation", data, metadata);
        return promise.then((data) => PurgeResult.decode(_m0.Reader.create(data)));
    }
    PurgeAppVersion(request, metadata) {
        const data = AppVersionPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeAppVersion", data, metadata);
        return promise.then((data) => PurgeResult.decode(_m0.Reader.create(data)));
    }
    PurgeApp(request, metadata) {
        const data = AppPurgeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "PurgeApp", data, metadata);
        return promise.then((data) => PurgeResult.decode(_m0.Reader.create(data)));
    }
}
export const CacheUtilsDefinition = {
    name: "CacheUtils",
    fullName: "devvit.gateway.utils.v1alpha.CacheUtils",
    methods: {
        purgeAccessKey: {
            name: "PurgeAccessKey",
            requestType: AccessKeyPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
        purgeInstallations: {
            name: "PurgeInstallations",
            requestType: InstallationsPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
        purgeInstallation: {
            name: "PurgeInstallation",
            requestType: InstallationPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
        purgeAppVersion: {
            name: "PurgeAppVersion",
            requestType: AppVersionPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
        purgeApp: {
            name: "PurgeApp",
            requestType: AppPurgeRequest,
            requestStream: false,
            responseType: PurgeResult,
            responseStream: false,
            options: {},
        },
    },
};
export const ModInviteUtilsServiceName = "devvit.gateway.utils.v1alpha.ModInviteUtils";
export class ModInviteUtilsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || ModInviteUtilsServiceName;
        this.rpc = rpc;
        this.AcceptInvite = this.AcceptInvite.bind(this);
    }
    AcceptInvite(request, metadata) {
        const data = AcceptInviteRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "AcceptInvite", data, metadata);
        return promise.then((data) => AcceptInviteResponse.decode(_m0.Reader.create(data)));
    }
}
export const ModInviteUtilsDefinition = {
    name: "ModInviteUtils",
    fullName: "devvit.gateway.utils.v1alpha.ModInviteUtils",
    methods: {
        acceptInvite: {
            name: "AcceptInvite",
            requestType: AcceptInviteRequest,
            requestStream: false,
            responseType: AcceptInviteResponse,
            responseStream: false,
            options: {},
        },
    },
};
function longToNumber(long) {
    if (long.gt(globalThis.Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
