// @devvit/protos/types/devvit/events/v1alpha/events.js
/**
 * #events.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Timestamp } from '../../../google/protobuf/timestamp.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { CommentV2 } from '../../reddit/v2alpha/commentv2.js';
import { PostV2 } from '../../reddit/v2alpha/postv2.js';
import { SubredditV2 } from '../../reddit/v2alpha/subredditv2.js';
import { UserV2 } from '../../reddit/v2alpha/userv2.js';
export var EventSource;
(function (EventSource) {
    EventSource[EventSource["UNKNOWN_EVENT_SOURCE"] = 0] = "UNKNOWN_EVENT_SOURCE";
    EventSource[EventSource["USER"] = 1] = "USER";
    EventSource[EventSource["ADMIN"] = 2] = "ADMIN";
    EventSource[EventSource["MODERATOR"] = 3] = "MODERATOR";
    EventSource[EventSource["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(EventSource || (EventSource = {}));
export function eventSourceFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN_EVENT_SOURCE":
            return EventSource.UNKNOWN_EVENT_SOURCE;
        case 1:
        case "USER":
            return EventSource.USER;
        case 2:
        case "ADMIN":
            return EventSource.ADMIN;
        case 3:
        case "MODERATOR":
            return EventSource.MODERATOR;
        case -1:
        case "UNRECOGNIZED":
        default:
            return EventSource.UNRECOGNIZED;
    }
}
export function eventSourceToJSON(object) {
    switch (object) {
        case EventSource.UNKNOWN_EVENT_SOURCE:
            return 0;
        case EventSource.USER:
            return 1;
        case EventSource.ADMIN:
            return 2;
        case EventSource.MODERATOR:
            return 3;
        case EventSource.UNRECOGNIZED:
        default:
            return -1;
    }
}
export var DeletionReason;
(function (DeletionReason) {
    DeletionReason[DeletionReason["UNSPECIFIED_DELETION_REASON"] = 0] = "UNSPECIFIED_DELETION_REASON";
    DeletionReason[DeletionReason["SPAM"] = 1] = "SPAM";
    DeletionReason[DeletionReason["LEGAL"] = 2] = "LEGAL";
    DeletionReason[DeletionReason["OTHER"] = 3] = "OTHER";
    DeletionReason[DeletionReason["UNKNOWN"] = 4] = "UNKNOWN";
    DeletionReason[DeletionReason["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(DeletionReason || (DeletionReason = {}));
export function deletionReasonFromJSON(object) {
    switch (object) {
        case 0:
        case "UNSPECIFIED_DELETION_REASON":
            return DeletionReason.UNSPECIFIED_DELETION_REASON;
        case 1:
        case "SPAM":
            return DeletionReason.SPAM;
        case 2:
        case "LEGAL":
            return DeletionReason.LEGAL;
        case 3:
        case "OTHER":
            return DeletionReason.OTHER;
        case 4:
        case "UNKNOWN":
            return DeletionReason.UNKNOWN;
        case -1:
        case "UNRECOGNIZED":
        default:
            return DeletionReason.UNRECOGNIZED;
    }
}
export function deletionReasonToJSON(object) {
    switch (object) {
        case DeletionReason.UNSPECIFIED_DELETION_REASON:
            return 0;
        case DeletionReason.SPAM:
            return 1;
        case DeletionReason.LEGAL:
            return 2;
        case DeletionReason.OTHER:
            return 3;
        case DeletionReason.UNKNOWN:
            return 4;
        case DeletionReason.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBasePostSubmit() {
    return { post: undefined, author: undefined, subreddit: undefined };
}
export const PostSubmit = {
    $type: "devvit.events.v1alpha.PostSubmit",
    encode(message, writer = _m0.Writer.create()) {
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(10).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(18).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePostSubmit();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePostSubmit();
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.PostSubmit", PostSubmit);
function createBasePostCreate() {
    return { post: undefined, author: undefined, subreddit: undefined };
}
export const PostCreate = {
    $type: "devvit.events.v1alpha.PostCreate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(10).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(18).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePostCreate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePostCreate();
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.PostCreate", PostCreate);
function createBasePostUpdate() {
    return { post: undefined, author: undefined, previousBody: "", subreddit: undefined };
}
export const PostUpdate = {
    $type: "devvit.events.v1alpha.PostUpdate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(10).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(18).fork()).ldelim();
        }
        if (message.previousBody !== "") {
            writer.uint32(26).string(message.previousBody);
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePostUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.previousBody = reader.string();
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            previousBody: isSet(object.previousBody) ? String(object.previousBody) : "",
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.previousBody !== undefined && (obj.previousBody = message.previousBody);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePostUpdate();
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.previousBody = object.previousBody ?? "";
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.PostUpdate", PostUpdate);
function createBasePostReport() {
    return { post: undefined, subreddit: undefined, reason: "" };
}
export const PostReport = {
    $type: "devvit.events.v1alpha.PostReport",
    encode(message, writer = _m0.Writer.create()) {
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(10).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        if (message.reason !== "") {
            writer.uint32(26).string(message.reason);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePostReport();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.reason = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
            reason: isSet(object.reason) ? String(object.reason) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        message.reason !== undefined && (obj.reason = message.reason);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePostReport();
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        message.reason = object.reason ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.PostReport", PostReport);
function createBasePostFlairUpdate() {
    return { post: undefined, author: undefined, subreddit: undefined };
}
export const PostFlairUpdate = {
    $type: "devvit.events.v1alpha.PostFlairUpdate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(10).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(18).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePostFlairUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePostFlairUpdate();
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.PostFlairUpdate", PostFlairUpdate);
function createBasePostDelete() {
    return {
        postId: "",
        deletedAt: undefined,
        author: undefined,
        createdAt: undefined,
        source: 0,
        reason: 0,
        subreddit: undefined,
    };
}
export const PostDelete = {
    $type: "devvit.events.v1alpha.PostDelete",
    encode(message, writer = _m0.Writer.create()) {
        if (message.postId !== "") {
            writer.uint32(10).string(message.postId);
        }
        if (message.deletedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.deletedAt), writer.uint32(18).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(26).fork()).ldelim();
        }
        if (message.createdAt !== undefined) {
            Timestamp.encode(toTimestamp(message.createdAt), writer.uint32(34).fork()).ldelim();
        }
        if (message.source !== 0) {
            writer.uint32(40).int32(message.source);
        }
        if (message.reason !== 0) {
            writer.uint32(48).int32(message.reason);
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePostDelete();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.postId = reader.string();
                    break;
                case 2:
                    message.deletedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.createdAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.source = reader.int32();
                    break;
                case 6:
                    message.reason = reader.int32();
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            postId: isSet(object.postId) ? String(object.postId) : "",
            deletedAt: isSet(object.deletedAt) ? fromJsonTimestamp(object.deletedAt) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            createdAt: isSet(object.createdAt) ? fromJsonTimestamp(object.createdAt) : undefined,
            source: isSet(object.source) ? eventSourceFromJSON(object.source) : 0,
            reason: isSet(object.reason) ? deletionReasonFromJSON(object.reason) : 0,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.postId !== undefined && (obj.postId = message.postId);
        message.deletedAt !== undefined && (obj.deletedAt = message.deletedAt.toISOString());
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.createdAt !== undefined && (obj.createdAt = message.createdAt.toISOString());
        message.source !== undefined && (obj.source = eventSourceToJSON(message.source));
        message.reason !== undefined && (obj.reason = deletionReasonToJSON(message.reason));
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePostDelete();
        message.postId = object.postId ?? "";
        message.deletedAt = object.deletedAt ?? undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.createdAt = object.createdAt ?? undefined;
        message.source = object.source ?? 0;
        message.reason = object.reason ?? 0;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.PostDelete", PostDelete);
function createBasePostApprove() {
    return { post: undefined, approvedAt: undefined, author: undefined, source: 0, subreddit: undefined };
}
export const PostApprove = {
    $type: "devvit.events.v1alpha.PostApprove",
    encode(message, writer = _m0.Writer.create()) {
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(10).fork()).ldelim();
        }
        if (message.approvedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.approvedAt), writer.uint32(18).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(26).fork()).ldelim();
        }
        if (message.source !== 0) {
            writer.uint32(40).int32(message.source);
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePostApprove();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.approvedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.source = reader.int32();
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            approvedAt: isSet(object.approvedAt) ? fromJsonTimestamp(object.approvedAt) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            source: isSet(object.source) ? eventSourceFromJSON(object.source) : 0,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.approvedAt !== undefined && (obj.approvedAt = message.approvedAt.toISOString());
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.source !== undefined && (obj.source = eventSourceToJSON(message.source));
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePostApprove();
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.approvedAt = object.approvedAt ?? undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.source = object.source ?? 0;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.PostApprove", PostApprove);
function createBaseCommentApprove() {
    return {
        comment: undefined,
        post: undefined,
        approvedAt: undefined,
        author: undefined,
        source: 0,
        subreddit: undefined,
    };
}
export const CommentApprove = {
    $type: "devvit.events.v1alpha.CommentApprove",
    encode(message, writer = _m0.Writer.create()) {
        if (message.comment !== undefined) {
            CommentV2.encode(message.comment, writer.uint32(10).fork()).ldelim();
        }
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(18).fork()).ldelim();
        }
        if (message.approvedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.approvedAt), writer.uint32(26).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(34).fork()).ldelim();
        }
        if (message.source !== 0) {
            writer.uint32(48).int32(message.source);
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommentApprove();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.comment = CommentV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.approvedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.source = reader.int32();
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            comment: isSet(object.comment) ? CommentV2.fromJSON(object.comment) : undefined,
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            approvedAt: isSet(object.approvedAt) ? fromJsonTimestamp(object.approvedAt) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            source: isSet(object.source) ? eventSourceFromJSON(object.source) : 0,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.comment !== undefined && (obj.comment = message.comment ? CommentV2.toJSON(message.comment) : undefined);
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.approvedAt !== undefined && (obj.approvedAt = message.approvedAt.toISOString());
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.source !== undefined && (obj.source = eventSourceToJSON(message.source));
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCommentApprove();
        message.comment = (object.comment !== undefined && object.comment !== null)
            ? CommentV2.fromPartial(object.comment)
            : undefined;
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.approvedAt = object.approvedAt ?? undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.source = object.source ?? 0;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.CommentApprove", CommentApprove);
function createBaseCommentSubmit() {
    return { comment: undefined, author: undefined, post: undefined, subreddit: undefined };
}
export const CommentSubmit = {
    $type: "devvit.events.v1alpha.CommentSubmit",
    encode(message, writer = _m0.Writer.create()) {
        if (message.comment !== undefined) {
            CommentV2.encode(message.comment, writer.uint32(10).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(18).fork()).ldelim();
        }
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(26).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommentSubmit();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.comment = CommentV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            comment: isSet(object.comment) ? CommentV2.fromJSON(object.comment) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.comment !== undefined && (obj.comment = message.comment ? CommentV2.toJSON(message.comment) : undefined);
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCommentSubmit();
        message.comment = (object.comment !== undefined && object.comment !== null)
            ? CommentV2.fromPartial(object.comment)
            : undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.CommentSubmit", CommentSubmit);
function createBaseCommentCreate() {
    return { comment: undefined, author: undefined, post: undefined, subreddit: undefined };
}
export const CommentCreate = {
    $type: "devvit.events.v1alpha.CommentCreate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.comment !== undefined) {
            CommentV2.encode(message.comment, writer.uint32(10).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(18).fork()).ldelim();
        }
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(26).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommentCreate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.comment = CommentV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            comment: isSet(object.comment) ? CommentV2.fromJSON(object.comment) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.comment !== undefined && (obj.comment = message.comment ? CommentV2.toJSON(message.comment) : undefined);
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCommentCreate();
        message.comment = (object.comment !== undefined && object.comment !== null)
            ? CommentV2.fromPartial(object.comment)
            : undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.CommentCreate", CommentCreate);
function createBaseCommentUpdate() {
    return { comment: undefined, author: undefined, post: undefined, previousBody: "", subreddit: undefined };
}
export const CommentUpdate = {
    $type: "devvit.events.v1alpha.CommentUpdate",
    encode(message, writer = _m0.Writer.create()) {
        if (message.comment !== undefined) {
            CommentV2.encode(message.comment, writer.uint32(10).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(18).fork()).ldelim();
        }
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(26).fork()).ldelim();
        }
        if (message.previousBody !== "") {
            writer.uint32(34).string(message.previousBody);
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommentUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.comment = CommentV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.previousBody = reader.string();
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            comment: isSet(object.comment) ? CommentV2.fromJSON(object.comment) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            previousBody: isSet(object.previousBody) ? String(object.previousBody) : "",
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.comment !== undefined && (obj.comment = message.comment ? CommentV2.toJSON(message.comment) : undefined);
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.previousBody !== undefined && (obj.previousBody = message.previousBody);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCommentUpdate();
        message.comment = (object.comment !== undefined && object.comment !== null)
            ? CommentV2.fromPartial(object.comment)
            : undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.previousBody = object.previousBody ?? "";
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.CommentUpdate", CommentUpdate);
function createBaseCommentReport() {
    return { comment: undefined, subreddit: undefined, reason: "" };
}
export const CommentReport = {
    $type: "devvit.events.v1alpha.CommentReport",
    encode(message, writer = _m0.Writer.create()) {
        if (message.comment !== undefined) {
            CommentV2.encode(message.comment, writer.uint32(10).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        if (message.reason !== "") {
            writer.uint32(26).string(message.reason);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommentReport();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.comment = CommentV2.decode(reader, reader.uint32());
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.reason = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            comment: isSet(object.comment) ? CommentV2.fromJSON(object.comment) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
            reason: isSet(object.reason) ? String(object.reason) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.comment !== undefined && (obj.comment = message.comment ? CommentV2.toJSON(message.comment) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        message.reason !== undefined && (obj.reason = message.reason);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCommentReport();
        message.comment = (object.comment !== undefined && object.comment !== null)
            ? CommentV2.fromPartial(object.comment)
            : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        message.reason = object.reason ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.CommentReport", CommentReport);
function createBaseCommentDelete() {
    return {
        commentId: "",
        postId: "",
        parentId: "",
        deletedAt: undefined,
        author: undefined,
        createdAt: undefined,
        source: 0,
        reason: 0,
        subreddit: undefined,
    };
}
export const CommentDelete = {
    $type: "devvit.events.v1alpha.CommentDelete",
    encode(message, writer = _m0.Writer.create()) {
        if (message.commentId !== "") {
            writer.uint32(10).string(message.commentId);
        }
        if (message.postId !== "") {
            writer.uint32(18).string(message.postId);
        }
        if (message.parentId !== "") {
            writer.uint32(26).string(message.parentId);
        }
        if (message.deletedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.deletedAt), writer.uint32(34).fork()).ldelim();
        }
        if (message.author !== undefined) {
            UserV2.encode(message.author, writer.uint32(42).fork()).ldelim();
        }
        if (message.createdAt !== undefined) {
            Timestamp.encode(toTimestamp(message.createdAt), writer.uint32(50).fork()).ldelim();
        }
        if (message.source !== 0) {
            writer.uint32(56).int32(message.source);
        }
        if (message.reason !== 0) {
            writer.uint32(64).int32(message.reason);
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCommentDelete();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.commentId = reader.string();
                    break;
                case 2:
                    message.postId = reader.string();
                    break;
                case 3:
                    message.parentId = reader.string();
                    break;
                case 4:
                    message.deletedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.author = UserV2.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.createdAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 7:
                    message.source = reader.int32();
                    break;
                case 8:
                    message.reason = reader.int32();
                    break;
                case 10:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            commentId: isSet(object.commentId) ? String(object.commentId) : "",
            postId: isSet(object.postId) ? String(object.postId) : "",
            parentId: isSet(object.parentId) ? String(object.parentId) : "",
            deletedAt: isSet(object.deletedAt) ? fromJsonTimestamp(object.deletedAt) : undefined,
            author: isSet(object.author) ? UserV2.fromJSON(object.author) : undefined,
            createdAt: isSet(object.createdAt) ? fromJsonTimestamp(object.createdAt) : undefined,
            source: isSet(object.source) ? eventSourceFromJSON(object.source) : 0,
            reason: isSet(object.reason) ? deletionReasonFromJSON(object.reason) : 0,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.commentId !== undefined && (obj.commentId = message.commentId);
        message.postId !== undefined && (obj.postId = message.postId);
        message.parentId !== undefined && (obj.parentId = message.parentId);
        message.deletedAt !== undefined && (obj.deletedAt = message.deletedAt.toISOString());
        message.author !== undefined && (obj.author = message.author ? UserV2.toJSON(message.author) : undefined);
        message.createdAt !== undefined && (obj.createdAt = message.createdAt.toISOString());
        message.source !== undefined && (obj.source = eventSourceToJSON(message.source));
        message.reason !== undefined && (obj.reason = deletionReasonToJSON(message.reason));
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseCommentDelete();
        message.commentId = object.commentId ?? "";
        message.postId = object.postId ?? "";
        message.parentId = object.parentId ?? "";
        message.deletedAt = object.deletedAt ?? undefined;
        message.author = (object.author !== undefined && object.author !== null)
            ? UserV2.fromPartial(object.author)
            : undefined;
        message.createdAt = object.createdAt ?? undefined;
        message.source = object.source ?? 0;
        message.reason = object.reason ?? 0;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.CommentDelete", CommentDelete);
function createBaseSubredditSubscribe() {
    return { subreddit: undefined, subscriber: undefined };
}
export const SubredditSubscribe = {
    $type: "devvit.events.v1alpha.SubredditSubscribe",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(10).fork()).ldelim();
        }
        if (message.subscriber !== undefined) {
            UserV2.encode(message.subscriber, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubredditSubscribe();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.subscriber = UserV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
            subscriber: isSet(object.subscriber) ? UserV2.fromJSON(object.subscriber) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        message.subscriber !== undefined &&
            (obj.subscriber = message.subscriber ? UserV2.toJSON(message.subscriber) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseSubredditSubscribe();
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        message.subscriber = (object.subscriber !== undefined && object.subscriber !== null)
            ? UserV2.fromPartial(object.subscriber)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.SubredditSubscribe", SubredditSubscribe);
function createBaseAppInstall() {
    return { subreddit: undefined, installer: undefined };
}
export const AppInstall = {
    $type: "devvit.events.v1alpha.AppInstall",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(10).fork()).ldelim();
        }
        if (message.installer !== undefined) {
            UserV2.encode(message.installer, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAppInstall();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.installer = UserV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
            installer: isSet(object.installer) ? UserV2.fromJSON(object.installer) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        message.installer !== undefined &&
            (obj.installer = message.installer ? UserV2.toJSON(message.installer) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAppInstall();
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        message.installer = (object.installer !== undefined && object.installer !== null)
            ? UserV2.fromPartial(object.installer)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.AppInstall", AppInstall);
function createBaseAppUpgrade() {
    return { subreddit: undefined, installer: undefined };
}
export const AppUpgrade = {
    $type: "devvit.events.v1alpha.AppUpgrade",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(10).fork()).ldelim();
        }
        if (message.installer !== undefined) {
            UserV2.encode(message.installer, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAppUpgrade();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.installer = UserV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
            installer: isSet(object.installer) ? UserV2.fromJSON(object.installer) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        message.installer !== undefined &&
            (obj.installer = message.installer ? UserV2.toJSON(message.installer) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAppUpgrade();
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        message.installer = (object.installer !== undefined && object.installer !== null)
            ? UserV2.fromPartial(object.installer)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.AppUpgrade", AppUpgrade);
function createBaseAccountDelete() {
    return { userId: "", deletedAt: undefined, user: undefined };
}
export const AccountDelete = {
    $type: "devvit.events.v1alpha.AccountDelete",
    encode(message, writer = _m0.Writer.create()) {
        if (message.userId !== "") {
            writer.uint32(10).string(message.userId);
        }
        if (message.deletedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.deletedAt), writer.uint32(18).fork()).ldelim();
        }
        if (message.user !== undefined) {
            UserV2.encode(message.user, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccountDelete();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.userId = reader.string();
                    break;
                case 2:
                    message.deletedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 3:
                    message.user = UserV2.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            userId: isSet(object.userId) ? String(object.userId) : "",
            deletedAt: isSet(object.deletedAt) ? fromJsonTimestamp(object.deletedAt) : undefined,
            user: isSet(object.user) ? UserV2.fromJSON(object.user) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.userId !== undefined && (obj.userId = message.userId);
        message.deletedAt !== undefined && (obj.deletedAt = message.deletedAt.toISOString());
        message.user !== undefined && (obj.user = message.user ? UserV2.toJSON(message.user) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseAccountDelete();
        message.userId = object.userId ?? "";
        message.deletedAt = object.deletedAt ?? undefined;
        message.user = (object.user !== undefined && object.user !== null) ? UserV2.fromPartial(object.user) : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.AccountDelete", AccountDelete);
function createBaseVote() {
    return { updatedAt: undefined, comment: undefined, post: undefined, subreddit: undefined, upvoteRatio: 0, score: 0 };
}
export const Vote = {
    $type: "devvit.events.v1alpha.Vote",
    encode(message, writer = _m0.Writer.create()) {
        if (message.updatedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.updatedAt), writer.uint32(34).fork()).ldelim();
        }
        if (message.comment !== undefined) {
            CommentV2.encode(message.comment, writer.uint32(42).fork()).ldelim();
        }
        if (message.post !== undefined) {
            PostV2.encode(message.post, writer.uint32(50).fork()).ldelim();
        }
        if (message.subreddit !== undefined) {
            SubredditV2.encode(message.subreddit, writer.uint32(58).fork()).ldelim();
        }
        if (message.upvoteRatio !== 0) {
            writer.uint32(69).float(message.upvoteRatio);
        }
        if (message.score !== 0) {
            writer.uint32(72).int32(message.score);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseVote();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 4:
                    message.updatedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.comment = CommentV2.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.post = PostV2.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.subreddit = SubredditV2.decode(reader, reader.uint32());
                    break;
                case 8:
                    message.upvoteRatio = reader.float();
                    break;
                case 9:
                    message.score = reader.int32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            updatedAt: isSet(object.updatedAt) ? fromJsonTimestamp(object.updatedAt) : undefined,
            comment: isSet(object.comment) ? CommentV2.fromJSON(object.comment) : undefined,
            post: isSet(object.post) ? PostV2.fromJSON(object.post) : undefined,
            subreddit: isSet(object.subreddit) ? SubredditV2.fromJSON(object.subreddit) : undefined,
            upvoteRatio: isSet(object.upvoteRatio) ? Number(object.upvoteRatio) : 0,
            score: isSet(object.score) ? Number(object.score) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.updatedAt !== undefined && (obj.updatedAt = message.updatedAt.toISOString());
        message.comment !== undefined && (obj.comment = message.comment ? CommentV2.toJSON(message.comment) : undefined);
        message.post !== undefined && (obj.post = message.post ? PostV2.toJSON(message.post) : undefined);
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? SubredditV2.toJSON(message.subreddit) : undefined);
        message.upvoteRatio !== undefined && (obj.upvoteRatio = message.upvoteRatio);
        message.score !== undefined && (obj.score = Math.round(message.score));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseVote();
        message.updatedAt = object.updatedAt ?? undefined;
        message.comment = (object.comment !== undefined && object.comment !== null)
            ? CommentV2.fromPartial(object.comment)
            : undefined;
        message.post = (object.post !== undefined && object.post !== null) ? PostV2.fromPartial(object.post) : undefined;
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? SubredditV2.fromPartial(object.subreddit)
            : undefined;
        message.upvoteRatio = object.upvoteRatio ?? 0;
        message.score = object.score ?? 0;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.Vote", Vote);
function toTimestamp(date) {
    const seconds = date.getTime() / 1000;
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = t.seconds * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
