// @devvit/protos/types/devvit/events/v1alpha/wqcleanup.js
/**
 * #wqcleanup.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import Long from "long";
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseWqCleanup() {
    return { cursor: 0 };
}
export const WqCleanup = {
    $type: "devvit.events.v1alpha.WqCleanup",
    encode(message, writer = _m0.Writer.create()) {
        if (message.cursor !== 0) {
            writer.uint32(8).uint64(message.cursor);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWqCleanup();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.cursor = longToNumber(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { cursor: isSet(object.cursor) ? Number(object.cursor) : 0 };
    },
    toJSON(message) {
        const obj = {};
        message.cursor !== undefined && (obj.cursor = Math.round(message.cursor));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseWqCleanup();
        message.cursor = object.cursor ?? 0;
        return message;
    },
};
messageTypeRegistry.set("devvit.events.v1alpha.WqCleanup", WqCleanup);
var globalThis = (() => {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function longToNumber(long) {
    if (long.gt(Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
