// @devvit/protos/types/devvit/plugin/devserver/testcases.js
/**
 * #testcases.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseTestCasesList() {
    return { actorName: "", testCases: [] };
}
export const TestCasesList = {
    $type: "devvit.plugin.devserver.TestCasesList",
    encode(message, writer = _m0.Writer.create()) {
        if (message.actorName !== "") {
            writer.uint32(10).string(message.actorName);
        }
        for (const v of message.testCases) {
            TestCase.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestCasesList();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.actorName = reader.string();
                    break;
                case 2:
                    message.testCases.push(TestCase.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            actorName: isSet(object.actorName) ? String(object.actorName) : "",
            testCases: Array.isArray(object?.testCases) ? object.testCases.map((e) => TestCase.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.actorName !== undefined && (obj.actorName = message.actorName);
        if (message.testCases) {
            obj.testCases = message.testCases.map((e) => e ? TestCase.toJSON(e) : undefined);
        }
        else {
            obj.testCases = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseTestCasesList();
        message.actorName = object.actorName ?? "";
        message.testCases = object.testCases?.map((e) => TestCase.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.TestCasesList", TestCasesList);
function createBaseTestCase() {
    return { methodName: "", testName: "", testInput: "" };
}
export const TestCase = {
    $type: "devvit.plugin.devserver.TestCase",
    encode(message, writer = _m0.Writer.create()) {
        if (message.methodName !== "") {
            writer.uint32(10).string(message.methodName);
        }
        if (message.testName !== "") {
            writer.uint32(18).string(message.testName);
        }
        if (message.testInput !== "") {
            writer.uint32(26).string(message.testInput);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestCase();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.methodName = reader.string();
                    break;
                case 2:
                    message.testName = reader.string();
                    break;
                case 3:
                    message.testInput = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            methodName: isSet(object.methodName) ? String(object.methodName) : "",
            testName: isSet(object.testName) ? String(object.testName) : "",
            testInput: isSet(object.testInput) ? String(object.testInput) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.methodName !== undefined && (obj.methodName = message.methodName);
        message.testName !== undefined && (obj.testName = message.testName);
        message.testInput !== undefined && (obj.testInput = message.testInput);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseTestCase();
        message.methodName = object.methodName ?? "";
        message.testName = object.testName ?? "";
        message.testInput = object.testInput ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.devserver.TestCase", TestCase);
function isSet(value) {
    return value !== null && value !== undefined;
}
