// @devvit/protos/types/devvit/actor/settings/v1alpha/shared.js
/**
 * #shared.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
import { Form } from '../../../ui/form_builder/v1alpha/form.js';
import { FormFieldValue } from '../../../ui/form_builder/v1alpha/value.js';
function createBaseGetFieldsRequest() {
    return { editing: false };
}
export const GetFieldsRequest = {
    $type: "devvit.actor.settings.v1alpha.GetFieldsRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.editing === true) {
            writer.uint32(8).bool(message.editing);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetFieldsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.editing = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { editing: isSet(object.editing) ? Boolean(object.editing) : false };
    },
    toJSON(message) {
        const obj = {};
        message.editing !== undefined && (obj.editing = message.editing);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetFieldsRequest();
        message.editing = object.editing ?? false;
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.settings.v1alpha.GetFieldsRequest", GetFieldsRequest);
function createBaseGetFieldsResponse() {
    return { fields: undefined };
}
export const GetFieldsResponse = {
    $type: "devvit.actor.settings.v1alpha.GetFieldsResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.fields !== undefined) {
            Form.encode(message.fields, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetFieldsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.fields = Form.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { fields: isSet(object.fields) ? Form.fromJSON(object.fields) : undefined };
    },
    toJSON(message) {
        const obj = {};
        message.fields !== undefined && (obj.fields = message.fields ? Form.toJSON(message.fields) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseGetFieldsResponse();
        message.fields = (object.fields !== undefined && object.fields !== null)
            ? Form.fromPartial(object.fields)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.settings.v1alpha.GetFieldsResponse", GetFieldsResponse);
function createBaseValidateFormRequest() {
    return { editing: false, fieldValues: {} };
}
export const ValidateFormRequest = {
    $type: "devvit.actor.settings.v1alpha.ValidateFormRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.editing === true) {
            writer.uint32(8).bool(message.editing);
        }
        Object.entries(message.fieldValues).forEach(([key, value]) => {
            ValidateFormRequest_FieldValuesEntry.encode({ key: key, value }, writer.uint32(18).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidateFormRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.editing = reader.bool();
                    break;
                case 2:
                    const entry2 = ValidateFormRequest_FieldValuesEntry.decode(reader, reader.uint32());
                    if (entry2.value !== undefined) {
                        message.fieldValues[entry2.key] = entry2.value;
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            editing: isSet(object.editing) ? Boolean(object.editing) : false,
            fieldValues: isObject(object.fieldValues)
                ? Object.entries(object.fieldValues).reduce((acc, [key, value]) => {
                    acc[key] = FormFieldValue.fromJSON(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        message.editing !== undefined && (obj.editing = message.editing);
        obj.fieldValues = {};
        if (message.fieldValues) {
            Object.entries(message.fieldValues).forEach(([k, v]) => {
                obj.fieldValues[k] = FormFieldValue.toJSON(v);
            });
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseValidateFormRequest();
        message.editing = object.editing ?? false;
        message.fieldValues = Object.entries(object.fieldValues ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = FormFieldValue.fromPartial(value);
            }
            return acc;
        }, {});
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.settings.v1alpha.ValidateFormRequest", ValidateFormRequest);
function createBaseValidateFormRequest_FieldValuesEntry() {
    return { key: "", value: undefined };
}
export const ValidateFormRequest_FieldValuesEntry = {
    $type: "devvit.actor.settings.v1alpha.ValidateFormRequest.FieldValuesEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== undefined) {
            FormFieldValue.encode(message.value, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidateFormRequest_FieldValuesEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.string();
                    break;
                case 2:
                    message.value = FormFieldValue.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? String(object.key) : "",
            value: isSet(object.value) ? FormFieldValue.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined && (obj.value = message.value ? FormFieldValue.toJSON(message.value) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseValidateFormRequest_FieldValuesEntry();
        message.key = object.key ?? "";
        message.value = (object.value !== undefined && object.value !== null)
            ? FormFieldValue.fromPartial(object.value)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.settings.v1alpha.ValidateFormRequest.FieldValuesEntry", ValidateFormRequest_FieldValuesEntry);
function createBaseValidateFormResponse() {
    return { success: false, errors: {} };
}
export const ValidateFormResponse = {
    $type: "devvit.actor.settings.v1alpha.ValidateFormResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.success === true) {
            writer.uint32(8).bool(message.success);
        }
        Object.entries(message.errors).forEach(([key, value]) => {
            ValidateFormResponse_ErrorsEntry.encode({ key: key, value }, writer.uint32(18).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidateFormResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.success = reader.bool();
                    break;
                case 2:
                    const entry2 = ValidateFormResponse_ErrorsEntry.decode(reader, reader.uint32());
                    if (entry2.value !== undefined) {
                        message.errors[entry2.key] = entry2.value;
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? Boolean(object.success) : false,
            errors: isObject(object.errors)
                ? Object.entries(object.errors).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        message.success !== undefined && (obj.success = message.success);
        obj.errors = {};
        if (message.errors) {
            Object.entries(message.errors).forEach(([k, v]) => {
                obj.errors[k] = v;
            });
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseValidateFormResponse();
        message.success = object.success ?? false;
        message.errors = Object.entries(object.errors ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = String(value);
            }
            return acc;
        }, {});
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.settings.v1alpha.ValidateFormResponse", ValidateFormResponse);
function createBaseValidateFormResponse_ErrorsEntry() {
    return { key: "", value: "" };
}
export const ValidateFormResponse_ErrorsEntry = {
    $type: "devvit.actor.settings.v1alpha.ValidateFormResponse.ErrorsEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidateFormResponse_ErrorsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.string();
                    break;
                case 2:
                    message.value = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { key: isSet(object.key) ? String(object.key) : "", value: isSet(object.value) ? String(object.value) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined && (obj.value = message.value);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseValidateFormResponse_ErrorsEntry();
        message.key = object.key ?? "";
        message.value = object.value ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.settings.v1alpha.ValidateFormResponse.ErrorsEntry", ValidateFormResponse_ErrorsEntry);
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
