// @devvit/protos/types/devvit/reddit/v2alpha/userv2.js
/**
 * #userv2.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseUserV2() {
    return { id: "", name: "", isGold: false, snoovatarImage: "", url: "" };
}
export const UserV2 = {
    $type: "devvit.reddit.v2alpha.UserV2",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.isGold === true) {
            writer.uint32(24).bool(message.isGold);
        }
        if (message.snoovatarImage !== "") {
            writer.uint32(34).string(message.snoovatarImage);
        }
        if (message.url !== "") {
            writer.uint32(42).string(message.url);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUserV2();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.string();
                    break;
                case 2:
                    message.name = reader.string();
                    break;
                case 3:
                    message.isGold = reader.bool();
                    break;
                case 4:
                    message.snoovatarImage = reader.string();
                    break;
                case 5:
                    message.url = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? String(object.id) : "",
            name: isSet(object.name) ? String(object.name) : "",
            isGold: isSet(object.isGold) ? Boolean(object.isGold) : false,
            snoovatarImage: isSet(object.snoovatarImage) ? String(object.snoovatarImage) : "",
            url: isSet(object.url) ? String(object.url) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        message.name !== undefined && (obj.name = message.name);
        message.isGold !== undefined && (obj.isGold = message.isGold);
        message.snoovatarImage !== undefined && (obj.snoovatarImage = message.snoovatarImage);
        message.url !== undefined && (obj.url = message.url);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseUserV2();
        message.id = object.id ?? "";
        message.name = object.name ?? "";
        message.isGold = object.isGold ?? false;
        message.snoovatarImage = object.snoovatarImage ?? "";
        message.url = object.url ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.reddit.v2alpha.UserV2", UserV2);
function isSet(value) {
    return value !== null && value !== undefined;
}
