// @devvit/runtime-lite/types/Commands.js
export var ActorCommand;
(function (ActorCommand) {
    /**
     * Collect and post app console logs in command responses. For a sandboxed
     * app, this would be every console.log() call made in the VM. When disabled,
     * logs are present but empty.
     */
    ActorCommand["ENABLE_APP_LOGGING"] = "enableAppLogging";
    /** Enable runtime debug logs. */
    ActorCommand["ENABLE_DEBUG_LOGGING"] = "enableDebugLogging";
    ActorCommand["ENABLE_SANDBOX"] = "enableSandbox";
    ActorCommand["ENABLE_PLUGINS"] = "enablePlugins";
    ActorCommand["LOAD_BUNDLE"] = "loadBundle";
    ActorCommand["CALL"] = "call";
    ActorCommand["PLUGIN_RESPONSE"] = "pluginResponse";
    ActorCommand["QUIT"] = "quit";
})(ActorCommand || (ActorCommand = {}));
export var ActorResponse;
(function (ActorResponse) {
    ActorResponse["READY"] = "ready";
    ActorResponse["SANDBOX_LOADED"] = "sandboxLoaded";
    ActorResponse["PLUGINS_CONFIGURED"] = "pluginsConfigured";
    ActorResponse["BUNDLE_LOADED"] = "bundleLoaded";
    ActorResponse["BUNDLE_LOAD_FAILED"] = "bundleLoadFailed";
    ActorResponse["RETURN"] = "return";
    ActorResponse["ERROR"] = "error";
    ActorResponse["CIRCUIT_BREAKER"] = "circuitBreaker";
    ActorResponse["PLUGIN_CALL"] = "pluginCall";
    /**
     * An un-awaited promise rejected. This response has no originating request.
     */
    ActorResponse["UNHANDLED_REJECTION"] = "unhandledRejection";
})(ActorResponse || (ActorResponse = {}));
