// @devvit/protos/types/devvit/ui/effects/v1alpha/reload_part.js
/**
 * #reload_part.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../../typeRegistry.js';
function createBaseReloadPartEffect() {
    return { subreddit: undefined, post: undefined, comment: undefined };
}
export const ReloadPartEffect = {
    $type: "devvit.ui.effects.v1alpha.ReloadPartEffect",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subreddit !== undefined) {
            ReloadPartEffect_Subreddit.encode(message.subreddit, writer.uint32(10).fork()).ldelim();
        }
        if (message.post !== undefined) {
            ReloadPartEffect_Post.encode(message.post, writer.uint32(18).fork()).ldelim();
        }
        if (message.comment !== undefined) {
            ReloadPartEffect_Comment.encode(message.comment, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReloadPartEffect();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subreddit = ReloadPartEffect_Subreddit.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.post = ReloadPartEffect_Post.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.comment = ReloadPartEffect_Comment.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subreddit: isSet(object.subreddit) ? ReloadPartEffect_Subreddit.fromJSON(object.subreddit) : undefined,
            post: isSet(object.post) ? ReloadPartEffect_Post.fromJSON(object.post) : undefined,
            comment: isSet(object.comment) ? ReloadPartEffect_Comment.fromJSON(object.comment) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.subreddit !== undefined &&
            (obj.subreddit = message.subreddit ? ReloadPartEffect_Subreddit.toJSON(message.subreddit) : undefined);
        message.post !== undefined && (obj.post = message.post ? ReloadPartEffect_Post.toJSON(message.post) : undefined);
        message.comment !== undefined &&
            (obj.comment = message.comment ? ReloadPartEffect_Comment.toJSON(message.comment) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseReloadPartEffect();
        message.subreddit = (object.subreddit !== undefined && object.subreddit !== null)
            ? ReloadPartEffect_Subreddit.fromPartial(object.subreddit)
            : undefined;
        message.post = (object.post !== undefined && object.post !== null)
            ? ReloadPartEffect_Post.fromPartial(object.post)
            : undefined;
        message.comment = (object.comment !== undefined && object.comment !== null)
            ? ReloadPartEffect_Comment.fromPartial(object.comment)
            : undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.ReloadPartEffect", ReloadPartEffect);
function createBaseReloadPartEffect_Subreddit() {
    return { subredditId: "" };
}
export const ReloadPartEffect_Subreddit = {
    $type: "devvit.ui.effects.v1alpha.ReloadPartEffect.Subreddit",
    encode(message, writer = _m0.Writer.create()) {
        if (message.subredditId !== "") {
            writer.uint32(10).string(message.subredditId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReloadPartEffect_Subreddit();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subredditId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { subredditId: isSet(object.subredditId) ? String(object.subredditId) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.subredditId !== undefined && (obj.subredditId = message.subredditId);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseReloadPartEffect_Subreddit();
        message.subredditId = object.subredditId ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.ReloadPartEffect.Subreddit", ReloadPartEffect_Subreddit);
function createBaseReloadPartEffect_Post() {
    return { postId: "", body: undefined, comments: undefined };
}
export const ReloadPartEffect_Post = {
    $type: "devvit.ui.effects.v1alpha.ReloadPartEffect.Post",
    encode(message, writer = _m0.Writer.create()) {
        if (message.postId !== "") {
            writer.uint32(10).string(message.postId);
        }
        if (message.body !== undefined) {
            writer.uint32(16).bool(message.body);
        }
        if (message.comments !== undefined) {
            writer.uint32(24).bool(message.comments);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReloadPartEffect_Post();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.postId = reader.string();
                    break;
                case 2:
                    message.body = reader.bool();
                    break;
                case 3:
                    message.comments = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            postId: isSet(object.postId) ? String(object.postId) : "",
            body: isSet(object.body) ? Boolean(object.body) : undefined,
            comments: isSet(object.comments) ? Boolean(object.comments) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.postId !== undefined && (obj.postId = message.postId);
        message.body !== undefined && (obj.body = message.body);
        message.comments !== undefined && (obj.comments = message.comments);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseReloadPartEffect_Post();
        message.postId = object.postId ?? "";
        message.body = object.body ?? undefined;
        message.comments = object.comments ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.ReloadPartEffect.Post", ReloadPartEffect_Post);
function createBaseReloadPartEffect_Comment() {
    return { postId: "", commentId: "", replies: undefined };
}
export const ReloadPartEffect_Comment = {
    $type: "devvit.ui.effects.v1alpha.ReloadPartEffect.Comment",
    encode(message, writer = _m0.Writer.create()) {
        if (message.postId !== "") {
            writer.uint32(10).string(message.postId);
        }
        if (message.commentId !== "") {
            writer.uint32(18).string(message.commentId);
        }
        if (message.replies !== undefined) {
            writer.uint32(24).bool(message.replies);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReloadPartEffect_Comment();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.postId = reader.string();
                    break;
                case 2:
                    message.commentId = reader.string();
                    break;
                case 3:
                    message.replies = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            postId: isSet(object.postId) ? String(object.postId) : "",
            commentId: isSet(object.commentId) ? String(object.commentId) : "",
            replies: isSet(object.replies) ? Boolean(object.replies) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.postId !== undefined && (obj.postId = message.postId);
        message.commentId !== undefined && (obj.commentId = message.commentId);
        message.replies !== undefined && (obj.replies = message.replies);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseReloadPartEffect_Comment();
        message.postId = object.postId ?? "";
        message.commentId = object.commentId ?? "";
        message.replies = object.replies ?? undefined;
        return message;
    },
};
messageTypeRegistry.set("devvit.ui.effects.v1alpha.ReloadPartEffect.Comment", ReloadPartEffect_Comment);
function isSet(value) {
    return value !== null && value !== undefined;
}
