// @devvit/protos/types/devvit/dev_portal/app/categories/categories.js
/**
 * #categories.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
/** Contains every valid category an App can have. */
export var Categories;
(function (Categories) {
    Categories[Categories["MODERATION"] = 0] = "MODERATION";
    Categories[Categories["CUSTOM_POSTS"] = 1] = "CUSTOM_POSTS";
    Categories[Categories["UTILITIES"] = 2] = "UTILITIES";
    Categories[Categories["OTHER_SITES"] = 3] = "OTHER_SITES";
    Categories[Categories["FUNNY"] = 4] = "FUNNY";
    Categories[Categories["WEIRD"] = 5] = "WEIRD";
    Categories[Categories["AWW"] = 6] = "AWW";
    Categories[Categories["LEARNING"] = 7] = "LEARNING";
    Categories[Categories["FINANCIAL"] = 8] = "FINANCIAL";
    Categories[Categories["SPORTS"] = 9] = "SPORTS";
    Categories[Categories["GAMES"] = 10] = "GAMES";
    Categories[Categories["MOVIES"] = 11] = "MOVIES";
    Categories[Categories["MUSIC"] = 12] = "MUSIC";
    Categories[Categories["TV"] = 13] = "TV";
    Categories[Categories["BOOKS"] = 14] = "BOOKS";
    Categories[Categories["ANIME"] = 15] = "ANIME";
    Categories[Categories["PROGRAMMING"] = 16] = "PROGRAMMING";
    Categories[Categories["COOKING"] = 17] = "COOKING";
    Categories[Categories["FITNESS"] = 18] = "FITNESS";
    Categories[Categories["TRAVEL"] = 19] = "TRAVEL";
    Categories[Categories["CONTEXT_MENU"] = 20] = "CONTEXT_MENU";
    Categories[Categories["MOD_CONTEXT"] = 21] = "MOD_CONTEXT";
    Categories[Categories["FILTER"] = 22] = "FILTER";
    Categories[Categories["POST_FILTER"] = 23] = "POST_FILTER";
    Categories[Categories["COMMENT_FILTER"] = 24] = "COMMENT_FILTER";
    Categories[Categories["ANTI_SPAM"] = 25] = "ANTI_SPAM";
    Categories[Categories["ANTI_HARASSMENT"] = 26] = "ANTI_HARASSMENT";
    Categories[Categories["NSFW_FILTER"] = 27] = "NSFW_FILTER";
    Categories[Categories["CITIZEN_FILTER"] = 28] = "CITIZEN_FILTER";
    Categories[Categories["POST_FLAIR"] = 29] = "POST_FLAIR";
    Categories[Categories["COMMENT_FLAIR"] = 30] = "COMMENT_FLAIR";
    Categories[Categories["USER_FLAIR"] = 31] = "USER_FLAIR";
    Categories[Categories["MESSAGING"] = 32] = "MESSAGING";
    Categories[Categories["MISCHIEF"] = 420] = "MISCHIEF";
    Categories[Categories["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Categories || (Categories = {}));
export function categoriesFromJSON(object) {
    switch (object) {
        case 0:
        case "MODERATION":
            return Categories.MODERATION;
        case 1:
        case "CUSTOM_POSTS":
            return Categories.CUSTOM_POSTS;
        case 2:
        case "UTILITIES":
            return Categories.UTILITIES;
        case 3:
        case "OTHER_SITES":
            return Categories.OTHER_SITES;
        case 4:
        case "FUNNY":
            return Categories.FUNNY;
        case 5:
        case "WEIRD":
            return Categories.WEIRD;
        case 6:
        case "AWW":
            return Categories.AWW;
        case 7:
        case "LEARNING":
            return Categories.LEARNING;
        case 8:
        case "FINANCIAL":
            return Categories.FINANCIAL;
        case 9:
        case "SPORTS":
            return Categories.SPORTS;
        case 10:
        case "GAMES":
            return Categories.GAMES;
        case 11:
        case "MOVIES":
            return Categories.MOVIES;
        case 12:
        case "MUSIC":
            return Categories.MUSIC;
        case 13:
        case "TV":
            return Categories.TV;
        case 14:
        case "BOOKS":
            return Categories.BOOKS;
        case 15:
        case "ANIME":
            return Categories.ANIME;
        case 16:
        case "PROGRAMMING":
            return Categories.PROGRAMMING;
        case 17:
        case "COOKING":
            return Categories.COOKING;
        case 18:
        case "FITNESS":
            return Categories.FITNESS;
        case 19:
        case "TRAVEL":
            return Categories.TRAVEL;
        case 20:
        case "CONTEXT_MENU":
            return Categories.CONTEXT_MENU;
        case 21:
        case "MOD_CONTEXT":
            return Categories.MOD_CONTEXT;
        case 22:
        case "FILTER":
            return Categories.FILTER;
        case 23:
        case "POST_FILTER":
            return Categories.POST_FILTER;
        case 24:
        case "COMMENT_FILTER":
            return Categories.COMMENT_FILTER;
        case 25:
        case "ANTI_SPAM":
            return Categories.ANTI_SPAM;
        case 26:
        case "ANTI_HARASSMENT":
            return Categories.ANTI_HARASSMENT;
        case 27:
        case "NSFW_FILTER":
            return Categories.NSFW_FILTER;
        case 28:
        case "CITIZEN_FILTER":
            return Categories.CITIZEN_FILTER;
        case 29:
        case "POST_FLAIR":
            return Categories.POST_FLAIR;
        case 30:
        case "COMMENT_FLAIR":
            return Categories.COMMENT_FLAIR;
        case 31:
        case "USER_FLAIR":
            return Categories.USER_FLAIR;
        case 32:
        case "MESSAGING":
            return Categories.MESSAGING;
        case 420:
        case "MISCHIEF":
            return Categories.MISCHIEF;
        case -1:
        case "UNRECOGNIZED":
        default:
            return Categories.UNRECOGNIZED;
    }
}
export function categoriesToJSON(object) {
    switch (object) {
        case Categories.MODERATION:
            return 0;
        case Categories.CUSTOM_POSTS:
            return 1;
        case Categories.UTILITIES:
            return 2;
        case Categories.OTHER_SITES:
            return 3;
        case Categories.FUNNY:
            return 4;
        case Categories.WEIRD:
            return 5;
        case Categories.AWW:
            return 6;
        case Categories.LEARNING:
            return 7;
        case Categories.FINANCIAL:
            return 8;
        case Categories.SPORTS:
            return 9;
        case Categories.GAMES:
            return 10;
        case Categories.MOVIES:
            return 11;
        case Categories.MUSIC:
            return 12;
        case Categories.TV:
            return 13;
        case Categories.BOOKS:
            return 14;
        case Categories.ANIME:
            return 15;
        case Categories.PROGRAMMING:
            return 16;
        case Categories.COOKING:
            return 17;
        case Categories.FITNESS:
            return 18;
        case Categories.TRAVEL:
            return 19;
        case Categories.CONTEXT_MENU:
            return 20;
        case Categories.MOD_CONTEXT:
            return 21;
        case Categories.FILTER:
            return 22;
        case Categories.POST_FILTER:
            return 23;
        case Categories.COMMENT_FILTER:
            return 24;
        case Categories.ANTI_SPAM:
            return 25;
        case Categories.ANTI_HARASSMENT:
            return 26;
        case Categories.NSFW_FILTER:
            return 27;
        case Categories.CITIZEN_FILTER:
            return 28;
        case Categories.POST_FLAIR:
            return 29;
        case Categories.COMMENT_FLAIR:
            return 30;
        case Categories.USER_FLAIR:
            return 31;
        case Categories.MESSAGING:
            return 32;
        case Categories.MISCHIEF:
            return 420;
        case Categories.UNRECOGNIZED:
        default:
            return -1;
    }
}
