// @devvit/protos/types/devvit/plugin/kvstore/kvstore.js
/**
 * #kvstore.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBaseMessageSet() {
    return { messages: {} };
}
export const MessageSet = {
    $type: "devvit.plugin.kvstore.MessageSet",
    encode(message, writer = _m0.Writer.create()) {
        Object.entries(message.messages).forEach(([key, value]) => {
            MessageSet_MessagesEntry.encode({ key: key, value }, writer.uint32(10).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMessageSet();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    const entry1 = MessageSet_MessagesEntry.decode(reader, reader.uint32());
                    if (entry1.value !== undefined) {
                        message.messages[entry1.key] = entry1.value;
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            messages: isObject(object.messages)
                ? Object.entries(object.messages).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        obj.messages = {};
        if (message.messages) {
            Object.entries(message.messages).forEach(([k, v]) => {
                obj.messages[k] = v;
            });
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMessageSet();
        message.messages = Object.entries(object.messages ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = String(value);
            }
            return acc;
        }, {});
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.kvstore.MessageSet", MessageSet);
function createBaseMessageSet_MessagesEntry() {
    return { key: "", value: "" };
}
export const MessageSet_MessagesEntry = {
    $type: "devvit.plugin.kvstore.MessageSet.MessagesEntry",
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMessageSet_MessagesEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.string();
                    break;
                case 2:
                    message.value = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { key: isSet(object.key) ? String(object.key) : "", value: isSet(object.value) ? String(object.value) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined && (obj.value = message.value);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseMessageSet_MessagesEntry();
        message.key = object.key ?? "";
        message.value = object.value ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.kvstore.MessageSet.MessagesEntry", MessageSet_MessagesEntry);
function createBaseKeySet() {
    return { keys: [] };
}
export const KeySet = {
    $type: "devvit.plugin.kvstore.KeySet",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.keys) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseKeySet();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.keys.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { keys: Array.isArray(object?.keys) ? object.keys.map((e) => String(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.keys) {
            obj.keys = message.keys.map((e) => e);
        }
        else {
            obj.keys = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseKeySet();
        message.keys = object.keys?.map((e) => e) || [];
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.kvstore.KeySet", KeySet);
function createBaseListFilter() {
    return { filter: "" };
}
export const ListFilter = {
    $type: "devvit.plugin.kvstore.ListFilter",
    encode(message, writer = _m0.Writer.create()) {
        if (message.filter !== "") {
            writer.uint32(10).string(message.filter);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.filter = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return { filter: isSet(object.filter) ? String(object.filter) : "" };
    },
    toJSON(message) {
        const obj = {};
        message.filter !== undefined && (obj.filter = message.filter);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseListFilter();
        message.filter = object.filter ?? "";
        return message;
    },
};
messageTypeRegistry.set("devvit.plugin.kvstore.ListFilter", ListFilter);
export class KVStoreClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.plugin.kvstore.KVStore";
        this.rpc = rpc;
        this.Put = this.Put.bind(this);
        this.Get = this.Get.bind(this);
        this.Del = this.Del.bind(this);
        this.List = this.List.bind(this);
    }
    Put(request) {
        const data = MessageSet.encode(request).finish();
        const promise = this.rpc.request(this.service, "Put", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    Get(request) {
        const data = KeySet.encode(request).finish();
        const promise = this.rpc.request(this.service, "Get", data);
        return promise.then((data) => MessageSet.decode(new _m0.Reader(data)));
    }
    Del(request) {
        const data = KeySet.encode(request).finish();
        const promise = this.rpc.request(this.service, "Del", data);
        return promise.then((data) => Empty.decode(new _m0.Reader(data)));
    }
    List(request) {
        const data = ListFilter.encode(request).finish();
        const promise = this.rpc.request(this.service, "List", data);
        return promise.then((data) => KeySet.decode(new _m0.Reader(data)));
    }
}
export const KVStoreDefinition = {
    name: "KVStore",
    fullName: "devvit.plugin.kvstore.KVStore",
    methods: {
        put: {
            name: "Put",
            requestType: MessageSet,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        get: {
            name: "Get",
            requestType: KeySet,
            requestStream: false,
            responseType: MessageSet,
            responseStream: false,
            options: {},
        },
        del: {
            name: "Del",
            requestType: KeySet,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        list: {
            name: "List",
            requestType: ListFilter,
            requestStream: false,
            responseType: KeySet,
            responseStream: false,
            options: {},
        },
    },
};
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
