// @devvit/shared-types/slug.js
export const APP_ACCOUNT_NAME_LENGTH_LIMIT = 20;
export const UNIQUE_SLUG_SUFFIX_LENGTH = 3;
// -1 for the dash between the slug and the suffix
export const APP_SLUG_BASE_MIN_LENGTH = 3;
export const APP_SLUG_BASE_MAX_LENGTH = APP_ACCOUNT_NAME_LENGTH_LIMIT - UNIQUE_SLUG_SUFFIX_LENGTH - 1;
const ALLOWED_SLUG_REGEX = /^[a-z][a-z0-9-]*$/;
export function sluggable(str) {
    const slug = str.replace(/\s+/g, '-').toLowerCase();
    if (slug.length < APP_SLUG_BASE_MIN_LENGTH || slug.length > APP_SLUG_BASE_MAX_LENGTH) {
        return false;
    }
    return ALLOWED_SLUG_REGEX.test(slug);
}
export function makeSlug(str) {
    if (!sluggable(str)) {
        const slug = str.replace(/\s+/g, '-').toLowerCase();
        let reason = '';
        if (slug.length < APP_SLUG_BASE_MIN_LENGTH) {
            reason = `name length must be longer than ${APP_SLUG_BASE_MIN_LENGTH} characters; length: ${slug.length}`;
        }
        else if (slug.length > APP_SLUG_BASE_MAX_LENGTH) {
            reason = `name length must be shorter than ${APP_SLUG_BASE_MAX_LENGTH} characters; length: ${slug.length}`;
        }
        else if (!ALLOWED_SLUG_REGEX.test(slug)) {
            reason = 'invalid characters; only use a-z, 0-9, and -';
        }
        throw new Error(`App name is not valid. (${reason})`);
    }
    return str.replace(/\s+/g, '-').toLowerCase();
}
