// @devvit/protos/types/devvit/actor/hello/hello.js
/**
 * #hello.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
function createBasePingMessage() {
    return { message: "", delayMillis: 0, successProbability: 0 };
}
export const PingMessage = {
    $type: "devvit.actor.hello.PingMessage",
    encode(message, writer = _m0.Writer.create()) {
        if (message.message !== "") {
            writer.uint32(10).string(message.message);
        }
        if (message.delayMillis !== 0) {
            writer.uint32(16).int32(message.delayMillis);
        }
        if (message.successProbability !== 0) {
            writer.uint32(29).float(message.successProbability);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePingMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.message = reader.string();
                    break;
                case 2:
                    message.delayMillis = reader.int32();
                    break;
                case 3:
                    message.successProbability = reader.float();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            message: isSet(object.message) ? String(object.message) : "",
            delayMillis: isSet(object.delayMillis) ? Number(object.delayMillis) : 0,
            successProbability: isSet(object.successProbability) ? Number(object.successProbability) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        message.delayMillis !== undefined && (obj.delayMillis = Math.round(message.delayMillis));
        message.successProbability !== undefined && (obj.successProbability = message.successProbability);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePingMessage();
        message.message = object.message ?? "";
        message.delayMillis = object.delayMillis ?? 0;
        message.successProbability = object.successProbability ?? 0;
        return message;
    },
};
messageTypeRegistry.set("devvit.actor.hello.PingMessage", PingMessage);
export class HelloClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || "devvit.actor.hello.Hello";
        this.rpc = rpc;
        this.Ping = this.Ping.bind(this);
    }
    Ping(request) {
        const data = PingMessage.encode(request).finish();
        const promise = this.rpc.request(this.service, "Ping", data);
        return promise.then((data) => PingMessage.decode(new _m0.Reader(data)));
    }
}
export const HelloDefinition = {
    name: "Hello",
    fullName: "devvit.actor.hello.Hello",
    methods: {
        ping: {
            name: "Ping",
            requestType: PingMessage,
            requestStream: false,
            responseType: PingMessage,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
