// @devvit/protos/types/devvit/runtime/actor/runtime.js
/**
 * #runtime.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from 'protobufjs/minimal.js';
import { Empty } from '../../../google/protobuf/empty.js';
import { messageTypeRegistry } from '../../../typeRegistry.js';
import { LinkedBundle } from '../bundle.js';
function createBaseWorkerQuery() {
    return { hostname: "" };
}
export const WorkerQuery = {
    $type: "devvit.runtime.actor.WorkerQuery",
    encode(message, writer = _m0.Writer.create()) {
        if (message.hostname !== "") {
            writer.uint32(10).string(message.hostname);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWorkerQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.hostname = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { hostname: isSet(object.hostname) ? globalThis.String(object.hostname) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.hostname !== "") {
            obj.hostname = message.hostname;
        }
        return obj;
    },
    create(base) {
        return WorkerQuery.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseWorkerQuery();
        message.hostname = object.hostname ?? "";
        return message;
    },
};
messageTypeRegistry.set(WorkerQuery.$type, WorkerQuery);
export const RuntimeActorServiceName = "devvit.runtime.actor.RuntimeActor";
export class RuntimeActorClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || RuntimeActorServiceName;
        this.rpc = rpc;
        this.NewWorker = this.NewWorker.bind(this);
        this.TerminateWorker = this.TerminateWorker.bind(this);
    }
    NewWorker(request, metadata) {
        const data = LinkedBundle.encode(request).finish();
        const promise = this.rpc.request(this.service, "NewWorker", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
    TerminateWorker(request, metadata) {
        const data = WorkerQuery.encode(request).finish();
        const promise = this.rpc.request(this.service, "TerminateWorker", data, metadata);
        return promise.then((data) => Empty.decode(_m0.Reader.create(data)));
    }
}
export const RuntimeActorDefinition = {
    name: "RuntimeActor",
    fullName: "devvit.runtime.actor.RuntimeActor",
    methods: {
        newWorker: {
            name: "NewWorker",
            requestType: LinkedBundle,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
        terminateWorker: {
            name: "TerminateWorker",
            requestType: WorkerQuery,
            requestStream: false,
            responseType: Empty,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
